{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Panel widgets exploiting MGnify API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import io\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    # clone the momics-demos repository to use the utils module from there\n",
    "    # TODO: eventually utils from momics will be used for that\n",
    "    try:\n",
    "        os.system('git clone https://github.com/palec87/momics-demos.git')\n",
    "        print(f\"Repository cloned\")\n",
    "    except OSError as e:\n",
    "        print(f\"An error occurred while cloning the repository: {e}\")\n",
    "\n",
    "    sys.path.insert(0,'/content/momics-demos')\n",
    "\n",
    "else:\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "\n",
    "from utils import init_setup, get_notebook_environment\n",
    "init_setup()\n",
    "\n",
    "# Initialize the environment variable\n",
    "notebook_environment = 'unknown'\n",
    "# Determine the notebook environment\n",
    "env = get_notebook_environment()\n",
    "print(f\"Environment: {env}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connection to MGnify API\n",
    "\n",
    "\n",
    "import os\n",
    "# this repo is not maintained, or less than jsonapi-requests\n",
    "# consider a dep change\n",
    "from jsonapi_client import Session as APISession\n",
    "from jsonapi_client import Modifier\n",
    "import requests\n",
    "import panel as pn\n",
    "\n",
    "# Dataframes and display\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Data transformation\n",
    "from functools import reduce\n",
    "\n",
    "# Plots\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "# import plotly.graph_objects as go\n",
    "%matplotlib inline\n",
    "\n",
    "import momics.plotting as pl\n",
    "\n",
    "# Create signature of MAGs for comparison against database\n",
    "# import sourmash\n",
    "# import glob\n",
    "# import time\n",
    "# from pathlib import PurePath as pp\n",
    "# from Bio import SeqIO\n",
    "\n",
    "# Warning verbosity\n",
    "import warnings \n",
    "warnings.filterwarnings(action=\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension()\n",
    "## Query and show endpoints ##\n",
    "##############################\n",
    "select_endpoint = pn.widgets.Select(\n",
    "    name=\"MGnify endpoints\",\n",
    "    value=\"\",\n",
    "    options=[],\n",
    "    description=\"Select endpoint to query\",\n",
    ")\n",
    "\n",
    "button0 = pn.widgets.Button(name='Get', button_type='primary')\n",
    "def update_endpoints(event):\n",
    "    if not event:\n",
    "        return\n",
    "    r = requests.get(f\"https://www.ebi.ac.uk/metagenomics/api/v1/\")\n",
    "    endpoint_dict = r.json()['data']\n",
    "    endpoints = [k for k in endpoint_dict.keys()]\n",
    "    select_endpoint.options = endpoints\n",
    "\n",
    "pn.bind(update_endpoints, button0, watch=True)\n",
    "pn.Row(button0, select_endpoint)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Query genomes ##\n",
    "###################\n",
    "styles = {\n",
    "    \"box-shadow\": \"rgba(50, 50, 93, 0.25) 0px 6px 12px -2px, rgba(0, 0, 0, 0.3) 0px 3px 7px -3px\",\n",
    "    \"border-radius\": \"4px\",\n",
    "    \"padding\": \"10px\",\n",
    "}\n",
    "pn.extension('tabulator')\n",
    "button1 = pn.widgets.Button(name='Query', button_type='primary')\n",
    "text_input = pn.widgets.TextInput(name='Text Input', placeholder='Enter a string here...')\n",
    "atable = pn.widgets.Tabulator(sizing_mode=\"stretch_both\", name=\"Data View\")\n",
    "\n",
    "def query_genomes(event):\n",
    "    if not event:\n",
    "        return\n",
    "    with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        search_filter = Modifier(f\"taxon_lineage={text_input.value}\")\n",
    "        resources = map(lambda r: r.json, mgnify.iterate(select_endpoint.value, filter=search_filter))\n",
    "        resources_df = pd.json_normalize(resources)\n",
    "\n",
    "    print('Queried', text_input.value, \"from\", select_endpoint.value)\n",
    "\n",
    "    # update table view\n",
    "    atable.value = resources_df\n",
    "    \n",
    "    # create data folder if it doesn't exist\n",
    "    os.system(\"mkdir -p data\")\n",
    "    # save to parquet\n",
    "    resources_df.to_parquet(os.path.join(\"data\", f\"{select_endpoint.value}_{text_input.value}.parquet\"))\n",
    "\n",
    "    # this is alert for the dsahboard, TODO: needs to be tested\n",
    "    pn.pane.Alert('## Data saved to data folder ##', alert_type='success', width=500).servable()\n",
    "\n",
    "\n",
    "pn.bind(query_genomes, button1, watch=True)\n",
    "tabs = pn.Tabs(\n",
    "    ('select', pn.Row(button1, text_input)),\n",
    "    ('view', atable),\n",
    "    styles=styles, sizing_mode=\"stretch_width\", height=500, margin=10\n",
    ")\n",
    "\n",
    "tabs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data transformation\n",
    "from functools import reduce\n",
    "\n",
    "def mgn_split_taxonomy(df):\n",
    "    features = ['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "    # all_genomes_df = atable.value\n",
    "\n",
    "    # Split the 'attributes.taxon-lineage' column and create new columns\n",
    "    lineage_split = df['attributes.taxon-lineage'].str.split(';', expand=True)\n",
    "    lineage_split.columns = features\n",
    "\n",
    "    # Concatenate the original DataFrame with the new columns\n",
    "    all_genomes_tax_df = pd.concat([df, lineage_split], axis=1)\n",
    "    return features, all_genomes_tax_df\n",
    "\n",
    "features, taxonomy_df = mgn_split_taxonomy(atable.value)\n",
    "sankey_df = taxonomy_df.groupby(features).size().reset_index(name='count')\n",
    "\n",
    "sankey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension()\n",
    "\n",
    "# panel plot widget with plotly sankey\n",
    "sankey = pl.get_sankey(sankey_df, cat_cols=features[0:6],\n",
    "                       value_cols='count', title='Taxon Lineage')\n",
    "\n",
    "# panel pane for plotly sankey\n",
    "sankey_pane = pn.pane.Plotly(sankey, config={'displayModeBar': False})\n",
    "sankey_pane"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momicsdem",
   "language": "python",
   "name": "momicsdem"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
