{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Panel dashboard to query and display MGnify data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Platform dependent part\n",
    "- Resolve platform setup\n",
    "- the difference to local imports should be resolved by setting the Blue Cloud VRE well, Colab will still be an issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import logging\n",
    "from IPython import get_ipython\n",
    "logger = logging.getLogger(name=\"MGnify API\")\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    print('Setting Google colab, you will need a ngrok account to make the dashboard display over the tunnel. \\\n",
    "    https://ngrok.com/')\n",
    "    # clone the momics-demos repository to use it to load data\n",
    "    try:\n",
    "        os.system('git clone https://github.com/palec87/momics-demos.git')\n",
    "        logger.info(f\"Repository cloned\")\n",
    "    except OSError as e:\n",
    "        logger.info(f\"An error occurred while cloning the repository: {e}\")\n",
    "\n",
    "    sys.path.insert(0,'/content/momics-demos')\n",
    "\n",
    "    # this step takes time beacause of many dependencies\n",
    "    os.system('pip install marine-omics')\n",
    "\n",
    "from momics.utils import (\n",
    "    memory_load, reconfig_logger,\n",
    "    init_setup, get_notebook_environment,\n",
    "    load_and_clean, taxonomy_common_preprocess01,\n",
    ")\n",
    "\n",
    "# Set up logging\n",
    "reconfig_logger()\n",
    "\n",
    "# Determine the notebook environment\n",
    "env = get_notebook_environment()\n",
    "\n",
    "init_setup()\n",
    "logger.info(f\"Environment: {env}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This needs to be repeated here for the Pannel dashboard to work, WEIRD\n",
    "# TODO: report as possible bug\n",
    "import sys\n",
    "import os\n",
    "import io\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "from jsonapi_client import Session as APISession\n",
    "from jsonapi_client import Modifier\n",
    "import requests\n",
    "\n",
    "# Dataframes and display\n",
    "import panel as pn\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "# Data transformation\n",
    "from functools import reduce\n",
    "\n",
    "# Plots\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.graph_objects as go\n",
    "# %matplotlib inline \n",
    "\n",
    "import momics.plotting as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parquet files\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    root_folder = os.path.abspath(os.path.join('/content/momics-demos'))\n",
    "else:\n",
    "    root_folder = os.path.abspath(os.path.join('../'))\n",
    "\n",
    "\n",
    "data_folder = os.path.join(root_folder, 'wf5_MGnify/data/parquet_files')\n",
    "assets_folder = os.path.join(root_folder, 'assets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mgn_split_taxonomy(df):\n",
    "    features = ['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "    # all_genomes_df = atable.value\n",
    "\n",
    "    # Split the 'attributes.taxon-lineage' column and create new columns\n",
    "    lineage_split = df['attributes.taxon-lineage'].str.split(';', expand=True)\n",
    "    lineage_split.columns = features\n",
    "\n",
    "    # Concatenate the original DataFrame with the new columns\n",
    "    all_genomes_tax_df = pd.concat([df, lineage_split], axis=1)\n",
    "    return features, all_genomes_tax_df\n",
    "\n",
    "# sankey_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## App setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "xx = np.linspace(-3.5, 3.5, 100)\n",
    "yy = np.linspace(-3.5, 3.5, 100)\n",
    "x, y = np.meshgrid(xx, yy)\n",
    "z = np.exp(-((x - 1) ** 2) - y**2) - (x**3 + y**4 - x / 5) * np.exp(-(x**2 + y**2))\n",
    "\n",
    "surface=go.Surface(z=z)\n",
    "fig = go.Figure(data=[surface])\n",
    "\n",
    "fig.update_layout(\n",
    "    title=\"Plotly 3D Plot\",\n",
    "    width=500,\n",
    "    height=500,\n",
    "    margin=dict(t=50, b=50, r=50, l=50),\n",
    ")\n",
    "plotly_pane = pn.pane.Plotly(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension(\"tabulator\")\n",
    "pn.extension(\"plotly\")\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    pn.extension(comms='colab')\n",
    "ACCENT = \"teal\"\n",
    "\n",
    "styles = {\n",
    "    \"box-shadow\": \"rgba(50, 50, 93, 0.25) 0px 6px 12px -2px, rgba(0, 0, 0, 0.3) 0px 3px 7px -3px\",\n",
    "    \"border-radius\": \"4px\",\n",
    "    \"padding\": \"10px\",\n",
    "}\n",
    "\n",
    "# TODO: there is a bug in the panel library that does not allow to open png files, renoming does not help \n",
    "image = pn.pane.PNG(os.path.join(assets_folder, \"figs/mgnify_banner.png\"),\n",
    "                    width=300,\n",
    "                    # height=100,\n",
    "                    )\n",
    "\n",
    "##############################\n",
    "## Query and show endpoints ##\n",
    "##############################\n",
    "# TODO: this goes to the utils\n",
    "select_endpoint = pn.widgets.Select(\n",
    "    name=\"MGnify endpoints\",\n",
    "    value=\"\",\n",
    "    options=[],\n",
    "    description=\"Select endpoint to query\",\n",
    ")\n",
    "\n",
    "r = requests.get(f\"https://www.ebi.ac.uk/metagenomics/api/v1/\")\n",
    "endpoint_dict = r.json()['data']\n",
    "endpoints = [k for k in endpoint_dict.keys()]\n",
    "select_endpoint.options = endpoints\n",
    "select_endpoint.value = endpoints[0]\n",
    "\n",
    "###################\n",
    "## Query genomes ##\n",
    "###################\n",
    "button1 = pn.widgets.Button(name='Query', button_type='primary')\n",
    "text_input = pn.widgets.TextInput(name='Text Input', placeholder='Enter a string here...')\n",
    "atable = pn.widgets.Tabulator(sizing_mode=\"stretch_both\", name=\"Data View\")\n",
    "\n",
    "def query_genomes(event):\n",
    "    if not event:\n",
    "        return\n",
    "    with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        search_filter = Modifier(f\"taxon_lineage={text_input.value}\")\n",
    "        resources = map(lambda r: r.json, mgnify.iterate(select_endpoint.value, filter=search_filter))\n",
    "        resources_df = pd.json_normalize(resources)\n",
    "        logger.info(f\"Queried {len(resources_df)} resources from {select_endpoint.value} with filter {text_input.value}\")\n",
    "\n",
    "    print('Queried', text_input.value, \"from\", select_endpoint.value)\n",
    "\n",
    "    # update table view\n",
    "    atable.value = resources_df\n",
    "    \n",
    "    # create data folder if it doesn't exist\n",
    "    os.system(\"mkdir -p data\")\n",
    "    # save to parquet\n",
    "    resources_df.to_parquet(os.path.join(\"data\", f\"{select_endpoint.value}_{text_input.value}.parquet\"))\n",
    "\n",
    "    # this is alert for the dsahboard, TODO: needs to be tested\n",
    "    pn.pane.Alert('## Data saved to data folder ##', alert_type='success', width=500).servable()\n",
    "\n",
    "pn.bind(query_genomes, button1, watch=True)\n",
    "\n",
    "#################\n",
    "## Plot sankey ##\n",
    "#################\n",
    "button2 = pn.widgets.Button(name='Display Sankey', button_type='primary')\n",
    "\n",
    "def plot_sankey(event):\n",
    "    if not event:\n",
    "        print('no event')\n",
    "        return\n",
    "    print('updating sankey plot')\n",
    "    features, taxonomy_df = mgn_split_taxonomy(atable.value)\n",
    "    sankey_df = taxonomy_df.groupby(features).size().reset_index(name='count')\n",
    "    # panel plot widget with plotly sankey\n",
    "    sankey = pl.get_sankey(sankey_df, cat_cols=features[0:6],\n",
    "                           value_cols='count', title='Taxon Lineage')\n",
    "\n",
    "    # panel pane for plotly sankey\n",
    "    display(sankey_df)\n",
    "    sankey_pane = pn.pane.Plotly(sankey, config={'displayModeBar': False})\n",
    "    tabs.objects[1] = sankey_pane\n",
    "\n",
    "\n",
    "pn.bind(plot_sankey, button2, watch=True)\n",
    "\n",
    "sankey_pane = pn.pane.Plotly()\n",
    "tabs = pn.Tabs(\n",
    "    ('Table view', atable),\n",
    "    ('Sankey plot', sankey_pane),\n",
    "    styles=styles, sizing_mode=\"stretch_width\", height=500, margin=10,\n",
    "    dynamic=True,\n",
    ")\n",
    "pn.bind(plot_sankey, tabs, watch=True)\n",
    "\n",
    "# template\n",
    "template = pn.template.FastListTemplate(\n",
    "    title=\"Query MGnify\",\n",
    "    sidebar=[image,\n",
    "             \"# Endpoint\", select_endpoint,\n",
    "             \"works only for 'genomes' endpoint and lineage queries\",\n",
    "             pn.layout.Divider(),\n",
    "             \"# Query\", pn.Column(text_input, button1),\n",
    "             \"# Plot\", pn.Row(button2),\n",
    "             ],\n",
    "    main=[pn.Column(\n",
    "        tabs,\n",
    "        sizing_mode=\"stretch_both\",\n",
    "    )],\n",
    "    main_layout=None,\n",
    "    accent=ACCENT,\n",
    ")\n",
    "\n",
    "if \"google.colab\" in str(get_ipython()) or env == \"vscode\":\n",
    "    server=pn.serve({\"\": template}, port=4040, address=\"127.0.0.1\", threaded=True, websocket_origin=\"*\")\n",
    "    os.system(\"curl http://localhost:4040\")\n",
    "    from pyngrok import ngrok\n",
    "\n",
    "    # Terminate open tunnels if exist\n",
    "    ngrok.kill()\n",
    "\n",
    "    # Setting the authtoken, get yours from https://dashboard.ngrok.com/auth\n",
    "    NGROK_AUTH_TOKEN = os.getenv(\"NGROK_TOKEN\")\n",
    "    ngrok.set_auth_token(NGROK_AUTH_TOKEN)\n",
    "\n",
    "    # Open an HTTPs tunnel on port 4040 for http://localhost:4040\n",
    "    if env == \"vscode\":\n",
    "        public_url = ngrok.connect(addr='4040')\n",
    "    else:\n",
    "        public_url = ngrok.connect(port='4040')\n",
    "    \n",
    "    print(\"Tracking URL:\", public_url)\n",
    "else:\n",
    "    print('serving locally')\n",
    "    template.servable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# server.stop()\n",
    "# ngrok.disconnect(server)\n",
    "# ngrok.kill()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
