{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f53f4f5e",
   "metadata": {},
   "source": [
    "## Example from MGnify summary data\n",
    "\n",
    "1. Study MGYS00006608, 16S rRNA amplicon sequencing from the Ocean Sampling Day (OSD) campaign June 2018 (ERP124424_taxonomy_abundances_SSU_v5.0)\n",
    "2. Study MGYS00006607, 16S rRNA amplicon sequencing from the Ocean Sampling Day (OSD) campaign June 2019 (ERP124431_taxonomy_abundances_SSU_v5.0)\n",
    "3. Study MGYS00000492, Amplicon sequencing of Tara Oceans DNA samples corresponding to size fractions for prokaryotes or protist. (ERP003634_taxonomy_abundances_SSU_v5.0)\n",
    "4. Study MGYS00006680, SOLA sampling point Raw sequence reads (SRP237882_taxonomy_abundances_SSU_v5.0)\n",
    "5. Study MGYS00006682, Vertical stratification of environmental DNA in the open ocean captures ecological patterns and behavior of deep-sea fishes (SRP334933_taxonomy_abundances_SSU_v5.0)\n",
    "6. Study MGYS00006678, Dataset on spatiotemporal variation of microbial plankton communities in the Baltic Sea (ERP140185_taxonomy_abundances_SSU_v5.0)\n",
    "7. Study MGYS00006675, 16S rRNA gene amplicon time-series in Blanes Bay Microbial Observatory (BBMO) (ERP122219_taxonomy_abundances_SSU_v5.0)\n",
    "8. Study MGYS00003725, Arctic microbiome along Svalbard Cross Shelf transects (ERP106348_taxonomy_abundances_SSU_v5.0)\n",
    "9. Study MGYS00006686, Environmental DNA and zooplankton samples taken at Helgoland Roads in June 2019 (ERP144826_taxonomy_abundances_SSU_v5.0)\n",
    "10. Study MGYS00006714, Regional and vertical patterns in microbial communities across Fram Strait (2015-2019) (ERP151329_taxonomy_abundances_SSU_v5.0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "81f0f2e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"esms-options\">{\"shimMode\": true}</script><style>*[data-root-id],\n",
       "*[data-root-id] > * {\n",
       "  box-sizing: border-box;\n",
       "  font-family: var(--jp-ui-font-family);\n",
       "  font-size: var(--jp-ui-font-size1);\n",
       "  color: var(--vscode-editor-foreground, var(--jp-ui-font-color1));\n",
       "}\n",
       "\n",
       "/* Override VSCode background color */\n",
       ".cell-output-ipywidget-background:has(\n",
       "    > .cell-output-ipywidget-background > .lm-Widget > *[data-root-id]\n",
       "  ),\n",
       ".cell-output-ipywidget-background:has(> .lm-Widget > *[data-root-id]) {\n",
       "  background-color: transparent !important;\n",
       "}\n",
       "</style>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n  const py_version = '3.7.3'.replace('rc', '-rc.').replace('.dev', '-dev.');\n  const reloading = false;\n  const Bokeh = root.Bokeh;\n\n  // Set a timeout for this load but only if we are not already initializing\n  if (typeof (root._bokeh_timeout) === \"undefined\" || (force || !root._bokeh_is_initializing)) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks;\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, js_modules, js_exports, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n    if (js_modules == null) js_modules = [];\n    if (js_exports == null) js_exports = {};\n\n    root._bokeh_onload_callbacks.push(callback);\n\n    if (root._bokeh_is_loading > 0) {\n      // Don't load bokeh if it is still initializing\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    } else if (js_urls.length === 0 && js_modules.length === 0 && Object.keys(js_exports).length === 0) {\n      // There is nothing to load\n      run_callbacks();\n      return null;\n    }\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n    window._bokeh_on_load = on_load\n\n    function on_error(e) {\n      const src_el = e.srcElement\n      console.error(\"failed to load \" + (src_el.href || src_el.src));\n    }\n\n    const skip = [];\n    if (window.requirejs) {\n      window.requirejs.config({'packages': {}, 'paths': {}, 'shim': {}});\n      root._bokeh_is_loading = css_urls.length + 0;\n    } else {\n      root._bokeh_is_loading = css_urls.length + js_urls.length + js_modules.length + Object.keys(js_exports).length;\n    }\n\n    const existing_stylesheets = []\n    const links = document.getElementsByTagName('link')\n    for (let i = 0; i < links.length; i++) {\n      const link = links[i]\n      if (link.href != null) {\n        existing_stylesheets.push(link.href)\n      }\n    }\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const escaped = encodeURI(url)\n      if (existing_stylesheets.indexOf(escaped) !== -1) {\n        on_load()\n        continue;\n      }\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }    var existing_scripts = []\n    const scripts = document.getElementsByTagName('script')\n    for (let i = 0; i < scripts.length; i++) {\n      var script = scripts[i]\n      if (script.src != null) {\n        existing_scripts.push(script.src)\n      }\n    }\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const escaped = encodeURI(url)\n      if (skip.indexOf(escaped) !== -1 || existing_scripts.indexOf(escaped) !== -1) {\n        if (!window.requirejs) {\n          on_load();\n        }\n        continue;\n      }\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (let i = 0; i < js_modules.length; i++) {\n      const url = js_modules[i];\n      const escaped = encodeURI(url)\n      if (skip.indexOf(escaped) !== -1 || existing_scripts.indexOf(escaped) !== -1) {\n        if (!window.requirejs) {\n          on_load();\n        }\n        continue;\n      }\n      var element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error;\n      element.async = false;\n      element.src = url;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n    for (const name in js_exports) {\n      const url = js_exports[name];\n      const escaped = encodeURI(url)\n      if (skip.indexOf(escaped) >= 0 || root[name] != null) {\n        if (!window.requirejs) {\n          on_load();\n        }\n        continue;\n      }\n      var element = document.createElement('script');\n      element.onerror = on_error;\n      element.async = false;\n      element.type = \"module\";\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      element.textContent = `\n      import ${name} from \"${url}\"\n      window.${name} = ${name}\n      window._bokeh_on_load()\n      `\n      document.head.appendChild(element);\n    }\n    if (!js_urls.length && !js_modules.length) {\n      on_load()\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.holoviz.org/panel/1.7.0/dist/bundled/reactiveesm/es-module-shims@^1.10.0/dist/es-module-shims.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-3.7.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.7.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.7.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.7.3.min.js\", \"https://cdn.holoviz.org/panel/1.7.0/dist/panel.min.js\"];\n  const js_modules = [];\n  const js_exports = {};\n  const css_urls = [];\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {} // ensure no trailing comma for IE\n  ];\n\n  function run_inline_js() {\n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (let i = 0; i < inline_js.length; i++) {\n        try {\n          inline_js[i].call(root, root.Bokeh);\n        } catch(e) {\n          if (!reloading) {\n            throw e;\n          }\n        }\n      }\n      // Cache old bokeh versions\n      if (Bokeh != undefined && !reloading) {\n        var NewBokeh = root.Bokeh;\n        if (Bokeh.versions === undefined) {\n          Bokeh.versions = new Map();\n        }\n        if (NewBokeh.version !== Bokeh.version) {\n          Bokeh.versions.set(NewBokeh.version, NewBokeh)\n        }\n        root.Bokeh = Bokeh;\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    }\n    root._bokeh_is_initializing = false\n  }\n\n  function load_or_wait() {\n    // Implement a backoff loop that tries to ensure we do not load multiple\n    // versions of Bokeh and its dependencies at the same time.\n    // In recent versions we use the root._bokeh_is_initializing flag\n    // to determine whether there is an ongoing attempt to initialize\n    // bokeh, however for backward compatibility we also try to ensure\n    // that we do not start loading a newer (Panel>=1.0 and Bokeh>3) version\n    // before older versions are fully initialized.\n    if (root._bokeh_is_initializing && Date.now() > root._bokeh_timeout) {\n      // If the timeout and bokeh was not successfully loaded we reset\n      // everything and try loading again\n      root._bokeh_timeout = Date.now() + 5000;\n      root._bokeh_is_initializing = false;\n      root._bokeh_onload_callbacks = undefined;\n      root._bokeh_is_loading = 0\n      console.log(\"Bokeh: BokehJS was loaded multiple times but one version failed to initialize.\");\n      load_or_wait();\n    } else if (root._bokeh_is_initializing || (typeof root._bokeh_is_initializing === \"undefined\" && root._bokeh_onload_callbacks !== undefined)) {\n      setTimeout(load_or_wait, 100);\n    } else {\n      root._bokeh_is_initializing = true\n      root._bokeh_onload_callbacks = []\n      const bokeh_loaded = root.Bokeh != null && (root.Bokeh.version === py_version || (root.Bokeh.versions !== undefined && root.Bokeh.versions.has(py_version)));\n      if (!reloading && !bokeh_loaded) {\n        if (root.Bokeh) {\n          root.Bokeh = undefined;\n        }\n        console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n      }\n      load_libs(css_urls, js_urls, js_modules, js_exports, function() {\n        console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n        run_inline_js();\n      });\n    }\n  }\n  // Give older versions of the autoload script a head-start to ensure\n  // they initialize before we start loading newer version.\n  setTimeout(load_or_wait, 100)\n}(window));",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "\nif ((window.PyViz === undefined) || (window.PyViz instanceof HTMLElement)) {\n  window.PyViz = {comms: {}, comm_status:{}, kernels:{}, receivers: {}, plot_index: []}\n}\n\n\n    function JupyterCommManager() {\n    }\n\n    JupyterCommManager.prototype.register_target = function(plot_id, comm_id, msg_handler) {\n      if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        comm_manager.register_target(comm_id, function(comm) {\n          comm.on_msg(msg_handler);\n        });\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        window.PyViz.kernels[plot_id].registerCommTarget(comm_id, function(comm) {\n          comm.onMsg = msg_handler;\n        });\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        google.colab.kernel.comms.registerTarget(comm_id, (comm) => {\n          var messages = comm.messages[Symbol.asyncIterator]();\n          function processIteratorResult(result) {\n            var message = result.value;\n            console.log(message)\n            var content = {data: message.data, comm_id};\n            var buffers = []\n            for (var buffer of message.buffers || []) {\n              buffers.push(new DataView(buffer))\n            }\n            var metadata = message.metadata || {};\n            var msg = {content, buffers, metadata}\n            msg_handler(msg);\n            return messages.next().then(processIteratorResult);\n          }\n          return messages.next().then(processIteratorResult);\n        })\n      }\n    }\n\n    JupyterCommManager.prototype.get_client_comm = function(plot_id, comm_id, msg_handler) {\n      if (comm_id in window.PyViz.comms) {\n        return window.PyViz.comms[comm_id];\n      } else if (window.comm_manager || ((window.Jupyter !== undefined) && (Jupyter.notebook.kernel != null))) {\n        var comm_manager = window.comm_manager || Jupyter.notebook.kernel.comm_manager;\n        var comm = comm_manager.new_comm(comm_id, {}, {}, {}, comm_id);\n        if (msg_handler) {\n          comm.on_msg(msg_handler);\n        }\n      } else if ((plot_id in window.PyViz.kernels) && (window.PyViz.kernels[plot_id])) {\n        var comm = window.PyViz.kernels[plot_id].connectToComm(comm_id);\n        comm.open();\n        if (msg_handler) {\n          comm.onMsg = msg_handler;\n        }\n      } else if (typeof google != 'undefined' && google.colab.kernel != null) {\n        var comm_promise = google.colab.kernel.comms.open(comm_id)\n        comm_promise.then((comm) => {\n          window.PyViz.comms[comm_id] = comm;\n          if (msg_handler) {\n            var messages = comm.messages[Symbol.asyncIterator]();\n            function processIteratorResult(result) {\n              var message = result.value;\n              var content = {data: message.data};\n              var metadata = message.metadata || {comm_id};\n              var msg = {content, metadata}\n              msg_handler(msg);\n              return messages.next().then(processIteratorResult);\n            }\n            return messages.next().then(processIteratorResult);\n          }\n        })\n        var sendClosure = (data, metadata, buffers, disposeOnDone) => {\n          return comm_promise.then((comm) => {\n            comm.send(data, metadata, buffers, disposeOnDone);\n          });\n        };\n        var comm = {\n          send: sendClosure\n        };\n      }\n      window.PyViz.comms[comm_id] = comm;\n      return comm;\n    }\n    window.PyViz.comm_manager = new JupyterCommManager();\n    \n\n\nvar JS_MIME_TYPE = 'application/javascript';\nvar HTML_MIME_TYPE = 'text/html';\nvar EXEC_MIME_TYPE = 'application/vnd.holoviews_exec.v0+json';\nvar CLASS_NAME = 'output';\n\n/**\n * Render data to the DOM node\n */\nfunction render(props, node) {\n  var div = document.createElement(\"div\");\n  var script = document.createElement(\"script\");\n  node.appendChild(div);\n  node.appendChild(script);\n}\n\n/**\n * Handle when a new output is added\n */\nfunction handle_add_output(event, handle) {\n  var output_area = handle.output_area;\n  var output = handle.output;\n  if ((output.data == undefined) || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n    return\n  }\n  var id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n  var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n  if (id !== undefined) {\n    var nchildren = toinsert.length;\n    var html_node = toinsert[nchildren-1].children[0];\n    html_node.innerHTML = output.data[HTML_MIME_TYPE];\n    var scripts = [];\n    var nodelist = html_node.querySelectorAll(\"script\");\n    for (var i in nodelist) {\n      if (nodelist.hasOwnProperty(i)) {\n        scripts.push(nodelist[i])\n      }\n    }\n\n    scripts.forEach( function (oldScript) {\n      var newScript = document.createElement(\"script\");\n      var attrs = [];\n      var nodemap = oldScript.attributes;\n      for (var j in nodemap) {\n        if (nodemap.hasOwnProperty(j)) {\n          attrs.push(nodemap[j])\n        }\n      }\n      attrs.forEach(function(attr) { newScript.setAttribute(attr.name, attr.value) });\n      newScript.appendChild(document.createTextNode(oldScript.innerHTML));\n      oldScript.parentNode.replaceChild(newScript, oldScript);\n    });\n    if (JS_MIME_TYPE in output.data) {\n      toinsert[nchildren-1].children[1].textContent = output.data[JS_MIME_TYPE];\n    }\n    output_area._hv_plot_id = id;\n    if ((window.Bokeh !== undefined) && (id in Bokeh.index)) {\n      window.PyViz.plot_index[id] = Bokeh.index[id];\n    } else {\n      window.PyViz.plot_index[id] = null;\n    }\n  } else if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n    var bk_div = document.createElement(\"div\");\n    bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n    var script_attrs = bk_div.children[0].attributes;\n    for (var i = 0; i < script_attrs.length; i++) {\n      toinsert[toinsert.length - 1].childNodes[1].setAttribute(script_attrs[i].name, script_attrs[i].value);\n    }\n    // store reference to server id on output_area\n    output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n  }\n}\n\n/**\n * Handle when an output is cleared or removed\n */\nfunction handle_clear_output(event, handle) {\n  var id = handle.cell.output_area._hv_plot_id;\n  var server_id = handle.cell.output_area._bokeh_server_id;\n  if (((id === undefined) || !(id in PyViz.plot_index)) && (server_id !== undefined)) { return; }\n  var comm = window.PyViz.comm_manager.get_client_comm(\"hv-extension-comm\", \"hv-extension-comm\", function () {});\n  if (server_id !== null) {\n    comm.send({event_type: 'server_delete', 'id': server_id});\n    return;\n  } else if (comm !== null) {\n    comm.send({event_type: 'delete', 'id': id});\n  }\n  delete PyViz.plot_index[id];\n  if ((window.Bokeh !== undefined) & (id in window.Bokeh.index)) {\n    var doc = window.Bokeh.index[id].model.document\n    doc.clear();\n    const i = window.Bokeh.documents.indexOf(doc);\n    if (i > -1) {\n      window.Bokeh.documents.splice(i, 1);\n    }\n  }\n}\n\n/**\n * Handle kernel restart event\n */\nfunction handle_kernel_cleanup(event, handle) {\n  delete PyViz.comms[\"hv-extension-comm\"];\n  window.PyViz.plot_index = {}\n}\n\n/**\n * Handle update_display_data messages\n */\nfunction handle_update_output(event, handle) {\n  handle_clear_output(event, {cell: {output_area: handle.output_area}})\n  handle_add_output(event, handle)\n}\n\nfunction register_renderer(events, OutputArea) {\n  function append_mime(data, metadata, element) {\n    // create a DOM node to render to\n    var toinsert = this.create_output_subarea(\n    metadata,\n    CLASS_NAME,\n    EXEC_MIME_TYPE\n    );\n    this.keyboard_manager.register_events(toinsert);\n    // Render to node\n    var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n    render(props, toinsert[0]);\n    element.append(toinsert);\n    return toinsert\n  }\n\n  events.on('output_added.OutputArea', handle_add_output);\n  events.on('output_updated.OutputArea', handle_update_output);\n  events.on('clear_output.CodeCell', handle_clear_output);\n  events.on('delete.Cell', handle_clear_output);\n  events.on('kernel_ready.Kernel', handle_kernel_cleanup);\n\n  OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n    safe: true,\n    index: 0\n  });\n}\n\nif (window.Jupyter !== undefined) {\n  try {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  } catch(err) {\n  }\n}\n",
      "application/vnd.holoviews_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.holoviews_exec.v0+json": "",
      "text/html": [
       "<div id='cebecef4-98fc-4147-811c-fe92adaf228c'>\n",
       "  <div id=\"e24676cf-f716-472e-a708-7cf0712dd91f\" data-root-id=\"cebecef4-98fc-4147-811c-fe92adaf228c\" style=\"display: contents;\"></div>\n",
       "</div>\n",
       "<script type=\"application/javascript\">(function(root) {\n",
       "  var docs_json = {\"800a3824-bf39-4438-b4e0-cfd80584fe91\":{\"version\":\"3.7.3\",\"title\":\"Bokeh Application\",\"roots\":[{\"type\":\"object\",\"name\":\"panel.models.browser.BrowserInfo\",\"id\":\"cebecef4-98fc-4147-811c-fe92adaf228c\"},{\"type\":\"object\",\"name\":\"panel.models.comm_manager.CommManager\",\"id\":\"024ca9b5-0502-46e2-83ac-dad8b8eacf16\",\"attributes\":{\"plot_id\":\"cebecef4-98fc-4147-811c-fe92adaf228c\",\"comm_id\":\"f25c1ae918be4eef8a9d52e159b3883f\",\"client_comm_id\":\"3fe30aea561f4c0792dc021e133d6764\"}}],\"defs\":[{\"type\":\"model\",\"name\":\"ReactiveHTML1\"},{\"type\":\"model\",\"name\":\"FlexBox1\",\"properties\":[{\"name\":\"align_content\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"align_items\",\"kind\":\"Any\",\"default\":\"flex-start\"},{\"name\":\"flex_direction\",\"kind\":\"Any\",\"default\":\"row\"},{\"name\":\"flex_wrap\",\"kind\":\"Any\",\"default\":\"wrap\"},{\"name\":\"gap\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"justify_content\",\"kind\":\"Any\",\"default\":\"flex-start\"}]},{\"type\":\"model\",\"name\":\"FloatPanel1\",\"properties\":[{\"name\":\"config\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"contained\",\"kind\":\"Any\",\"default\":true},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"right-top\"},{\"name\":\"offsetx\",\"kind\":\"Any\",\"default\":null},{\"name\":\"offsety\",\"kind\":\"Any\",\"default\":null},{\"name\":\"theme\",\"kind\":\"Any\",\"default\":\"primary\"},{\"name\":\"status\",\"kind\":\"Any\",\"default\":\"normalized\"}]},{\"type\":\"model\",\"name\":\"GridStack1\",\"properties\":[{\"name\":\"ncols\",\"kind\":\"Any\",\"default\":null},{\"name\":\"nrows\",\"kind\":\"Any\",\"default\":null},{\"name\":\"allow_resize\",\"kind\":\"Any\",\"default\":true},{\"name\":\"allow_drag\",\"kind\":\"Any\",\"default\":true},{\"name\":\"state\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"drag1\",\"properties\":[{\"name\":\"slider_width\",\"kind\":\"Any\",\"default\":5},{\"name\":\"slider_color\",\"kind\":\"Any\",\"default\":\"black\"},{\"name\":\"value\",\"kind\":\"Any\",\"default\":50}]},{\"type\":\"model\",\"name\":\"click1\",\"properties\":[{\"name\":\"terminal_output\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"debug_name\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"clears\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"FastWrapper1\",\"properties\":[{\"name\":\"object\",\"kind\":\"Any\",\"default\":null},{\"name\":\"style\",\"kind\":\"Any\",\"default\":null}]},{\"type\":\"model\",\"name\":\"NotificationArea1\",\"properties\":[{\"name\":\"js_events\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}},{\"name\":\"max_notifications\",\"kind\":\"Any\",\"default\":5},{\"name\":\"notifications\",\"kind\":\"Any\",\"default\":[]},{\"name\":\"position\",\"kind\":\"Any\",\"default\":\"bottom-right\"},{\"name\":\"_clear\",\"kind\":\"Any\",\"default\":0},{\"name\":\"types\",\"kind\":\"Any\",\"default\":[{\"type\":\"map\",\"entries\":[[\"type\",\"warning\"],[\"background\",\"#ffc107\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-exclamation-triangle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]},{\"type\":\"map\",\"entries\":[[\"type\",\"info\"],[\"background\",\"#007bff\"],[\"icon\",{\"type\":\"map\",\"entries\":[[\"className\",\"fas fa-info-circle\"],[\"tagName\",\"i\"],[\"color\",\"white\"]]}]]}]}]},{\"type\":\"model\",\"name\":\"Notification\",\"properties\":[{\"name\":\"background\",\"kind\":\"Any\",\"default\":null},{\"name\":\"duration\",\"kind\":\"Any\",\"default\":3000},{\"name\":\"icon\",\"kind\":\"Any\",\"default\":null},{\"name\":\"message\",\"kind\":\"Any\",\"default\":\"\"},{\"name\":\"notification_type\",\"kind\":\"Any\",\"default\":null},{\"name\":\"_rendered\",\"kind\":\"Any\",\"default\":false},{\"name\":\"_destroyed\",\"kind\":\"Any\",\"default\":false}]},{\"type\":\"model\",\"name\":\"TemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"BootstrapTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"TemplateEditor1\",\"properties\":[{\"name\":\"layout\",\"kind\":\"Any\",\"default\":[]}]},{\"type\":\"model\",\"name\":\"MaterialTemplateActions1\",\"properties\":[{\"name\":\"open_modal\",\"kind\":\"Any\",\"default\":0},{\"name\":\"close_modal\",\"kind\":\"Any\",\"default\":0}]},{\"type\":\"model\",\"name\":\"ReactiveESM1\",\"properties\":[{\"name\":\"esm_constants\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}}]},{\"type\":\"model\",\"name\":\"JSComponent1\",\"properties\":[{\"name\":\"esm_constants\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}}]},{\"type\":\"model\",\"name\":\"ReactComponent1\",\"properties\":[{\"name\":\"esm_constants\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}}]},{\"type\":\"model\",\"name\":\"AnyWidgetComponent1\",\"properties\":[{\"name\":\"esm_constants\",\"kind\":\"Any\",\"default\":{\"type\":\"map\"}}]},{\"type\":\"model\",\"name\":\"request_value1\",\"properties\":[{\"name\":\"fill\",\"kind\":\"Any\",\"default\":\"none\"},{\"name\":\"_synced\",\"kind\":\"Any\",\"default\":null},{\"name\":\"_request_sync\",\"kind\":\"Any\",\"default\":0}]}]}};\n",
       "  var render_items = [{\"docid\":\"800a3824-bf39-4438-b4e0-cfd80584fe91\",\"roots\":{\"cebecef4-98fc-4147-811c-fe92adaf228c\":\"e24676cf-f716-472e-a708-7cf0712dd91f\"},\"root_ids\":[\"cebecef4-98fc-4147-811c-fe92adaf228c\"]}];\n",
       "  var docs = Object.values(docs_json)\n",
       "  if (!docs) {\n",
       "    return\n",
       "  }\n",
       "  const py_version = docs[0].version.replace('rc', '-rc.').replace('.dev', '-dev.')\n",
       "  async function embed_document(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    await Bokeh.embed.embed_items_notebook(docs_json, render_items);\n",
       "    for (const render_item of render_items) {\n",
       "      for (const root_id of render_item.root_ids) {\n",
       "\tconst id_el = document.getElementById(root_id)\n",
       "\tif (id_el.children.length && id_el.children[0].hasAttribute('data-root-id')) {\n",
       "\t  const root_el = id_el.children[0]\n",
       "\t  root_el.id = root_el.id + '-rendered'\n",
       "\t  for (const child of root_el.children) {\n",
       "            // Ensure JupyterLab does not capture keyboard shortcuts\n",
       "            // see: https://jupyterlab.readthedocs.io/en/4.1.x/extension/notebook.html#keyboard-interaction-model\n",
       "\t    child.setAttribute('data-lm-suppress-shortcuts', 'true')\n",
       "\t  }\n",
       "\t}\n",
       "      }\n",
       "    }\n",
       "  }\n",
       "  function get_bokeh(root) {\n",
       "    if (root.Bokeh === undefined) {\n",
       "      return null\n",
       "    } else if (root.Bokeh.version !== py_version) {\n",
       "      if (root.Bokeh.versions === undefined || !root.Bokeh.versions.has(py_version)) {\n",
       "\treturn null\n",
       "      }\n",
       "      return root.Bokeh.versions.get(py_version);\n",
       "    } else if (root.Bokeh.version === py_version) {\n",
       "      return root.Bokeh\n",
       "    }\n",
       "    return null\n",
       "  }\n",
       "  function is_loaded(root) {\n",
       "    var Bokeh = get_bokeh(root)\n",
       "    return (Bokeh != null && Bokeh.Panel !== undefined)\n",
       "  }\n",
       "  if (is_loaded(root)) {\n",
       "    embed_document(root);\n",
       "  } else {\n",
       "    var attempts = 0;\n",
       "    var timer = setInterval(function(root) {\n",
       "      if (is_loaded(root)) {\n",
       "        clearInterval(timer);\n",
       "        embed_document(root);\n",
       "      } else if (document.readyState == \"complete\") {\n",
       "        attempts++;\n",
       "        if (attempts > 200) {\n",
       "          clearInterval(timer);\n",
       "\t  var Bokeh = get_bokeh(root)\n",
       "\t  if (Bokeh == null || Bokeh.Panel == null) {\n",
       "            console.warn(\"Panel: ERROR: Unable to run Panel code because Bokeh or Panel library is missing\");\n",
       "\t  } else {\n",
       "\t    console.warn(\"Panel: WARNING: Attempting to render but not all required libraries could be resolved.\")\n",
       "\t    embed_document(root)\n",
       "\t  }\n",
       "        }\n",
       "      }\n",
       "    }, 25, root)\n",
       "  }\n",
       "})(window);</script>"
      ]
     },
     "metadata": {
      "application/vnd.holoviews_exec.v0+json": {
       "id": "cebecef4-98fc-4147-811c-fe92adaf228c"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# This needs to be repeated here for the Pannel dashboard to work, WEIRD\n",
    "# TODO: report as possible bug\n",
    "import sys\n",
    "import os\n",
    "import io\n",
    "import warnings\n",
    "import holoviews as hv\n",
    "\n",
    "from functools import partial\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import pandas as pd\n",
    "import panel as pn\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "from mgo.udal import UDAL\n",
    "\n",
    "# All low level functions are imported from the momics package\n",
    "from momics.loader import load_parquets_udal\n",
    "from momics.metadata import get_metadata_udal, enhance_metadata\n",
    "import momics.plotting as pl\n",
    "from momics.panel_utils import (\n",
    "    diversity_select_widgets, create_indicators_diversity,\n",
    "    serve_app, close_server,\n",
    ")\n",
    "from momics.diversity import (\n",
    "    beta_diversity_parametrized,\n",
    ")\n",
    "\n",
    "from momics.taxonomy import (\n",
    "    fill_taxonomy_placeholders,\n",
    "    pivot_taxonomic_data,\n",
    "    remove_high_taxa,\n",
    "    prevalence_cutoff_taxonomy,\n",
    "    prevalence_cutoff,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfe0935f",
   "metadata": {},
   "source": [
    "## Loading EMO-BON (meta)data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "55ff6b58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parquet files\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    root_folder = os.path.abspath(os.path.join('/content/momics-demos'))\n",
    "else:\n",
    "    root_folder = os.path.abspath(os.path.join('../'))\n",
    "\n",
    "assets_folder = os.path.join(root_folder, 'assets')\n",
    "data_folder = os.path.join(root_folder, 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "daef4d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "@pn.cache()\n",
    "def get_data():\n",
    "    return load_parquets_udal()\n",
    "\n",
    "# Load and merge metadata\n",
    "@pn.cache()\n",
    "def get_full_metadata():\n",
    "    return get_metadata_udal()\n",
    "\n",
    "@pn.cache()\n",
    "def get_valid_samples():\n",
    "    df_valid = pd.read_csv(\n",
    "        os.path.join(root_folder, 'data/shipment_b1b2_181.csv')\n",
    "    )\n",
    "    return df_valid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "63f22dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata\n",
    "full_metadata = get_full_metadata()\n",
    "\n",
    "# filter the metadata only for valid 181 samples\n",
    "valid_samples = get_valid_samples()\n",
    "full_metadata = enhance_metadata(full_metadata, valid_samples)\n",
    "\n",
    "# LOADing data\n",
    "mgf_parquet_dfs = get_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "419cf6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep only ssu\n",
    "ssu = mgf_parquet_dfs['ssu'].copy()\n",
    "\n",
    "del mgf_parquet_dfs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5da2f8c",
   "metadata": {},
   "source": [
    "### MGnify: Get study metadata\n",
    "- study contains many samples, metadata need to be per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e49a55a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from jsonapi_client import Session as APISession\n",
    "import requests\n",
    "from tqdm import tqdm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1255735",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to get metadata for MGnify studies\n",
    "def get_mgnify_metadata(study_id):\n",
    "    with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as session:\n",
    "\n",
    "        samples = map(lambda r: r.json, session.iterate(f'studies/{study_id}/samples?page_size=1000'))\n",
    "\n",
    "        sample_list = []\n",
    "        for sample_json in tqdm(samples):\n",
    "            # Flatten sample-metadata list into a dictionary\n",
    "            # 1. Extract sample-metadata (allowing None)\n",
    "            metadata_fields = {\n",
    "                item.get(\"key\"): item.get(\"value\", None)\n",
    "                for item in sample_json[\"attributes\"].get(\"sample-metadata\", [])\n",
    "            }\n",
    "\n",
    "            # 2. Extract all other attributes (including None)\n",
    "            attributes_fields = {\n",
    "                k: v for k, v in sample_json[\"attributes\"].items()\n",
    "                if k != \"sample-metadata\"  # already unpacked separately\n",
    "            }\n",
    "\n",
    "            # 3. Merge everything including top-level id\n",
    "            flat_data = {\n",
    "                \"id\": sample_json.get(\"id\"),\n",
    "                **attributes_fields,\n",
    "                **metadata_fields\n",
    "            }\n",
    "\n",
    "            # 4. Create DataFrame\n",
    "            df = pd.DataFrame([flat_data])\n",
    "            sample_list.append(df)\n",
    "\n",
    "        # Concatenate all DataFrames into one\n",
    "        df = pd.concat(sample_list, ignore_index=True)\n",
    "        df['study'] = study_id\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "bd1d37d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_studies = [\n",
    "    'MGYS00006608',\n",
    "    'MGYS00006607',\n",
    "    'MGYS00000492',\n",
    "    'MGYS00006680',\n",
    "    'MGYS00006682',\n",
    "    'MGYS00006678',\n",
    "    'MGYS00006675',\n",
    "    'MGYS00003725',\n",
    "    'MGYS00006686',\n",
    "    'MGYS00006714',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7ce504e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006608/samples', params='', query='page_size=1000', fragment='')\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006608/samples', params='', query='page_size=1000', fragment='')\n",
      "62it [00:00, 81.21it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006607/samples', params='', query='page_size=1000', fragment='')\n",
      "48it [00:01, 41.32it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00000492/samples', params='', query='page_size=1000', fragment='')\n",
      "7it [00:00, 24.87it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006680/samples', params='', query='page_size=1000', fragment='')\n",
      "1it [00:01,  1.09s/it]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006680/samples', params='', query='page=2&page_size=1000', fragment='')\n",
      "283it [00:01, 156.16it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006682/samples', params='', query='page_size=1000', fragment='')\n",
      "52it [00:00, 118.62it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006678/samples', params='', query='page_size=1000', fragment='')\n",
      "197it [00:01, 148.68it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006678/samples', params='', query='page=2&page_size=1000', fragment='')\n",
      "313it [00:03, 112.50it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006678/samples', params='', query='page=3&page_size=1000', fragment='')\n",
      "665it [00:04, 153.34it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006675/samples', params='', query='page_size=1000', fragment='')\n",
      "249it [00:02, 116.43it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00003725/samples', params='', query='page_size=1000', fragment='')\n",
      "81it [00:42,  1.91it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006686/samples', params='', query='page_size=1000', fragment='')\n",
      "206it [00:03, 89.99it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006686/samples', params='', query='page=2&page_size=1000', fragment='')\n",
      "357it [00:04, 74.07it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n",
      "INFO | jsonapi_client.session | Entering session\n",
      "0it [00:00, ?it/s]INFO | jsonapi_client.session | Fetching document from url ParseResult(scheme='https', netloc='www.ebi.ac.uk', path='/metagenomics/api/v1/studies/MGYS00006714/samples', params='', query='page_size=1000', fragment='')\n",
      "205it [00:01, 143.82it/s]\n",
      "INFO | jsonapi_client.session | Exiting session\n",
      "INFO | jsonapi_client.session | Committing dirty resources\n"
     ]
    }
   ],
   "source": [
    "metadata_list = []\n",
    "for study in all_studies:\n",
    "    time.sleep(1)  # To avoid hitting the API too fast\n",
    "    try:\n",
    "        metadata_list.append(get_mgnify_metadata(study))\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching metadata for {study}: {e}\")\n",
    "# Concatenate all DataFrames into one\n",
    "metadata = pd.concat(metadata_list, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "52a444cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2009, 56)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "b6b22892",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2009 entries, 0 to 2008\n",
      "Data columns (total 56 columns):\n",
      " #   Column                                           Non-Null Count  Dtype  \n",
      "---  ------                                           --------------  -----  \n",
      " 0   id                                               2009 non-null   object \n",
      " 1   longitude                                        2009 non-null   float64\n",
      " 2   biosample                                        2009 non-null   object \n",
      " 3   latitude                                         2009 non-null   float64\n",
      " 4   accession                                        2009 non-null   object \n",
      " 5   sample-desc                                      2009 non-null   object \n",
      " 6   sample-alias                                     2009 non-null   object \n",
      " 7   sample-name                                      2009 non-null   object \n",
      " 8   geographic location (longitude)                  2009 non-null   object \n",
      " 9   study                                            2009 non-null   object \n",
      " 10  last-update                                      2009 non-null   object \n",
      " 11  geographic location (latitude)                   2009 non-null   object \n",
      " 12  collection date                                  2003 non-null   object \n",
      " 13  geographic location (country and/or sea,region)  1877 non-null   object \n",
      " 14  ENA checklist                                    1674 non-null   object \n",
      " 15  environment-feature                              1425 non-null   object \n",
      " 16  environment-material                             1425 non-null   object \n",
      " 17  environment (feature)                            1425 non-null   object \n",
      " 18  environment (material)                           1425 non-null   object \n",
      " 19  environment (biome)                              1425 non-null   object \n",
      " 20  project name                                     1425 non-null   object \n",
      " 21  environment-biome                                1425 non-null   object \n",
      " 22  sequencing method                                1403 non-null   object \n",
      " 23  depth                                            1266 non-null   object \n",
      " 24  target gene                                      914 non-null    object \n",
      " 25  environmental package                            423 non-null    object \n",
      " 26  last update date                                 88 non-null     object \n",
      " 27  analysis-completed                               88 non-null     object \n",
      " 28  instrument model                                 85 non-null     object \n",
      " 29  collection-date                                  84 non-null     object \n",
      " 30  geo-loc-name                                     66 non-null     object \n",
      " 31  investigation type                               66 non-null     object \n",
      " 32  temperature                                      25 non-null     object \n",
      " 33  salinity                                         22 non-null     object \n",
      " 34  marine region                                    7 non-null      object \n",
      " 35  protocol label                                   7 non-null      object \n",
      " 36  geographic location (depth)                      7 non-null      object \n",
      " 37  NCBI sample classification                       7 non-null      object \n",
      " 38  latitude end                                     7 non-null      object \n",
      " 39  longitude end                                    7 non-null      object \n",
      " 40  event date/time start                            7 non-null      object \n",
      " 41  event date/time end                              7 non-null      object \n",
      " 42  citation                                         7 non-null      object \n",
      " 43  chlorophyll sensor                               7 non-null      object \n",
      " 44  sampling platform                                7 non-null      object \n",
      " 45  sampling campaign                                7 non-null      object \n",
      " 46  further details                                  7 non-null      object \n",
      " 47  event label                                      7 non-null      object \n",
      " 48  sample status                                    7 non-null      object \n",
      " 49  sample collection device                         7 non-null      object \n",
      " 50  salinity sensor                                  7 non-null      object \n",
      " 51  oxygen sensor                                    7 non-null      object \n",
      " 52  nitrate sensor                                   7 non-null      object \n",
      " 53  size fraction upper threshold                    7 non-null      object \n",
      " 54  sampling station                                 7 non-null      object \n",
      " 55  size fraction lower threshold                    7 non-null      object \n",
      "dtypes: float64(2), object(54)\n",
      "memory usage: 879.1+ KB\n"
     ]
    }
   ],
   "source": [
    "non_null_counts = metadata.notnull().sum()\n",
    "\n",
    "# Step 2: Sort column names by that count (descending)\n",
    "ordered_columns = non_null_counts.sort_values(ascending=False).index.tolist()\n",
    "\n",
    "# Step 3: Reorder DataFrame\n",
    "df_ordered = metadata[ordered_columns]\n",
    "\n",
    "# Step 4 (optional): Show info\n",
    "df_ordered.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4fd4aa96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "study\n",
       "MGYS00006678    665\n",
       "MGYS00006686    357\n",
       "MGYS00006680    283\n",
       "MGYS00006675    249\n",
       "MGYS00006714    205\n",
       "MGYS00003725     81\n",
       "MGYS00006608     62\n",
       "MGYS00006682     52\n",
       "MGYS00006607     48\n",
       "MGYS00000492      7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.study.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "047664c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>longitude</th>\n",
       "      <th>biosample</th>\n",
       "      <th>latitude</th>\n",
       "      <th>accession</th>\n",
       "      <th>sample-desc</th>\n",
       "      <th>environment-biome</th>\n",
       "      <th>environment-feature</th>\n",
       "      <th>environment-material</th>\n",
       "      <th>sample-name</th>\n",
       "      <th>...</th>\n",
       "      <th>sample collection device</th>\n",
       "      <th>sample status</th>\n",
       "      <th>sampling station</th>\n",
       "      <th>size fraction lower threshold</th>\n",
       "      <th>size fraction upper threshold</th>\n",
       "      <th>salinity</th>\n",
       "      <th>geographic location (country and/or sea,region)</th>\n",
       "      <th>target gene</th>\n",
       "      <th>geo-loc-name</th>\n",
       "      <th>investigation type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERS5150811</td>\n",
       "      <td>14.34</td>\n",
       "      <td>SAMEA7392422</td>\n",
       "      <td>40.80</td>\n",
       "      <td>ERS5150811</td>\n",
       "      <td>surface water sample from the Tyrrhenian Sea</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>microbial community [PCO:1000004]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>32513:f016dd88-3b2f-48a1-b740-b5a51edf1232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERS5150815</td>\n",
       "      <td>-3.89</td>\n",
       "      <td>SAMEA7392426</td>\n",
       "      <td>48.67</td>\n",
       "      <td>ERS5150815</td>\n",
       "      <td>surface water sample from the English Channel</td>\n",
       "      <td>estuarine biome [ENVO:01000020]</td>\n",
       "      <td>microbial community [PCO:1000004]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>32509:f016dd88-3b2f-48a1-b740-b5a51edf1232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERS5150818</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>SAMEA7392429</td>\n",
       "      <td>53.24</td>\n",
       "      <td>ERS5150818</td>\n",
       "      <td>surface water sample from the North Atlantic O...</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>microbial community [PCO:1000004]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>32506:f016dd88-3b2f-48a1-b740-b5a51edf1232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERS5150819</td>\n",
       "      <td>-60.00</td>\n",
       "      <td>SAMEA7392430</td>\n",
       "      <td>43.94</td>\n",
       "      <td>ERS5150819</td>\n",
       "      <td>surface water sample from the Northwest Atlant...</td>\n",
       "      <td>neritic epipelagic zone biome [ENVO:01000042]</td>\n",
       "      <td>microbial community [PCO:1000004]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>32505:f016dd88-3b2f-48a1-b740-b5a51edf1232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERS5150823</td>\n",
       "      <td>12.58</td>\n",
       "      <td>SAMEA7392434</td>\n",
       "      <td>55.73</td>\n",
       "      <td>ERS5150823</td>\n",
       "      <td>surface water sample from the Baltic Sea</td>\n",
       "      <td>marine biome [ENVO:00000447]</td>\n",
       "      <td>microbial community [PCO:1000004]</td>\n",
       "      <td>sea water [ENVO:00002149]</td>\n",
       "      <td>32501:f016dd88-3b2f-48a1-b740-b5a51edf1232</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  longitude     biosample  latitude   accession  \\\n",
       "0  ERS5150811      14.34  SAMEA7392422     40.80  ERS5150811   \n",
       "1  ERS5150815      -3.89  SAMEA7392426     48.67  ERS5150815   \n",
       "2  ERS5150818      -9.31  SAMEA7392429     53.24  ERS5150818   \n",
       "3  ERS5150819     -60.00  SAMEA7392430     43.94  ERS5150819   \n",
       "4  ERS5150823      12.58  SAMEA7392434     55.73  ERS5150823   \n",
       "\n",
       "                                         sample-desc  \\\n",
       "0       surface water sample from the Tyrrhenian Sea   \n",
       "1      surface water sample from the English Channel   \n",
       "2  surface water sample from the North Atlantic O...   \n",
       "3  surface water sample from the Northwest Atlant...   \n",
       "4           surface water sample from the Baltic Sea   \n",
       "\n",
       "                               environment-biome  \\\n",
       "0                   marine biome [ENVO:00000447]   \n",
       "1                estuarine biome [ENVO:01000020]   \n",
       "2                   marine biome [ENVO:00000447]   \n",
       "3  neritic epipelagic zone biome [ENVO:01000042]   \n",
       "4                   marine biome [ENVO:00000447]   \n",
       "\n",
       "                 environment-feature       environment-material  \\\n",
       "0  microbial community [PCO:1000004]  sea water [ENVO:00002149]   \n",
       "1  microbial community [PCO:1000004]  sea water [ENVO:00002149]   \n",
       "2  microbial community [PCO:1000004]  sea water [ENVO:00002149]   \n",
       "3  microbial community [PCO:1000004]  sea water [ENVO:00002149]   \n",
       "4  microbial community [PCO:1000004]  sea water [ENVO:00002149]   \n",
       "\n",
       "                                  sample-name  ... sample collection device  \\\n",
       "0  32513:f016dd88-3b2f-48a1-b740-b5a51edf1232  ...                      NaN   \n",
       "1  32509:f016dd88-3b2f-48a1-b740-b5a51edf1232  ...                      NaN   \n",
       "2  32506:f016dd88-3b2f-48a1-b740-b5a51edf1232  ...                      NaN   \n",
       "3  32505:f016dd88-3b2f-48a1-b740-b5a51edf1232  ...                      NaN   \n",
       "4  32501:f016dd88-3b2f-48a1-b740-b5a51edf1232  ...                      NaN   \n",
       "\n",
       "  sample status sampling station size fraction lower threshold  \\\n",
       "0           NaN              NaN                           NaN   \n",
       "1           NaN              NaN                           NaN   \n",
       "2           NaN              NaN                           NaN   \n",
       "3           NaN              NaN                           NaN   \n",
       "4           NaN              NaN                           NaN   \n",
       "\n",
       "  size fraction upper threshold salinity  \\\n",
       "0                           NaN      NaN   \n",
       "1                           NaN      NaN   \n",
       "2                           NaN      NaN   \n",
       "3                           NaN      NaN   \n",
       "4                           NaN      NaN   \n",
       "\n",
       "  geographic location (country and/or sea,region) target gene geo-loc-name  \\\n",
       "0                                             NaN         NaN          NaN   \n",
       "1                                             NaN         NaN          NaN   \n",
       "2                                             NaN         NaN          NaN   \n",
       "3                                             NaN         NaN          NaN   \n",
       "4                                             NaN         NaN          NaN   \n",
       "\n",
       "  investigation type  \n",
       "0                NaN  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3                NaN  \n",
       "4                NaN  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "4b38c777",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save metadata to parquet\n",
    "metadata.to_parquet(\n",
    "    os.path.join(data_folder, 'mgnify_metadata.parquet'),\n",
    "    index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5c72e77",
   "metadata": {},
   "source": [
    "### ENA: EMO-BON project name is PRJEB51688\n",
    "- this needs to be searched at ENA, because it is not present on MGnify\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "67ed13a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step 1: Get all available fields for 'read_run'\n",
    "fields_url = \"https://www.ebi.ac.uk/ena/portal/api/returnFields\"\n",
    "fields_params = {\n",
    "    \"result\": \"read_run\",\n",
    "    \"format\": \"json\"\n",
    "}\n",
    "fields_response = requests.get(fields_url, params=fields_params)\n",
    "fields = [f[\"columnId\"] for f in fields_response.json()]\n",
    "\n",
    "##############################\n",
    "# Define base URL\n",
    "base_url = \"https://www.ebi.ac.uk/ena/portal/api/search\"\n",
    "\n",
    "# Define query parameters\n",
    "params = {\n",
    "    \"result\": \"read_run\",  # result type\n",
    "    \"query\": '\"study_accession=PRJEB51688\"',  # search condition\n",
    "    \"fields\": \",\".join(fields),\n",
    "    \"limit\": 1000,  # limit results\n",
    "    \"format\": \"json\"  # return as JSON\n",
    "}\n",
    "\n",
    "# Send GET request\n",
    "response = requests.get(base_url, params=params)\n",
    "\n",
    "# Check response\n",
    "if response.ok:\n",
    "    data = response.json()\n",
    "    df = pd.DataFrame(data)\n",
    "    # print(df)\n",
    "else:\n",
    "    print(\"Error:\", response.status_code, response.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "24148b0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "df.replace(\"\", np.nan, inplace=True)\n",
    "df_cleaned = df.dropna(axis=1, how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "b8c6b6f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 188 entries, 0 to 187\n",
      "Data columns (total 78 columns):\n",
      " #   Column                             Non-Null Count  Dtype \n",
      "---  ------                             --------------  ----- \n",
      " 0   run_accession                      188 non-null    object\n",
      " 1   bam_aspera                         188 non-null    object\n",
      " 2   bam_bytes                          188 non-null    object\n",
      " 3   bam_ftp                            188 non-null    object\n",
      " 4   bam_galaxy                         188 non-null    object\n",
      " 5   bam_md5                            188 non-null    object\n",
      " 6   base_count                         188 non-null    object\n",
      " 7   broad_scale_environmental_context  188 non-null    object\n",
      " 8   center_name                        188 non-null    object\n",
      " 9   checklist                          188 non-null    object\n",
      " 10  collection_date                    188 non-null    object\n",
      " 11  collection_date_end                188 non-null    object\n",
      " 12  collection_date_start              188 non-null    object\n",
      " 13  country                            188 non-null    object\n",
      " 14  datahub                            188 non-null    object\n",
      " 15  depth                              188 non-null    object\n",
      " 16  description                        188 non-null    object\n",
      " 17  elevation                          31 non-null     object\n",
      " 18  environment_biome                  188 non-null    object\n",
      " 19  environment_feature                188 non-null    object\n",
      " 20  environment_material               188 non-null    object\n",
      " 21  environmental_medium               188 non-null    object\n",
      " 22  experiment_accession               188 non-null    object\n",
      " 23  experiment_alias                   188 non-null    object\n",
      " 24  experiment_title                   188 non-null    object\n",
      " 25  fastq_aspera                       188 non-null    object\n",
      " 26  fastq_bytes                        188 non-null    object\n",
      " 27  fastq_ftp                          188 non-null    object\n",
      " 28  fastq_galaxy                       188 non-null    object\n",
      " 29  fastq_md5                          188 non-null    object\n",
      " 30  first_created                      188 non-null    object\n",
      " 31  first_public                       188 non-null    object\n",
      " 32  instrument_model                   188 non-null    object\n",
      " 33  instrument_platform                188 non-null    object\n",
      " 34  last_updated                       188 non-null    object\n",
      " 35  lat                                188 non-null    object\n",
      " 36  library_construction_protocol      188 non-null    object\n",
      " 37  library_layout                     188 non-null    object\n",
      " 38  library_name                       188 non-null    object\n",
      " 39  library_selection                  188 non-null    object\n",
      " 40  library_source                     188 non-null    object\n",
      " 41  library_strategy                   188 non-null    object\n",
      " 42  local_environmental_context        188 non-null    object\n",
      " 43  location                           188 non-null    object\n",
      " 44  location_end                       188 non-null    object\n",
      " 45  location_start                     188 non-null    object\n",
      " 46  lon                                188 non-null    object\n",
      " 47  nominal_length                     1 non-null      object\n",
      " 48  project_name                       15 non-null     object\n",
      " 49  read_count                         188 non-null    object\n",
      " 50  run_alias                          188 non-null    object\n",
      " 51  run_date                           188 non-null    object\n",
      " 52  sample_accession                   188 non-null    object\n",
      " 53  sample_alias                       188 non-null    object\n",
      " 54  sample_collection                  185 non-null    object\n",
      " 55  sample_description                 188 non-null    object\n",
      " 56  sample_title                       188 non-null    object\n",
      " 57  scientific_name                    188 non-null    object\n",
      " 58  secondary_sample_accession         188 non-null    object\n",
      " 59  secondary_study_accession          188 non-null    object\n",
      " 60  sequencing_method                  188 non-null    object\n",
      " 61  status                             188 non-null    object\n",
      " 62  study_accession                    188 non-null    object\n",
      " 63  study_alias                        188 non-null    object\n",
      " 64  study_title                        188 non-null    object\n",
      " 65  submission_accession               188 non-null    object\n",
      " 66  submitted_aspera                   188 non-null    object\n",
      " 67  submitted_bytes                    188 non-null    object\n",
      " 68  submitted_format                   188 non-null    object\n",
      " 69  submitted_ftp                      188 non-null    object\n",
      " 70  submitted_galaxy                   188 non-null    object\n",
      " 71  submitted_md5                      188 non-null    object\n",
      " 72  submitted_read_type                188 non-null    object\n",
      " 73  tag                                188 non-null    object\n",
      " 74  target_gene                        188 non-null    object\n",
      " 75  tax_id                             188 non-null    object\n",
      " 76  tax_lineage                        188 non-null    object\n",
      " 77  temperature                        8 non-null      object\n",
      "dtypes: object(78)\n",
      "memory usage: 114.7+ KB\n"
     ]
    }
   ],
   "source": [
    "df_cleaned.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "9f366e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       1;2787823;12908;408169;410657;412755\n",
       "1      1;2787823;12908;408169;410657;1874687\n",
       "2      1;2787823;12908;408169;410657;1874687\n",
       "3      1;2787823;12908;408169;410657;1874687\n",
       "4      1;2787823;12908;408169;410657;1874687\n",
       "                       ...                  \n",
       "183    1;2787823;12908;408169;410657;1874687\n",
       "184    1;2787823;12908;408169;410657;1874687\n",
       "185     1;2787823;12908;408169;410657;412755\n",
       "186     1;2787823;12908;408169;410657;412755\n",
       "187     1;2787823;12908;408169;410657;412755\n",
       "Name: tax_lineage, Length: 188, dtype: object"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cleaned['tax_lineage']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8d970d",
   "metadata": {},
   "source": [
    "### Merge the two\n",
    "- check how many columns share the same name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "d0b8c19e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'depth', 'temperature'}"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(metadata.columns).intersection(set(df_cleaned.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7094984f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momics-demos",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
