{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Jupyter to Galaxy for nextflow with data from MGnify\n",
    "\n",
    "* need to upload nextflow directory with the files and dependencies\n",
    "* trigger the Galaxy workflow\n",
    "* retrieve results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Platform: local Linux\n",
      "Environment: vscode\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import io\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    # clone the momics-demos repository to use the utils module from there\n",
    "    # TODO: eventually utils from momics will be used for that\n",
    "    try:\n",
    "        os.system('git clone https://github.com/palec87/momics-demos.git')\n",
    "        print(f\"Repository cloned\")\n",
    "    except OSError as e:\n",
    "        print(f\"An error occurred while cloning the repository: {e}\")\n",
    "\n",
    "    sys.path.insert(0,'/content/momics-demos')\n",
    "\n",
    "else:\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "\n",
    "    # downside of this is that all the deps need to be installed in the current (momics-demos) environment\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '../../../marine-omics')))  # local momics package, to be removed too\n",
    "\n",
    "from utils import init_setup, get_notebook_environment\n",
    "init_setup()\n",
    "\n",
    "# Initialize the environment variable\n",
    "notebook_environment = 'unknown'\n",
    "# Determine the notebook environment\n",
    "env = get_notebook_environment()\n",
    "print(f\"Environment: {env}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## XML for galaxy\n",
    "\n",
    "**XML taken from [here](https://galaxyproject.org/blog/2022-08-15-making-nextflow-work-with-galaxy-at-cfsan-fda/)**\n",
    "```xml\n",
    "<requirements>\n",
    "    <requirement type=\"package\" version=\"22.04\">nextflow</requirement>\n",
    "    <requirement type=\"package\">graphviz</requirement>\n",
    "</requirements>\n",
    "<command detect_errors=\"exit_code\"><![CDATA[\n",
    "    nextflow run $__tool_directory__/nf-pipelines/my_pipeline.nf --input /path/to/input\n",
    "]]></command>\n",
    "```\n",
    "\n",
    "**nf-core/rnaseq**\n",
    "```bash\n",
    "nextflow run nf-core/rnaseq \\\n",
    "    --input <SAMPLESHEET> \\\n",
    "    --outdir <OUTDIR> \\\n",
    "    --gtf <GTF> \\\n",
    "    --fasta <GENOME FASTA> \\\n",
    "    -profile <docker/singularity/.../institute>\n",
    "```\n",
    "\n",
    "**with the input sample `.csv` sheet as follows**\n",
    "```csv\n",
    "sample,fastq_1,fastq_2,strandedness\n",
    "CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz,auto\n",
    "CONTROL_REP1,AEG588A1_S1_L003_R1_001.fastq.gz,AEG588A1_S1_L003_R2_001.fastq.gz,auto\n",
    "CONTROL_REP1,AEG588A1_S1_L004_R1_001.fastq.gz,AEG588A1_S1_L004_R2_001.fastq.gz,auto\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "from datetime import datetime\n",
    "from platform import python_version\n",
    "import logging\n",
    "\n",
    "from jsonapi_client import Session as APISession\n",
    "from jsonapi_client import Modifier\n",
    "import requests\n",
    "\n",
    "# Import\n",
    "import bioblend.galaxy as g  # BioBlend is a Python library, wrapping the functionality of Galaxy and CloudMan APIs\n",
    "# import boto3\n",
    "import pandas as pd\n",
    "from bioblend.galaxy import GalaxyInstance\n",
    "from bioblend.galaxy.datasets import DatasetClient\n",
    "\n",
    "from momics.galaxy.blue_cloud import BCGalaxy\n",
    "# instead of the jupyter magic, you can also use\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read your secrets from the .env file\n",
    "# To see your API key login -> click 'user' (top right) -> click 'preferences' -> click 'Manage API Key' (menu to the left) -> click the icon to 'copy key'\n",
    "GALAXY_URL = os.getenv(\"GALAXY_EARTH_URL\")  # alternatively os.environ.get('GALAXY_URL'), \"https://earth-system.usegalaxy.eu/\"\n",
    "GALAXY_KEY = os.getenv(\"GALAXY_EARTH_KEY\")  # alternatively os.environ.get('GALAXY_KEY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# my class init, but not using it for anything here\n",
    "exp = BCGalaxy(\"GALAXY_EARTH_URL\", \"GALAXY_EARTH_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = exp.gi.config.get_config()\n",
    "# config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds_client = g.datasets.DatasetClient(exp.gi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['id', 'name', 'history_id', 'hid', 'deleted', 'visible', 'type_id', 'type', 'create_time', 'update_time', 'url', 'tags', 'history_content_type', 'dataset_id', 'state', 'extension', 'purged', 'genome_build', 'object_store_id', 'quota_source_label'])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds_list = ds_client.get_datasets()\n",
    "ds_list[0].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': '4838ba20a6d867655f7d33d07e6b9e78',\n",
       "  'name': 'Concatenate datasets on data 1 and data 2',\n",
       "  'history_id': '7d721489019e2043',\n",
       "  'hid': 1,\n",
       "  'deleted': False,\n",
       "  'visible': True,\n",
       "  'type_id': 'dataset-4838ba20a6d867655f7d33d07e6b9e78',\n",
       "  'type': 'file',\n",
       "  'create_time': '2025-02-14T14:33:14.729402',\n",
       "  'update_time': '2025-02-14T14:33:32.064405',\n",
       "  'url': '/api/histories/7d721489019e2043/contents/4838ba20a6d867655f7d33d07e6b9e78',\n",
       "  'tags': [],\n",
       "  'history_content_type': 'dataset',\n",
       "  'dataset_id': '4838ba20a6d8676553252dc7f1253d16',\n",
       "  'state': 'queued',\n",
       "  'extension': 'txt',\n",
       "  'purged': False,\n",
       "  'genome_build': '?',\n",
       "  'object_store_id': 'files27',\n",
       "  'quota_source_label': None},\n",
       " {'id': '4838ba20a6d8676558e4e84c5b89daec',\n",
       "  'name': 'Concatenate datasets on data 1 and data 2',\n",
       "  'history_id': '2e8a4915680c5db1',\n",
       "  'hid': 3,\n",
       "  'deleted': False,\n",
       "  'visible': True,\n",
       "  'type_id': 'dataset-4838ba20a6d8676558e4e84c5b89daec',\n",
       "  'type': 'file',\n",
       "  'create_time': '2025-02-13T15:43:38.597584',\n",
       "  'update_time': '2025-02-13T15:44:25.717816',\n",
       "  'url': '/api/histories/2e8a4915680c5db1/contents/4838ba20a6d8676558e4e84c5b89daec',\n",
       "  'tags': [],\n",
       "  'history_content_type': 'dataset',\n",
       "  'dataset_id': '4838ba20a6d867657f631dca3324369f',\n",
       "  'state': 'ok',\n",
       "  'extension': 'txt',\n",
       "  'purged': False,\n",
       "  'genome_build': '?',\n",
       "  'object_store_id': 'files28',\n",
       "  'quota_source_label': None},\n",
       " {'id': '4838ba20a6d867658edecd7071a62aef',\n",
       "  'name': 'dataset2.txt',\n",
       "  'history_id': '2e8a4915680c5db1',\n",
       "  'hid': 2,\n",
       "  'deleted': False,\n",
       "  'visible': True,\n",
       "  'type_id': 'dataset-4838ba20a6d867658edecd7071a62aef',\n",
       "  'type': 'file',\n",
       "  'create_time': '2025-02-13T15:42:28.038460',\n",
       "  'update_time': '2025-02-13T15:43:31.691163',\n",
       "  'url': '/api/histories/2e8a4915680c5db1/contents/4838ba20a6d867658edecd7071a62aef',\n",
       "  'tags': [],\n",
       "  'history_content_type': 'dataset',\n",
       "  'dataset_id': '4838ba20a6d86765c27d49db44672e67',\n",
       "  'state': 'ok',\n",
       "  'extension': 'txt',\n",
       "  'purged': False,\n",
       "  'genome_build': '?',\n",
       "  'object_store_id': 'files28',\n",
       "  'quota_source_label': None},\n",
       " {'id': '4838ba20a6d867654ef0f38728567a4a',\n",
       "  'name': 'dataset1.txt',\n",
       "  'history_id': '2e8a4915680c5db1',\n",
       "  'hid': 1,\n",
       "  'deleted': False,\n",
       "  'visible': True,\n",
       "  'type_id': 'dataset-4838ba20a6d867654ef0f38728567a4a',\n",
       "  'type': 'file',\n",
       "  'create_time': '2025-02-13T15:40:56.315560',\n",
       "  'update_time': '2025-02-13T15:42:27.390917',\n",
       "  'url': '/api/histories/2e8a4915680c5db1/contents/4838ba20a6d867654ef0f38728567a4a',\n",
       "  'tags': [],\n",
       "  'history_content_type': 'dataset',\n",
       "  'dataset_id': '4838ba20a6d86765867c5ebba93c9263',\n",
       "  'state': 'ok',\n",
       "  'extension': 'txt',\n",
       "  'purged': False,\n",
       "  'genome_build': '?',\n",
       "  'object_store_id': 'files28',\n",
       "  'quota_source_label': None}]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get only the datasets (dataset word is specific for the test workflow)\n",
    "[k for k in ds_list if 'dataset' in k['name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'model_class': 'HistoryDatasetAssociation',\n",
       "  'id': '4838ba20a6d867658edecd7071a62aef',\n",
       "  'name': 'dataset2.txt',\n",
       "  'history_id': '2e8a4915680c5db1',\n",
       "  'hid': 2,\n",
       "  'deleted': False,\n",
       "  'visible': True,\n",
       "  'type_id': 'dataset-4838ba20a6d867658edecd7071a62aef',\n",
       "  'type': 'file',\n",
       "  'create_time': '2025-02-13T15:42:28.038460',\n",
       "  'update_time': '2025-02-13T15:43:31.691163',\n",
       "  'url': '/api/histories/2e8a4915680c5db1/contents/4838ba20a6d867658edecd7071a62aef',\n",
       "  'tags': [],\n",
       "  'history_content_type': 'dataset',\n",
       "  'copied_from_ldda_id': None,\n",
       "  'dataset_id': '4838ba20a6d86765c27d49db44672e67',\n",
       "  'state': 'ok',\n",
       "  'extension': 'txt',\n",
       "  'purged': False,\n",
       "  'genome_build': '?',\n",
       "  'hda_ldda': 'hda',\n",
       "  'accessible': True,\n",
       "  'misc_info': 'uploaded txt file',\n",
       "  'misc_blurb': '3 lines',\n",
       "  'file_ext': 'txt',\n",
       "  'file_size': 18,\n",
       "  'resubmitted': False,\n",
       "  'meta_files': [],\n",
       "  'data_type': 'galaxy.datatypes.data.Text',\n",
       "  'peek': '<table cellspacing=\"0\" cellpadding=\"3\"><tr><td>this</td></tr><tr><td>is</td></tr><tr><td>dataset 2</td></tr></table>',\n",
       "  'creating_job': '11ac94870d0bb33a83cf798ca735de65',\n",
       "  'rerunnable': False,\n",
       "  'uuid': '96a8c4ae-1d81-43b9-a860-3f50323a3180',\n",
       "  'permissions': {'manage': ['b87365187f298056'],\n",
       "   'access': ['b87365187f298056']},\n",
       "  'file_name': '/data/dnb10/galaxy_db/files/9/6/a/dataset_96a8c4ae-1d81-43b9-a860-3f50323a3180.dat',\n",
       "  'display_apps': [],\n",
       "  'display_types': [],\n",
       "  'validated_state': 'unknown',\n",
       "  'validated_state_message': None,\n",
       "  'annotation': None,\n",
       "  'download_url': '/api/histories/2e8a4915680c5db1/contents/4838ba20a6d867658edecd7071a62aef/display',\n",
       "  'api_type': 'file',\n",
       "  'created_from_basename': None,\n",
       "  'hashes': [],\n",
       "  'drs_id': 'hda-5c4deb8360d3463bfdcabba6b94200a6',\n",
       "  'sources': [],\n",
       "  'quota_source_label': None,\n",
       "  'object_store_id': 'files28',\n",
       "  'metadata_dbkey': '?',\n",
       "  'metadata_data_lines': 3},\n",
       " dict)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# history_id, dataset_id\n",
    "# TODO: find the right way to query those\n",
    "# do i HAVE to initiate the dataset client for that?\n",
    "dataset = exp.gi.histories.show_dataset('7011ecc923ced444', '4838ba20a6d867658edecd7071a62aef')\n",
    "dataset, type(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6,\n",
       " dict_keys(['model_class', 'id', 'name', 'deleted', 'purged', 'archived', 'url', 'published', 'count', 'annotation', 'tags', 'update_time', 'preferred_object_store_id']),\n",
       " [('History from planemo run tutorial workflow', '7d721489019e2043'),\n",
       "  ('History from planemo run tutorial workflow', '7011ecc923ced444'),\n",
       "  ('CWL Target History', '2e8a4915680c5db1'),\n",
       "  ('GECCO Run', '115816a852b41534'),\n",
       "  ('Unnamed history', '227d4225a27186af'),\n",
       "  ('GECCO Run', '8d8d4bf21253beda')])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "histories = exp.gi.histories.get_histories(history_id=None, name=None, deleted=False)\n",
    "len(histories), histories[0].keys(), [(k['name'], k['id']) for k in histories]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow and invocations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(dict_keys(['model_class', 'id', 'latest_workflow_id', 'name', 'create_time', 'update_time', 'published', 'importable', 'deleted', 'hidden', 'tags', 'latest_workflow_uuid', 'annotations', 'url', 'owner', 'source_metadata', 'number_of_steps', 'show_in_tool_panel']),\n",
       " [('planemo run tutorial', '7cc90b8a12c8c643'),\n",
       "  ('test_nextflow', '659ce0d877aab15e')])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "workflows = exp.gi.workflows.get_workflows(workflow_id=None, name=None, published=False)\n",
    "workflows[0].keys(), [(k['name'], k['id']) for k in workflows]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'0': {'label': 'Dataset 2',\n",
       "  'value': '',\n",
       "  'uuid': '004a6ce7-a8f3-4a54-b513-e1456a3695e8'},\n",
       " '1': {'label': 'Dataset 1',\n",
       "  'value': '',\n",
       "  'uuid': 'b42b2137-033e-4e7e-849e-21def05f4a70'},\n",
       " '2': {'label': 'Number of lines',\n",
       "  'value': '',\n",
       "  'uuid': '442ca31b-cdcf-46eb-815c-c23baf2c4dc5'}}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# to select the planemo workflow\n",
    "planemo_id = [k['id'] for k in workflows if 'planemo' in k['name']][0]  # going thourgh WFs and mathces on name string\n",
    "wf = exp.gi.workflows.show_workflow(planemo_id)\n",
    "wf['inputs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# invoke\n",
    "\n",
    "# datamap constructed manually with the data ids\n",
    "datamap = {'0': {'src':'hda', 'id':'4838ba20a6d867658edecd7071a62aef'},\n",
    "           '1': {'src':'hda', 'id':'4838ba20a6d867654ef0f38728567a4a'},\n",
    "           '2': {'value': '3'},\n",
    "}\n",
    "w = exp.gi.workflows.invoke_workflow(\n",
    "    planemo_id,\n",
    "    inputs=datamap,\n",
    "    params=None,\n",
    "    history_id=None,\n",
    "    history_name=None,\n",
    "    import_inputs_to_history=False,\n",
    "    replacement_params=None,\n",
    "    allow_tool_state_corrections=None,\n",
    "    )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'new': ['4838ba20a6d867655cd296330f06d83a'],\n",
       " 'upload': [],\n",
       " 'queued': ['4838ba20a6d867655f7d33d07e6b9e78'],\n",
       " 'running': [],\n",
       " 'ok': [],\n",
       " 'empty': [],\n",
       " 'error': [],\n",
       " 'paused': [],\n",
       " 'setting_metadata': [],\n",
       " 'failed_metadata': [],\n",
       " 'deferred': [],\n",
       " 'discarded': []}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp.gi.histories.show_history('7d721489019e2043')['state_ids']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use planemo from here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_planemo():\n",
    "    # invoke\n",
    "\n",
    "    # datamap constructed manually with the data ids\n",
    "    os.system('planemo run tutorial.ga tutorial-job.yml \\\n",
    "        --engine external_galaxy \\\n",
    "        --galaxy_url https://earth-system.usegalaxy.eu/ \\\n",
    "        --galaxy_user_key 154f80453cb792de3ccb35f6be337475')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Usage: planemo run [OPTIONS] RUNNABLE_PATH_OR_ID JOB_PATH\n",
      "Try 'planemo run --help' for help.\n",
      "\n",
      "Error: Invalid value for 'JOB_PATH': File 'tutorial-job.yml' does not exist.\n"
     ]
    }
   ],
   "source": [
    "run_planemo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to package nextflow file into the workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9,\n",
       " dict_keys(['model_class', 'id', 'history_id', 'tool_id', 'state', 'exit_code', 'create_time', 'update_time', 'galaxy_version', 'external_id', 'handler', 'job_runner_name', 'command_line', 'user_email']),\n",
       " [('11ac94870d0bb33ad6fc7dbb812f1217', 'queued'),\n",
       "  ('11ac94870d0bb33adedcde9f82791389', 'new'),\n",
       "  ('11ac94870d0bb33a1219c0f07f0981f8', 'ok'),\n",
       "  ('11ac94870d0bb33ad39f874d7ca97db4', 'ok'),\n",
       "  ('11ac94870d0bb33a83cf798ca735de65', 'ok'),\n",
       "  ('11ac94870d0bb33a072bd9a33dc8c14d', 'ok'),\n",
       "  ('11ac94870d0bb33ac814be21ef6fbc2d', 'deleted'),\n",
       "  ('11ac94870d0bb33a4a74056d2ffeb889', 'ok'),\n",
       "  ('11ac94870d0bb33aea8210495860d0f4', 'ok')])"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#jobs\n",
    "jobs = exp.gi.jobs.get_jobs()\n",
    "len(jobs), jobs[0].keys(), [(k['id'], k['state']) for k in jobs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'model_class': 'Job',\n",
       " 'id': '11ac94870d0bb33ad6fc7dbb812f1217',\n",
       " 'history_id': '7d721489019e2043',\n",
       " 'tool_id': 'cat1',\n",
       " 'state': 'queued',\n",
       " 'exit_code': None,\n",
       " 'create_time': '2025-02-14T14:33:14.709536',\n",
       " 'update_time': '2025-02-14T14:33:32.393319',\n",
       " 'galaxy_version': '24.2',\n",
       " 'external_id': None,\n",
       " 'handler': None,\n",
       " 'job_runner_name': None,\n",
       " 'command_line': None,\n",
       " 'user_email': None}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jobs[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13585,\n",
       " dict_keys(['model_class', 'id', 'name', 'version', 'description', 'labels', 'edam_operations', 'edam_topics', 'hidden', 'is_workflow_compatible', 'xrefs', 'link', 'min_width', 'target', 'panel_section_id', 'panel_section_name', 'form_style']),\n",
       " [('Upload File', 'upload1'),\n",
       "  ('SEEK test', 'ds_seek_test'),\n",
       "  ('UCSC Main', 'ucsc_table_direct1'),\n",
       "  ('UCSC Archaea', 'ucsc_table_direct_archaea1'),\n",
       "  ('NCBI Datasets Genomes', 'ncbi_datasets_source'),\n",
       "  ('downloads', 'lftp'),\n",
       "  ('EBI SRA', 'ebi_sra_main'),\n",
       "  ('SRA', 'sra_source'),\n",
       "  ('Get Microbial Data', 'microbial_import1'),\n",
       "  ('BioMart', 'biomart'),\n",
       "  ('BioMart', 'biomart_test'),\n",
       "  ('CBI Rice Mart', 'cbi_rice_mart'),\n",
       "  ('GrameneMart', 'gramenemart'),\n",
       "  ('modENCODE fly', 'modENCODEfly'),\n",
       "  ('InterMine', 'intermine'),\n",
       "  ('Flymine', 'flymine'),\n",
       "  ('modENCODE modMine', 'modmine'),\n",
       "  ('MouseMine', 'mousemine'),\n",
       "  ('Ratmine', 'ratmine'),\n",
       "  ('YeastMine', 'yeastmine'),\n",
       "  ('metabolicMine', 'metabolicmine'),\n",
       "  ('modENCODE worm', 'modENCODEworm'),\n",
       "  ('WormBase', 'wormbase'),\n",
       "  ('ZebrafishMine', 'zebrafishmine'),\n",
       "  ('EuPathDB', 'eupathdb'),\n",
       "  ('HbVar', 'hbvar'),\n",
       "  ('OmicsDI', 'omicsdi'),\n",
       "  ('Export datasets', 'export_remote'),\n",
       "  ('Unzip collection', '__UNZIP_COLLECTION__'),\n",
       "  ('Zip collections', '__ZIP_COLLECTION__'),\n",
       "  ('Filter failed datasets', '__FILTER_FAILED_DATASETS__'),\n",
       "  ('Filter empty datasets', '__FILTER_EMPTY_DATASETS__'),\n",
       "  ('Filter null elements', '__FILTER_NULL__'),\n",
       "  ('Flatten collection', '__FLATTEN__'),\n",
       "  ('Merge collections', '__MERGE_COLLECTION__'),\n",
       "  ('Relabel identifiers', '__RELABEL_FROM_FILE__'),\n",
       "  ('Filter collection', '__FILTER_FROM_FILE__'),\n",
       "  ('Sort collection', '__SORTLIST__'),\n",
       "  ('Harmonize two collections', '__HARMONIZELISTS__'),\n",
       "  ('Flat Cross Product', '__CROSS_PRODUCT_FLAT__'),\n",
       "  ('Nested Cross Product', '__CROSS_PRODUCT_NESTED__'),\n",
       "  ('Tag elements', '__TAG_FROM_FILE__'),\n",
       "  ('Apply rules', '__APPLY_RULES__'),\n",
       "  ('Build list', '__BUILD_LIST__'),\n",
       "  ('Extract dataset', '__EXTRACT_DATASET__'),\n",
       "  ('Duplicate file to collection', '__DUPLICATE_FILE_TO_COLLECTION__'),\n",
       "  ('Parse parameter value', 'param_value_from_file'),\n",
       "  ('Add column', 'addValue'),\n",
       "  ('Concatenate datasets', 'cat1'),\n",
       "  ('Cut', 'Cut1'),\n",
       "  ('Convert', 'Convert characters1'),\n",
       "  ('Create single interval', 'createInterval'),\n",
       "  ('Change Case', 'ChangeCase'),\n",
       "  ('Paste', 'Paste1'),\n",
       "  ('Remove beginning', 'Remove beginning1'),\n",
       "  ('Select random lines', 'random_lines1'),\n",
       "  ('Select first', 'Show beginning1'),\n",
       "  ('Select last', 'Show tail1'),\n",
       "  ('Trim', 'trimmer'),\n",
       "  ('Line/Word/Character count', 'wc_gnu'),\n",
       "  ('Secure Hash / Message Digest', 'secure_hash_message_digest'),\n",
       "  ('BED-to-GFF', 'bed2gff1'),\n",
       "  ('GFF-to-BED', 'gff2bed1'),\n",
       "  ('MAF to BED', 'MAF_To_BED1'),\n",
       "  ('MAF to Interval', 'MAF_To_Interval1'),\n",
       "  ('MAF to FASTA', 'MAF_To_Fasta1'),\n",
       "  ('SFF converter', 'Sff_extractor'),\n",
       "  ('BED-to-bigBed', 'bed_to_bigBed'),\n",
       "  ('Filter', 'Filter1'),\n",
       "  ('Sort', 'sort1'),\n",
       "  ('Select', 'Grep1'),\n",
       "  ('Extract features', 'Extract_features1'),\n",
       "  ('Filter GFF data by attribute', 'gff_filter_by_attribute'),\n",
       "  ('Filter GFF data by feature count', 'gff_filter_by_feature_count'),\n",
       "  ('Filter GTF data by attribute values_list',\n",
       "   'gtf_filter_by_attribute_values_list'),\n",
       "  ('Join two Datasets', 'join1'),\n",
       "  ('Compare two Datasets', 'comp1'),\n",
       "  ('Group', 'Grouping1'),\n",
       "  ('AXT to concatenated FASTA', 'axt_to_concat_fasta'),\n",
       "  ('AXT to FASTA', 'axt_to_fasta'),\n",
       "  ('AXT to LAV', 'axt_to_lav_1'),\n",
       "  ('LAV to BED', 'lav_to_bed1'),\n",
       "  ('Wiggle-to-Interval', 'wiggle2simple1'),\n",
       "  ('GTF-to-BEDGraph', 'gtf2bedgraph'),\n",
       "  ('Convert genome coordinates', 'liftOver1'),\n",
       "  ('Convert', 'solid2fastq'),\n",
       "  ('Compute quality statistics', 'solid_qual_stats'),\n",
       "  ('Draw quality score boxplot', 'solid_qual_boxplot'),\n",
       "  ('Aggregate datapoints', 'aggregate_scores_in_intervals2'),\n",
       "  ('Gene BED To Exon/Intron/Codon BED', 'gene2exon1'),\n",
       "  ('Extract Pairwise MAF blocks', 'Interval2Maf_pairwise1'),\n",
       "  ('Extract MAF blocks', 'Interval2Maf1'),\n",
       "  ('Split MAF blocks', 'MAF_split_blocks_by_species1'),\n",
       "  ('Stitch MAF blocks', 'Interval_Maf_Merged_Fasta2'),\n",
       "  ('Stitch Gene blocks', 'GeneBed_Maf_Fasta2'),\n",
       "  ('MAF Coverage Stats', 'maf_stats1'),\n",
       "  ('Join MAF blocks', 'MAF_Thread_For_Species1'),\n",
       "  ('Filter MAF blocks', 'MAF_Limit_To_Species1'),\n",
       "  ('Filter MAF blocks', 'maf_limit_size1'),\n",
       "  ('Extract MAF by block number', 'maf_by_block_number1'),\n",
       "  ('Reverse Complement', 'MAF_Reverse_Complement_1'),\n",
       "  ('Filter MAF', 'MAF_filter'),\n",
       "  ('SnpSift GeneSets', 'snpSift_geneSets'),\n",
       "  ('velvetg', 'velvetg'),\n",
       "  ('velveth', 'velveth'),\n",
       "  ('Map with BFAST', 'bfast_wrapper'),\n",
       "  ('Map with PerM', 'PerM'),\n",
       "  ('Re-align with SRMA', 'srma_wrapper'),\n",
       "  ('Map with Mosaik', 'mosaik_wrapper'),\n",
       "  ('blockbuster', 'blockbuster'),\n",
       "  ('g:Profiler', 'hgv_linkToGProfile'),\n",
       "  ('DAVID', 'hgv_david'),\n",
       "  ('LD', 'hgv_ldtools'),\n",
       "  ('PASS', 'hgv_pass'),\n",
       "  ('GPASS', 'hgv_gpass'),\n",
       "  ('BEAM', 'hgv_beam'),\n",
       "  ('LPS', 'hgv_lps'),\n",
       "  ('MasterVar to pgSnp', 'master2pgSnp'),\n",
       "  ('X!Tandem MSMS Search', 'proteomics_search_tandem_1'),\n",
       "  ('X!Tandem MSMS Search', 'proteomics_search_tandem_2'),\n",
       "  ('Fragmenter', 'fragmenter'),\n",
       "  ('Merging', 'bgchem_fragment_merger'),\n",
       "  ('Summary Statistics', 'Summary_Statistics1'),\n",
       "  ('Count', 'Count1'),\n",
       "  ('Bar chart', 'barchart_gnuplot'),\n",
       "  ('Boxplot', 'qual_stats_boxplot'),\n",
       "  ('VCF to MAF Custom Track', 'vcf_to_maf_customtrack1'),\n",
       "  ('Interactive JupyterLab Notebook', 'interactive_tool_jupyter_notebook'),\n",
       "  ('RStudio', 'interactive_tool_rstudio'),\n",
       "  ('Interactive CellXgene Environment', 'interactive_tool_cellxgene'),\n",
       "  ('Interactive CellXgene VIP Environment', 'interactive_tool_cellxgene_vip'),\n",
       "  ('iSEE', 'interactive_tool_isee'),\n",
       "  ('NetCDF to Zarr', 'netcdf2zarr'),\n",
       "  ('zeppelin', 'interactive_zeppelin_notebook'),\n",
       "  ('Concatenate Icenet Data', 'concatenate_icenet_data'),\n",
       "  ('Icenet Forecast', 'icenet_forecast'),\n",
       "  ('Preprocess Icenet Data', 'preprocess_icenet_data'),\n",
       "  ('create API requests for Icenet Data',\n",
       "   'create_API_requests_for_icenet_data'),\n",
       "  ('Visualize Icenet Forecast', 'visualize_icenet_forecast'),\n",
       "  ('PyIron', 'pyiron_meta'),\n",
       "  ('Personal Cancer Genome Reporter', 'pcgr'),\n",
       "  ('Tombo Re-squiggle', 'tombo_resquiggle'),\n",
       "  ('Tombo create', 'tombo_text_output'),\n",
       "  ('Tombo detect modifications', 'tombo_detect_modifications'),\n",
       "  ('Tombo plot most signficant', 'tombo_plot_most_significant'),\n",
       "  ('Tombo plot genome locations', 'tombo_plot_genome_locations'),\n",
       "  ('Peptide Prophet', 'proteomics_search_peptide_prophet_1'),\n",
       "  ('Protein Prophet', 'proteomics_search_protein_prophet_1'),\n",
       "  ('InterProphet', 'proteomics_search_interprophet_1'),\n",
       "  ('bwtool-lift', 'bwtool-lift'),\n",
       "  ('Calculate protein properties', 'bg_protein_properties'),\n",
       "  ('Base Count', 'biopython_base_distribution'),\n",
       "  ('Codon count', 'codon_count'),\n",
       "  ('GenBank and EBML', 'genbank_to_gff'),\n",
       "  ('Extract sequences from GenBank', 'gbk_to_orf'),\n",
       "  ('GFF converter', 'gff_to_sequence'),\n",
       "  ('Sequence converter', 'sequence_convert'),\n",
       "  ('Translate nucleotides', 'translate_nucleotides'),\n",
       "  ('RNALfoldz', 'rnalfoldz'),\n",
       "  ('Terrabyte OGC API Processes', 'terrabyte_ogc_api_processes'),\n",
       "  ('Interactive Source Notebooks', 'interactive_tool_source'),\n",
       "  ('Run Napari', 'interactive_tool_napari'),\n",
       "  ('Run CellProfiler', 'interactive_tool_cellprofiler'),\n",
       "  ('tadviewer', 'interactive_tool_tadviewer'),\n",
       "  ('QGIS', 'interactive_tool_qgis'),\n",
       "  ('QuPath', 'interactive_tool_qupath'),\n",
       "  ('pampa', 'interactive_tool_pampa'),\n",
       "  ('ODV', 'interactive_tool_odv'),\n",
       "  ('Ilastik', 'interactive_tool_ilastik'),\n",
       "  ('metashrimps', 'interactive_tool_metashrimps'),\n",
       "  ('Interactive Hdfview tool', 'interactive_tool_hdfview'),\n",
       "  ('Scoop3-Argo', 'interactive_tool_scoop3_argo'),\n",
       "  ('Interactive DIVAnd Notebooks', 'interactive_tool_divand'),\n",
       "  ('RStudio AskoR', 'interactive_tool_rstudio_askor'),\n",
       "  ('Interactive BlobToolKit', 'interactive_tool_blobtoolkit'),\n",
       "  ('Pavian', 'interactive_tool_pavian'),\n",
       "  ('PhysiCell Studio', 'interactive_tool_pcstudio'),\n",
       "  ('COPASI', 'interactive_tool_copasi'),\n",
       "  ('Interactive MGnify Notebook', 'interactive_tool_mgnify_notebook'),\n",
       "  ('Interactive JupyterGIS Notebook', 'interactive_tool_jupytergis_notebook'),\n",
       "  ('GPU-enabled Interactive Jupyter Notebook for Machine Learning',\n",
       "   'interactive_tool_ml_jupyter_notebook'),\n",
       "  ('Interactive Climate Notebook', 'interactive_tool_climate_notebook'),\n",
       "  ('Interactive Pangeo Notebook', 'interactive_tool_pangeo_notebook'),\n",
       "  ('Interactive Pangeo Machine Learning Notebook',\n",
       "   'interactive_tool_pangeo_ml_notebook'),\n",
       "  ('CellPose Interactive', 'interactive_tool_cellpose'),\n",
       "  ('PyIron Interactive Jupyter Notebook', 'interactive_tool_pyiron'),\n",
       "  ('HiGlass', 'interactive_tool_higlass'),\n",
       "  ('OpenRefine', 'interactive_tool_openrefine'),\n",
       "  ('Panoply', 'interactive_tool_panoply'),\n",
       "  ('AskOmics', 'interactive_tool_askomics'),\n",
       "  ('bam.iobio visualisation', 'interactive_tool_bam_iobio'),\n",
       "  ('VCF (iobio) Visualisation', 'interactive_tool_vcf_iobio'),\n",
       "  ('Neo4j (Graph Database)', 'interactive_tool_neo4j'),\n",
       "  ('Phinch Visualisation', 'interactive_tool_phinch'),\n",
       "  ('Paraview', 'interactive_tool_paraview'),\n",
       "  ('Wilson', 'interactive_tool_wilson'),\n",
       "  ('Wallace', 'interactive_tool_wallace'),\n",
       "  ('geoexplorer', 'interactive_tool_geoexplorer'),\n",
       "  ('radiant', 'interactive_tool_radiant'),\n",
       "  ('EtherCalc', 'interactive_tool_ethercalc'),\n",
       "  ('VRM Editor', 'interactive_tool_vrm_editor'),\n",
       "  ('SimText', 'interactive_tool_simtext_app'),\n",
       "  ('IDEAL', 'interactive_tool_ideal'),\n",
       "  ('Phyloseq', 'interactive_tool_phyloseq'),\n",
       "  ('metashark', 'interactive_tool_metashark'),\n",
       "  ('NEAL', 'interactive_tool_audiolabeler'),\n",
       "  ('GeneNoteBook', 'interactive_tool_genenotebook'),\n",
       "  ('Holoviz', 'interactive_tool_holoviz'),\n",
       "  ('Copernicus Data Space Ecosystem', 'interactive_tool_copernicus_notebook'),\n",
       "  ('Qiskit Jupyter notebook', 'interactive_tool_qiskit_jupyter_notebook'),\n",
       "  ('STAC catalog', 'interactive_tool_stac'),\n",
       "  ('RNA STAR', 'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.8a'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.8a'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.17+galaxy4'),\n",
       "  ('LASTZ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_wrapper_2/1.3.3'),\n",
       "  ('LASTZ_D',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_d_wrapper/1.3.2'),\n",
       "  ('Winnowmap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/winnowmap/winnowmap/2.02+galaxy0'),\n",
       "  ('Map with BWA-MEM',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bwa/bwa_mem/0.7.17.2'),\n",
       "  ('Map with BWA', 'toolshed.g2.bx.psu.edu/repos/devteam/bwa/bwa/0.7.17.4'),\n",
       "  ('Winnowmap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/winnowmap/winnowmap/2.03+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.20+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.20.4+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.19.7.0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.20+galaxy1'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.20+galaxy2'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.8a+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.20.1+galaxy0'),\n",
       "  ('Winnowmap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/winnowmap/winnowmap/2.03+galaxy1'),\n",
       "  ('HISAT2', 'toolshed.g2.bx.psu.edu/repos/iuc/hisat2/hisat2/2.2.1+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.20.1+galaxy1'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.20.1+galaxy2'),\n",
       "  ('BWA-MEM2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bwa_mem2/bwa_mem2/2.2.1+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.22+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.20.2+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.23+galaxy0'),\n",
       "  ('LASTZ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_wrapper_2/1.04.15+galaxy0'),\n",
       "  ('LASTZ_D',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_d_wrapper/1.04.15+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.24+galaxy0'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/1.0.0+galaxy3'),\n",
       "  ('Bowtie2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/bowtie2/2.4.5+galaxy0'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/1.0.0+galaxy4'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/1.0.0+galaxy1.0.0'),\n",
       "  ('HISAT2', 'toolshed.g2.bx.psu.edu/repos/iuc/hisat2/hisat2/2.2.1+galaxy1'),\n",
       "  ('Bowtie2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/bowtie2/2.4.5+galaxy1'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.8a+galaxy1'),\n",
       "  ('Bowtie2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/bowtie2/2.5.0+galaxy0'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.10b+galaxy1'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.01+galaxy0'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.10b+galaxy1'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.10b+galaxy2'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.10b+galaxy2'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.10b+galaxy3'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.10b+galaxy3'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.26+galaxy0'),\n",
       "  ('LASTZ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_wrapper_2/1.04.22+galaxy0'),\n",
       "  ('LASTZ_D',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/lastz/lastz_d_wrapper/1.04.22+galaxy0'),\n",
       "  ('BWA-MEM2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bwa_mem2/bwa_mem2/2.2.1+galaxy1'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.10b+galaxy4'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.10b+galaxy4'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.06+galaxy0'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.11a+galaxy0'),\n",
       "  ('pbmm2', 'toolshed.g2.bx.psu.edu/repos/iuc/pbmm2/pbmm2/1.10.0+galaxy0'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.11a+galaxy0'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.21.2+galaxy0'),\n",
       "  ('Bowtie2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/bowtie2/2.5.3+galaxy0'),\n",
       "  ('pbmm2', 'toolshed.g2.bx.psu.edu/repos/iuc/pbmm2/pbmm2/1.13.1+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.27+galaxy0'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.28+galaxy0'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.06+galaxy1'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.06+galaxy2'),\n",
       "  ('bwameth',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bwameth/bwameth/0.2.7+galaxy0'),\n",
       "  ('Bowtie2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/bowtie2/2.5.3+galaxy1'),\n",
       "  ('Map with BWA-MEM',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/bwa/bwa_mem/0.7.18'),\n",
       "  ('Map with BWA', 'toolshed.g2.bx.psu.edu/repos/devteam/bwa/bwa/0.7.18'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.21.7+galaxy0'),\n",
       "  ('Map with KMA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/kma/kma_map/1.4.14+galaxy0'),\n",
       "  ('pbmm2', 'toolshed.g2.bx.psu.edu/repos/iuc/pbmm2/pbmm2/1.14.99+galaxy0'),\n",
       "  ('RNA STAR',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rgrnastar/rna_star/2.7.11a+galaxy1'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.08+galaxy0'),\n",
       "  ('RNA STARSolo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rna_starsolo/rna_starsolo/2.7.11a+galaxy1'),\n",
       "  ('Align sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bioext_bealign/bioext_bealign/0.21.9+galaxy0'),\n",
       "  ('pbmm2', 'toolshed.g2.bx.psu.edu/repos/iuc/pbmm2/pbmm2/1.16.0+galaxy0'),\n",
       "  ('BBTools: BBMap',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmap/bbtools_bbmap/39.08+galaxy1'),\n",
       "  ('Map with minimap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/minimap2/minimap2/2.28+galaxy1'),\n",
       "  ('pbmm2', 'toolshed.g2.bx.psu.edu/repos/iuc/pbmm2/pbmm2/1.16.99+galaxy0'),\n",
       "  ('Translate BED Sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/jjohnson/translate_bed_sequences/translate_bed_sequences/0.0.1'),\n",
       "  ('Filter BED on splice junctions',\n",
       "   'toolshed.g2.bx.psu.edu/repos/jjohnson/filter_bed_on_splice_junctions/filter_bed_on_splice_junctions/0.0.1'),\n",
       "  ('Venn Diagram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/venn_list/venn_list/0.0.3'),\n",
       "  ('JBrowse - Data Directory to Standalone',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/jbrowse/jbrowse_to_standalone/1.16.11+galaxy0'),\n",
       "  ('JBrowse',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/jbrowse/jbrowse/1.16.11+galaxy0'),\n",
       "  ('Paired-end histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pe_histogram/pe_histogram/1.0.1'),\n",
       "  ('Pretext Snapshot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pretext_snapshot/pretext_snapshot/0.0.3+galaxy0'),\n",
       "  ('JBrowse - Data Directory to Standalone',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/jbrowse/jbrowse_to_standalone/1.16.11+galaxy1'),\n",
       "  ('JBrowse',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/jbrowse/jbrowse/1.16.11+galaxy1'),\n",
       "  ('Volcano Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/volcanoplot/volcanoplot/0.0.4'),\n",
       "  ('Volcano Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/volcanoplot/volcanoplot/0.0.5'),\n",
       "  ('Pairwise intersection',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/intervene/intervene_pairwise/0.6.5'),\n",
       "  ('UpSet diagram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/intervene/intervene_upset/0.6.5'),\n",
       "  ('Bandage Image',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bandage/bandage_image/0.8.1+galaxy3'),\n",
       "  ('Bandage Info',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bandage/bandage_info/0.8.1+galaxy2'),\n",
       "  ('Circos: Alignments to links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_aln_to_links/0.69.8+galaxy10'),\n",
       "  ('Circos: Link Density Track',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_binlinks/0.69.8+galaxy10'),\n",
       "  ('Circos: Bundle Links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_bundlelinks/0.69.8+galaxy10'),\n",
       "  ('Circos', 'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos/0.69.8+galaxy10'),\n",
       "  ('GC Skew',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_gc_skew/0.69.8+galaxy10'),\n",
       "  ('Circos: Interval to Tiles',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_tile/0.69.8+galaxy10'),\n",
       "  ('Circos: Resample 1/2D data',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_resample/0.69.8+galaxy10'),\n",
       "  ('Circos: bigWig to Scatter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_scatter/0.69.8+galaxy10'),\n",
       "  ('Circos: Stack bigWigs as Histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_stacked/0.69.8+galaxy10'),\n",
       "  ('Circos: Table viewer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_tableviewer/0.69.8+galaxy10'),\n",
       "  ('Circos: Interval to Circos Text Labels',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_text/0.69.8+galaxy10'),\n",
       "  ('Circos: Alignments to links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_aln_to_links/0.69.8+galaxy8'),\n",
       "  ('Circos: Link Density Track',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_binlinks/0.69.8+galaxy8'),\n",
       "  ('Circos: Bundle Links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_bundlelinks/0.69.8+galaxy8'),\n",
       "  ('GC Skew',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_gc_skew/0.69.8+galaxy8'),\n",
       "  ('Circos: Resample 1/2D data',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_resample/0.69.8+galaxy8'),\n",
       "  ('Circos: bigWig to Scatter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_scatter/0.69.8+galaxy8'),\n",
       "  ('Circos: Stack bigWigs as Histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_stacked/0.69.8+galaxy8'),\n",
       "  ('Circos: Table viewer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_tableviewer/0.69.8+galaxy8'),\n",
       "  ('Circos: Interval to Circos Text Labels',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_text/0.69.8+galaxy8'),\n",
       "  ('Circos: Interval to Tiles',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_tile/0.69.8+galaxy8'),\n",
       "  ('Pretext Snapshot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pretext_snapshot/pretext_snapshot/0.0.3+galaxy1'),\n",
       "  ('Circos: Alignments to links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_aln_to_links/0.69.8+galaxy9'),\n",
       "  ('Circos: Link Density Track',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_binlinks/0.69.8+galaxy9'),\n",
       "  ('Circos: Bundle Links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_bundlelinks/0.69.8+galaxy9'),\n",
       "  ('Circos', 'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos/0.69.8+galaxy9'),\n",
       "  ('GC Skew',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_gc_skew/0.69.8+galaxy9'),\n",
       "  ('Circos: Resample 1/2D data',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_resample/0.69.8+galaxy9'),\n",
       "  ('Circos: bigWig to Scatter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_scatter/0.69.8+galaxy9'),\n",
       "  ('Circos: Stack bigWigs as Histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_stacked/0.69.8+galaxy9'),\n",
       "  ('Circos: Table viewer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_tableviewer/0.69.8+galaxy9'),\n",
       "  ('Circos: Interval to Circos Text Labels',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_text/0.69.8+galaxy9'),\n",
       "  ('Circos: Interval to Tiles',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_tile/0.69.8+galaxy9'),\n",
       "  ('PCA plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_pca/ggplot2_pca/3.3.5+galaxy0'),\n",
       "  ('pyGenomeTracks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pygenometracks/pygenomeTracks/3.6+galaxy1'),\n",
       "  ('Build custom track',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/ucsc_custom_track/build_ucsc_custom_track_1/1.0.1'),\n",
       "  ('Pathview',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pathview/pathview/1.24.0+galaxy2'),\n",
       "  ('heatmap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap2/ggplot2_heatmap2/3.1.1+galaxy1'),\n",
       "  ('Pathview',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pathview/pathview/1.34.0+galaxy0'),\n",
       "  ('Pairwise intersection',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/intervene/intervene_pairwise/0.6.5+galaxy1'),\n",
       "  ('Pathifier',\n",
       "   'toolshed.g2.bx.psu.edu/repos/artbio/pathifier/pathifier/1.0.1'),\n",
       "  ('pyGenomeTracks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pygenometracks/pygenomeTracks/3.7+galaxy0'),\n",
       "  ('Pairwise intersection',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/intervene/intervene_pairwise/0.6.5+galaxy2'),\n",
       "  ('UpSet diagram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/intervene/intervene_upset/0.6.5+galaxy2'),\n",
       "  ('Bandage Image',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bandage/bandage_image/0.8.1+galaxy4'),\n",
       "  ('PCA plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_pca/ggplot2_pca/3.3.5+galaxy1'),\n",
       "  ('Heatmap w ggplot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap/ggplot2_heatmap/3.3.5+galaxy1'),\n",
       "  ('Violin plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_violin/ggplot2_violin/3.3.5+galaxy1'),\n",
       "  ('Scatterplot with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_point/ggplot2_point/3.3.5+galaxy1'),\n",
       "  ('Histogram with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_histogram/ggplot2_histogram/3.3.5+galaxy1'),\n",
       "  ('PCA plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_pca/ggplot2_pca/3.3.5+galaxy2'),\n",
       "  ('Heatmap w ggplot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap/ggplot2_heatmap/3.3.5+galaxy2'),\n",
       "  ('Violin plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_violin/ggplot2_violin/3.3.5+galaxy2'),\n",
       "  ('Scatterplot with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_point/ggplot2_point/3.3.5+galaxy2'),\n",
       "  ('Histogram with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_histogram/ggplot2_histogram/3.3.5+galaxy2'),\n",
       "  ('Bandage Image',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bandage/bandage_image/2022.09+galaxy4'),\n",
       "  ('Bandage Info',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bandage/bandage_info/2022.09+galaxy2'),\n",
       "  ('PCA plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_pca/ggplot2_pca/3.4.0+galaxy0'),\n",
       "  ('Heatmap w ggplot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap/ggplot2_heatmap/3.4.0+galaxy0'),\n",
       "  ('Violin plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_violin/ggplot2_violin/3.4.0+galaxy0'),\n",
       "  ('Scatterplot with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_point/ggplot2_point/3.4.0+galaxy0'),\n",
       "  ('Histogram with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_histogram/ggplot2_histogram/3.4.0+galaxy0'),\n",
       "  ('pyGenomeTracks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pygenometracks/pygenomeTracks/3.8+galaxy0'),\n",
       "  ('Scatterplot with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_point/ggplot2_point/3.4.0+galaxy1'),\n",
       "  ('heatmap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap2/ggplot2_heatmap2/3.1.3+galaxy0'),\n",
       "  ('pyGenomeTracks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pygenometracks/pygenomeTracks/3.8+galaxy1'),\n",
       "  ('Circos: Alignments to links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_aln_to_links/0.69.8+galaxy12'),\n",
       "  ('Circos: Link Density Track',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_binlinks/0.69.8+galaxy12'),\n",
       "  ('Circos: Bundle Links',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_bundlelinks/0.69.8+galaxy12'),\n",
       "  ('Circos', 'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos/0.69.8+galaxy12'),\n",
       "  ('GC Skew',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_gc_skew/0.69.8+galaxy12'),\n",
       "  ('Circos: Resample 1/2D data',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_resample/0.69.8+galaxy12'),\n",
       "  ('Circos: bigWig to Scatter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_scatter/0.69.8+galaxy12'),\n",
       "  ('Circos: Stack bigWigs as Histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_wiggle_to_stacked/0.69.8+galaxy12'),\n",
       "  ('Circos: Table viewer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_tableviewer/0.69.8+galaxy12'),\n",
       "  ('Circos: Interval to Circos Text Labels',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_text/0.69.8+galaxy12'),\n",
       "  ('Circos: Interval to Tiles',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/circos/circos_interval_to_tile/0.69.8+galaxy12'),\n",
       "  ('pyGenomeTracks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pygenometracks/pygenomeTracks/3.8+galaxy2'),\n",
       "  ('gene.iobio visualisation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/geneiobio/gene_iobio_display_generation_iframe/4.7.1'),\n",
       "  ('gene.iobio visualisation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/geneiobio/gene_iobio_display_generation_iframe/4.7.1+galaxy1'),\n",
       "  ('Linear Genome Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_linear_genome_plot/edu.tamu.cpt.genome_viz.linear_genome_plot/1.0'),\n",
       "  ('Pretext Snapshot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pretext_snapshot/pretext_snapshot/0.0.3+galaxy2'),\n",
       "  ('heatmap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap2/ggplot2_heatmap2/3.1.3.1+galaxy0'),\n",
       "  ('Plot confusion matrix, precision, recall and ROC and AUC curves',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/plotly_ml_performance_plots/plotly_ml_performance_plots/0.3'),\n",
       "  ('Volcano Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/volcanoplot/volcanoplot/0.0.6'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.11.1+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.12.2+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.12.3+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.13.0+galaxy0'),\n",
       "  ('Plot confusion matrix, precision, recall and ROC and AUC curves',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/plotly_ml_performance_plots/plotly_ml_performance_plots/0.4'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.13.1+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.13.1+galaxy1'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.14.0+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.15.4+galaxy0'),\n",
       "  ('PCA plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_pca/ggplot2_pca/3.4.0+galaxy1'),\n",
       "  ('Heatmap w ggplot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap/ggplot2_heatmap/3.4.0+galaxy1'),\n",
       "  ('Violin plot w ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_violin/ggplot2_violin/3.4.0+galaxy1'),\n",
       "  ('Histogram with ggplot2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_histogram/ggplot2_histogram/3.4.0+galaxy1'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.15.4+galaxy4'),\n",
       "  ('Paired-end histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pe_histogram/pe_histogram/1.0.2'),\n",
       "  ('heatmap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap2/ggplot2_heatmap2/3.2.0+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.15.4+galaxy5'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.15.4+galaxy7'),\n",
       "  ('Pretext Snapshot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pretext_snapshot/pretext_snapshot/0.0.4+galaxy0'),\n",
       "  ('heatmap2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ggplot2_heatmap2/ggplot2_heatmap2/3.2.0+galaxy1'),\n",
       "  ('Generate a word cloud',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/wordcloud/wordcloud/1.9.4+galaxy0'),\n",
       "  ('JBrowse2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/fubar/jbrowse2/jbrowse2/2.17.0+galaxy0'),\n",
       "  ('Volcano Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/volcanoplot/volcanoplot/0.0.7'),\n",
       "  ('Histogram',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/histogram/histogram_rpy/1.0.5'),\n",
       "  ('FastQC',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastqc/fastqc/0.74+galaxy0'),\n",
       "  ('Trim Galore',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/trim_galore/trim_galore/0.2.8.1'),\n",
       "  ('filtlong',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/filtlong/filtlong/0.2.0+galaxy1'),\n",
       "  ('Fasta Statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/fasta_stats/fasta-stats/1.0.2'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.4+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.4+galaxy1'),\n",
       "  ('Fasta Statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/fasta_stats/fasta-stats/1.0.3'),\n",
       "  ('FASTQE', 'toolshed.g2.bx.psu.edu/repos/iuc/fastqe/fastqe/0.2.6+galaxy2'),\n",
       "  ('RNA-seq Rcorrector',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/rcorrector/rcorrector/1.0.3+galaxy1'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.2+galaxy0'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.2+galaxy0'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.2+galaxy0'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.2+galaxy0'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.2+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.4+galaxy2'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.2+galaxy1'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.2+galaxy1'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.2+galaxy1'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.2+galaxy1'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.2+galaxy1'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.5+galaxy0'),\n",
       "  ('Fasta Statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/fasta_stats/fasta-stats/2.0'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.2+galaxy2'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.2+galaxy2'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.2+galaxy2'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.2+galaxy2'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.2+galaxy2'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.23.2+galaxy0'),\n",
       "  ('faSplit', 'toolshed.g2.bx.psu.edu/repos/iuc/ucsc_fasplit/fasplit/377'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.5+galaxy2'),\n",
       "  ('filtlong',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/filtlong/filtlong/0.2.1+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/3.7+galaxy0'),\n",
       "  ('FASTQ splitter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_splitter/fastq_paired_end_splitter/1.1.5+galaxy1'),\n",
       "  ('Filter sequences by mapping',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/seq_filter_by_mapping/seq_filter_by_mapping/0.0.7'),\n",
       "  ('Pear', 'toolshed.g2.bx.psu.edu/repos/iuc/pear/iuc_pear/0.9.6.2'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.0+galaxy0'),\n",
       "  ('Calculate sequence complexity',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqcomplexity/seqcomplexity/0.1.0'),\n",
       "  ('SeqKit statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_stats/seqkit_stats/2.2.0+galaxy0'),\n",
       "  ('SeqKit fx2tab',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_fx2tab/seqkit_fx2tab/2.2.0+galaxy0'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.0+galaxy0'),\n",
       "  ('AdapterRemoval: remove adapter sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/adapter_removal/adapter_removal/2.3.3+galaxy0'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.4+galaxy0'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.5+galaxy0'),\n",
       "  ('Calculate sequence complexity',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqcomplexity/seqcomplexity/0.1.2'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.0+galaxy1'),\n",
       "  ('SeqKit fx2tab',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_fx2tab/seqkit_fx2tab/2.3.1+galaxy0'),\n",
       "  ('SeqKit statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_stats/seqkit_stats/2.3.1+galaxy0'),\n",
       "  ('SeqKit fx2tab',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_fx2tab/seqkit_fx2tab/2.3.1+galaxy1'),\n",
       "  ('SeqKit statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_stats/seqkit_stats/2.3.1+galaxy1'),\n",
       "  ('SeqKit locate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_locate/seqkit_locate/2.3.1+galaxy1'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.5+galaxy2'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.6+galaxy0'),\n",
       "  ('Fasta regular expression finder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/mbernt/fasta_regex_finder/fasta_regex_finder/0.1.0'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.01+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.4+galaxy0'),\n",
       "  ('FASTA Width',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fasta_formatter/cshl_fasta_formatter/1.0.1+galaxy2'),\n",
       "  ('RNA/DNA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fasta_nucleotide_changer/cshl_fasta_nucleotides_changer/1.0.2+galaxy2'),\n",
       "  ('Draw quality score boxplot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_quality_boxplot/cshl_fastq_quality_boxplot/1.0.1+galaxy2'),\n",
       "  ('Quality format converter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_quality_converter/cshl_fastq_quality_converter/1.0.1+galaxy2'),\n",
       "  ('Filter by quality',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_quality_filter/cshl_fastq_quality_filter/1.0.2+galaxy2'),\n",
       "  ('FASTQ to FASTA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_to_fasta/cshl_fastq_to_fasta/1.0.2+galaxy2'),\n",
       "  ('Remove sequencing artifacts',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_artifacts_filter/cshl_fastx_artifacts_filter/1.0.1+galaxy2'),\n",
       "  ('Barcode Splitter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_barcode_splitter/cshl_fastx_barcode_splitter/1.0.1+galaxy2'),\n",
       "  ('Clip',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_clipper/cshl_fastx_clipper/1.0.3+galaxy2'),\n",
       "  ('Collapse',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_collapser/cshl_fastx_collapser/1.0.1+galaxy2'),\n",
       "  ('Draw nucleotides distribution chart',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_nucleotides_distribution/cshl_fastx_nucleotides_distribution/1.0.1+galaxy2'),\n",
       "  ('Compute quality statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_quality_statistics/cshl_fastx_quality_statistics/1.0.1+galaxy2'),\n",
       "  ('Rename sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_renamer/cshl_fastx_renamer/0.0.14+galaxy2'),\n",
       "  ('Reverse-Complement',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_reverse_complement/cshl_fastx_reverse_complement/1.0.2+galaxy2'),\n",
       "  ('Trim sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastx_trimmer/cshl_fastx_trimmer/1.0.2+galaxy2'),\n",
       "  ('Remove Description',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_fasta_remove_id/edu.tamu.cpt.fasta.remove_desc/19.1.0.0'),\n",
       "  ('Trimmomatic',\n",
       "   'toolshed.g2.bx.psu.edu/repos/pjbriggs/trimmomatic/trimmomatic/0.39+galaxy0'),\n",
       "  ('Filter sequences by mapping',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/seq_filter_by_mapping/seq_filter_by_mapping/0.0.8'),\n",
       "  ('seqtk_comp',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_comp/1.4+galaxy0'),\n",
       "  ('seqtk_cutN',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_cutN/1.4+galaxy0'),\n",
       "  ('seqtk_dropse',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_dropse/1.4+galaxy0'),\n",
       "  ('seqtk_fqchk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_fqchk/1.4+galaxy0'),\n",
       "  ('seqtk_hety',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_hety/1.4+galaxy0'),\n",
       "  ('seqtk_listhet',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_listhet/1.4+galaxy0'),\n",
       "  ('seqtk_mergefa',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_mergefa/1.4+galaxy0'),\n",
       "  ('seqtk_mergepe',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_mergepe/1.4+galaxy0'),\n",
       "  ('seqtk_mutfa',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_mutfa/1.4+galaxy0'),\n",
       "  ('seqtk_randbase',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_randbase/1.4+galaxy0'),\n",
       "  ('seqtk_sample',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_sample/1.4+galaxy0'),\n",
       "  ('seqtk_seq',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_seq/1.4+galaxy0'),\n",
       "  ('seqtk_subseq',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_subseq/1.4+galaxy0'),\n",
       "  ('seqtk_telo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_telo/1.4+galaxy0'),\n",
       "  ('seqtk_trimfq',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_trimfq/1.4+galaxy0'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.01+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.6+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.6+galaxy1'),\n",
       "  ('FASTQE', 'toolshed.g2.bx.psu.edu/repos/iuc/fastqe/fastqe/0.3.1+galaxy0'),\n",
       "  ('BBTools: BBduk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbduk/bbtools_bbduk/39.06+galaxy0'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.06+galaxy0'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.06+galaxy0'),\n",
       "  ('Merging paired-end Illumina reads (SeqPrep)',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqprep/seqprep/1.3.2+galaxy0'),\n",
       "  ('Trimmomatic',\n",
       "   'toolshed.g2.bx.psu.edu/repos/pjbriggs/trimmomatic/trimmomatic/0.39+galaxy2'),\n",
       "  ('rasusa', 'toolshed.g2.bx.psu.edu/repos/iuc/rasusa/rasusa/0.8.0+galaxy0'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.23.4+galaxy0'),\n",
       "  ('NCBI FCS GX',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ncbi_fcs_gx/ncbi_fcs_gx/0.5.0+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.7+galaxy0'),\n",
       "  ('BBTools: BBduk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbduk/bbtools_bbduk/39.06+galaxy1'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.06+galaxy1'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.06+galaxy1'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.8+galaxy0'),\n",
       "  ('BBTools: BBduk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbduk/bbtools_bbduk/39.06+galaxy2'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.06+galaxy2'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.06+galaxy2'),\n",
       "  ('ART 454', 'toolshed.g2.bx.psu.edu/repos/iuc/art/art_454/2014.11.03.0'),\n",
       "  ('ART Illumina',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/art/art_illumina/2014.11.03.0'),\n",
       "  ('ART SOLiD', 'toolshed.g2.bx.psu.edu/repos/iuc/art/art_solid/2014.11.03.0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.8+galaxy1'),\n",
       "  ('SeqKit fx2tab',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_fx2tab/seqkit_fx2tab/2.8.1+galaxy0'),\n",
       "  ('SeqKit locate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_locate/seqkit_locate/2.8.1+galaxy0'),\n",
       "  ('SeqKit statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_stats/seqkit_stats/2.8.1+galaxy0'),\n",
       "  ('Falco', 'toolshed.g2.bx.psu.edu/repos/iuc/falco/falco/1.2.2+galaxy0'),\n",
       "  ('Falco', 'toolshed.g2.bx.psu.edu/repos/iuc/falco/falco/1.2.2+galaxy1'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.9+galaxy0'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/4.9+galaxy1'),\n",
       "  ('SeqKit translate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_translate/seqkit_translate/2.8.1+galaxy0'),\n",
       "  ('SeqKit sort',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_sort/seqkit_sort/2.8.1+galaxy0'),\n",
       "  ('FastQC',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastqc/fastqc/0.74+galaxy1'),\n",
       "  ('Pear', 'toolshed.g2.bx.psu.edu/repos/iuc/pear/iuc_pear/0.9.6.3'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.23.4+galaxy1'),\n",
       "  ('NCBI FCS GX',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ncbi_fcs_gx/ncbi_fcs_gx/0.5.4+galaxy0'),\n",
       "  ('AdapterRemoval: remove adapter sequences',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/adapter_removal/adapter_removal/2.3.4+galaxy0'),\n",
       "  ('BBTools: BBduk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbduk/bbtools_bbduk/39.08+galaxy0'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.08+galaxy0'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.08+galaxy0'),\n",
       "  ('Pear', 'toolshed.g2.bx.psu.edu/repos/iuc/pear/iuc_pear/0.9.6.4'),\n",
       "  ('Falco', 'toolshed.g2.bx.psu.edu/repos/iuc/falco/falco/1.2.3+galaxy0'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.2.2+galaxy0'),\n",
       "  ('Length Distribution',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fasta_clipping_histogram/cshl_fasta_clipping_histogram/1.0.1+galaxy2'),\n",
       "  ('rasusa', 'toolshed.g2.bx.psu.edu/repos/iuc/rasusa/rasusa/2.0.0+galaxy0'),\n",
       "  ('Falco', 'toolshed.g2.bx.psu.edu/repos/iuc/falco/falco/1.2.4+galaxy0'),\n",
       "  ('faSplit', 'toolshed.g2.bx.psu.edu/repos/iuc/ucsc_fasplit/fasplit/469'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.5+galaxy0'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.5+galaxy0'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.5+galaxy0'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.5+galaxy0'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.5+galaxy0'),\n",
       "  ('FASTQ Groomer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_groomer/fastq_groomer/1.1.5+galaxy2'),\n",
       "  ('Combine FASTA and QUAL',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_combiner/fastq_combiner/1.1.5+galaxy2'),\n",
       "  ('Compute sequence length',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fasta_compute_length/fasta_compute_length/1.0.4'),\n",
       "  ('Filter FASTQ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_filter/fastq_filter/1.1.5+galaxy2'),\n",
       "  ('FASTQ Masker',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_masker_by_quality/fastq_masker_by_quality/1.1.5+galaxy2'),\n",
       "  ('FASTQ de-interlacer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_deinterlacer/fastq_paired_end_deinterlacer/1.1.5+galaxy2'),\n",
       "  ('Manipulate FASTQ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_manipulation/fastq_manipulation/1.1.5+galaxy2'),\n",
       "  ('FASTQ joiner',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_joiner/fastq_paired_end_joiner/2.0.1.1+galaxy1'),\n",
       "  ('FASTQ splitter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_splitter/fastq_paired_end_splitter/1.1.5+galaxy2'),\n",
       "  ('FASTQ interlacer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_paired_end_interlacer/fastq_paired_end_interlacer/1.1.5+galaxy2'),\n",
       "  ('FASTQ to Tabular',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_to_tabular/fastq_to_tabular/1.1.5+galaxy2'),\n",
       "  ('FASTQ Trimmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_trimmer/fastq_trimmer/1.1.5+galaxy2'),\n",
       "  ('FASTQ to FASTA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastqtofasta/fastq_to_fasta_python/1.1.5+galaxy2'),\n",
       "  ('FASTQ Summary Statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/fastq_stats/fastq_stats/1.1.5+galaxy2'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.5+galaxy1'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.5+galaxy1'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.5+galaxy1'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.5+galaxy1'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.5+galaxy1'),\n",
       "  ('Tabular to FASTQ',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/tabular_to_fastq/tabular_to_fastq/1.1.5+galaxy2'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.8+galaxy0'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.23.4+galaxy2'),\n",
       "  ('seqtk_mergefa',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_mergefa/1.4+galaxy1'),\n",
       "  ('seqtk_seq',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqtk/seqtk_seq/1.4+galaxy1'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.9+galaxy0'),\n",
       "  ('SeqKit fx2tab',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_fx2tab/seqkit_fx2tab/2.9.0+galaxy0'),\n",
       "  ('SeqKit locate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_locate/seqkit_locate/2.9.0+galaxy0'),\n",
       "  ('SeqKit translate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_translate/seqkit_translate/2.9.0+galaxy0'),\n",
       "  ('SeqKit statistics',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_stats/seqkit_stats/2.9.0+galaxy0'),\n",
       "  ('SeqKit sort',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_sort/seqkit_sort/2.9.0+galaxy0'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.24.0+galaxy3'),\n",
       "  ('BBTools: BBduk',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbduk/bbtools_bbduk/39.08+galaxy1'),\n",
       "  ('BBTools: BBMerge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbmerge/bbtools_bbmerge/39.08+galaxy1'),\n",
       "  ('BBTools: Tadpole',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_tadpole/bbtools_tadpole/39.08+galaxy1'),\n",
       "  ('NCBI FCS GX',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/ncbi_fcs_gx/ncbi_fcs_gx/0.5.4+galaxy1'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.9+galaxy1'),\n",
       "  ('UMI-tools count',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_count/umi_tools_count/1.1.6+galaxy0'),\n",
       "  ('UMI-tools deduplicate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_dedup/umi_tools_dedup/1.1.6+galaxy0'),\n",
       "  ('UMI-tools extract',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_extract/umi_tools_extract/1.1.6+galaxy0'),\n",
       "  ('UMI-tools group',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_group/umi_tools_group/1.1.6+galaxy0'),\n",
       "  ('UMI-tools whitelist',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/umi_tools_whitelist/umi_tools_whitelist/1.1.6+galaxy0'),\n",
       "  ('gfastats',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/gfastats/gfastats/1.3.9+galaxy2'),\n",
       "  ('Cutadapt',\n",
       "   'toolshed.g2.bx.psu.edu/repos/lparsons/cutadapt/cutadapt/5.0+galaxy0'),\n",
       "  ('fastp', 'toolshed.g2.bx.psu.edu/repos/iuc/fastp/fastp/0.24.0+galaxy4'),\n",
       "  ('SeqKit Head',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seqkit_head/seqkit_head/2.9.0+galaxy0'),\n",
       "  ('BBTools: BBNorm',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/bbtools_bbnorm/bbtools_bbnorm/39.08+galaxy1'),\n",
       "  ('MDS Scatter Plot',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/ctb_machine_learning/ctb_mds_plot/0.1'),\n",
       "  ('Online data',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/chemical_data_sources/ctb_online_data_fetch/0.2'),\n",
       "  ('PubChem Assay Downloader',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/chemical_data_sources/ctb_pubchem_download_assays/1.0.0'),\n",
       "  ('Open Molecule Generator',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openmg/ctb_openmg/0.1'),\n",
       "  ('Docking',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/autodock_vina/docking/0.1.0'),\n",
       "  ('Prepare receptor',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/autodock_vina_prepare_receptor/prepare_receptor/0.1.0'),\n",
       "  ('Trajectory select and merge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/traj_selections_and_merge/traj_selections_and_merge/1.9.5+galaxy0'),\n",
       "  ('MDTraj file converter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/md_converter/md_converter/1.9.5+galaxy0'),\n",
       "  ('Convert Parameters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/parmconv/parmconv/20.15+galaxy1'),\n",
       "  ('Search ChEMBL database',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/chembl/chembl/0.10.1+galaxy4'),\n",
       "  ('ChEMBL structure pipeline',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/chembl/chembl_structure_pipeline/1.0.0+galaxy0'),\n",
       "  ('Score docked poses using SuCOS',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/sucos_docking_scoring/sucos_docking_scoring/2020.03.4+galaxy1'),\n",
       "  ('GROMACS production simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_md/gmx_md/2018.2'),\n",
       "  ('dcTMD friction correction',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/biomd_neqgamma/biomd_neqgamma/0.1.5.2+galaxy1'),\n",
       "  ('MMPBSA/MMGBSA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/mmpbsa_mmgbsa/mmpbsa_mmgbsa/21.0+galaxy0'),\n",
       "  ('Generate MD topologies for small molecules',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_acpype/ambertools_acpype/21.0+galaxy0'),\n",
       "  ('AnteChamber',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_antechamber/ambertools_antechamber/21.0+galaxy0'),\n",
       "  ('ParmChk2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_parmchk2/ambertools_parmchk2/21.0+galaxy0'),\n",
       "  ('Convert Parameters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/parmconv/parmconv/21.0+galaxy0'),\n",
       "  ('Slice MD trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/md_converter/md_slicer/1.9.6+galaxy0'),\n",
       "  ('Trajectory select and merge',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/traj_selections_and_merge/traj_selections_and_merge/1.9.6+galaxy0'),\n",
       "  ('MMPBSA/MMGBSA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/mmpbsa_mmgbsa/mmpbsa_mmgbsa/21.3+galaxy0'),\n",
       "  ('MDTraj file converter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/md_converter/md_converter/1.9.6+galaxy0'),\n",
       "  ('Generate MD topologies for small molecules',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_acpype/ambertools_acpype/21.3+galaxy0'),\n",
       "  ('AnteChamber',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_antechamber/ambertools_antechamber/21.3+galaxy0'),\n",
       "  ('ParmChk2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_parmchk2/ambertools_parmchk2/21.3+galaxy0'),\n",
       "  ('Convert Parameters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/parmconv/parmconv/21.3+galaxy0'),\n",
       "  ('Calculate the box parameters using RDKit',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/autodock_vina_prepare_box/prepare_box/2021.03.4+galaxy0'),\n",
       "  ('Drug-likeness',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/qed/ctb_silicos_qed/2021.03.4+galaxy0'),\n",
       "  ('GROMACS simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/2020.4+galaxy1'),\n",
       "  ('Pharmacophore alignment',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/align_it/ctb_alignit/1.0.3.1'),\n",
       "  ('Pharmacophore',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/align_it/ctb_alignit_create_db/0.1'),\n",
       "  ('Calculate the box parameters using RDKit',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/autodock_vina_prepare_box/prepare_box/2021.03.5+galaxy0'),\n",
       "  ('Pharmacophore alignment',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/align_it/ctb_alignit/1.0.4+galaxy0'),\n",
       "  ('dpocket',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/fpocket/dpocket/4.0.0+galaxy0'),\n",
       "  ('fpocket',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/fpocket/fpocket/4.0.0+galaxy0'),\n",
       "  ('MMPBSA/MMGBSA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/mmpbsa_mmgbsa/mmpbsa_mmgbsa/21.7+galaxy0'),\n",
       "  ('Generate MD topologies for small molecules',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_acpype/ambertools_acpype/21.7+galaxy0'),\n",
       "  ('AnteChamber',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_antechamber/ambertools_antechamber/21.7+galaxy0'),\n",
       "  ('ParmChk2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_parmchk2/ambertools_parmchk2/21.7+galaxy0'),\n",
       "  ('Convert Parameters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/parmconv/parmconv/21.7+galaxy0'),\n",
       "  ('MMPBSA/MMGBSA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/mmpbsa_mmgbsa/mmpbsa_mmgbsa/21.10+galaxy0'),\n",
       "  ('GROMACS initial setup',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_setup/gmx_setup/2021.3+galaxy0'),\n",
       "  ('GROMACS energy minimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_em/gmx_em/2021.3+galaxy0'),\n",
       "  ('GROMACS solvation and adding ions',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_solvate/gmx_solvate/2021.3+galaxy0'),\n",
       "  ('Merge GROMACS topologies',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_merge_topology_files/gmx_merge_topology_files/3.4.3+galaxy0'),\n",
       "  ('GROMACS simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/2021.3+galaxy1'),\n",
       "  ('Modify/convert GROMACS trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_trj/gmx_trj/2021.3+galaxy0'),\n",
       "  ('Extract energy components with GROMACS',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_energy/gmx_energy/2021.3+galaxy0'),\n",
       "  ('Create GROMACS position restraints files',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_restraints/gmx_restraints/2021.3+galaxy0'),\n",
       "  ('Create GROMACS index files',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_makendx/gmx_makendx/2021.3+galaxy0'),\n",
       "  ('GROMACS structure configuration',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_editconf/gmx_editconf/2021.3+galaxy0'),\n",
       "  ('Generate MD topologies for small molecules',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_acpype/ambertools_acpype/21.10+galaxy0'),\n",
       "  ('AnteChamber',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_antechamber/ambertools_antechamber/21.10+galaxy0'),\n",
       "  ('ParmChk2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/ambertools_parmchk2/ambertools_parmchk2/21.10+galaxy0'),\n",
       "  ('Convert Parameters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/parmconv/parmconv/21.10+galaxy0'),\n",
       "  ('GROMACS energy minimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_em/gmx_em/2021.3+galaxy1'),\n",
       "  ('GROMACS simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/2021.3+galaxy2'),\n",
       "  ('GROMACS solvation and adding ions',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_solvate/gmx_solvate/2021.3+galaxy1'),\n",
       "  ('Slice MD trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/md_converter/md_slicer/1.9.7+galaxy0'),\n",
       "  ('GROMACS simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/2021.3+galaxy3'),\n",
       "  ('GROMACS Radius of Gyration',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_rg/gmx_rg/2021.3+galaxy0'),\n",
       "  ('Convert Amber topology and coordinate files to GROMACS format',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/acpype_amber2gromacs/acpype_Amber2Gromacs/21.10+galaxy0'),\n",
       "  ('VINA Docking',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/autodock_vina/docking/1.2.3+galaxy0'),\n",
       "  ('Build tLEaP',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/tleap/tleap/21.10+galaxy0'),\n",
       "  ('alphafold',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.0.0+galaxy0'),\n",
       "  ('Adding New Topology Information',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gromacs_modify_topology/gromacs_modify_topology/0+galaxy0'),\n",
       "  ('Extracting Topology Information',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gromacs_extract_topology/gromacs_extract_topology/0+galaxy0'),\n",
       "  ('GROMACS copy file',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_get_builtin_file/gmx_get_builtin_file/2021.3+galaxy0'),\n",
       "  ('alphafold',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.0.0+galaxy1'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.1.2+galaxy0'),\n",
       "  ('GROMACS initial setup',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_setup/gmx_setup/2022+galaxy0'),\n",
       "  ('GROMACS energy minimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_em/gmx_em/2022+galaxy0'),\n",
       "  ('GROMACS solvation and adding ions',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_solvate/gmx_solvate/2022+galaxy0'),\n",
       "  ('GROMACS simulation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/2022+galaxy0'),\n",
       "  ('Modify/convert GROMACS trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_trj/gmx_trj/2022+galaxy0'),\n",
       "  ('Extract energy components with GROMACS',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_energy/gmx_energy/2022+galaxy0'),\n",
       "  ('Create GROMACS position restraints files',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_restraints/gmx_restraints/2022+galaxy0'),\n",
       "  ('Create GROMACS index files',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_makendx/gmx_makendx/2022+galaxy0'),\n",
       "  ('GROMACS structure configuration',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_editconf/gmx_editconf/2022+galaxy0'),\n",
       "  ('GROMACS Radius of Gyration',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_rg/gmx_rg/2022+galaxy0'),\n",
       "  ('GROMACS copy file',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_get_builtin_file/gmx_get_builtin_file/2022+galaxy0'),\n",
       "  ('Extract energy components with GROMACS',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_energy/gmx_energy/2022+galaxy1'),\n",
       "  ('PDBFixer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/pdbfixer/pdbfixer/1.8.1+galaxy0'),\n",
       "  ('Modify/convert GROMACS trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_trj/gmx_trj/2022+galaxy1'),\n",
       "  ('GROMACS check',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_check/gmx_check/2022+galaxy0'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.1.2+galaxy1'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.1.2+galaxy3'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.1.2+galaxy4'),\n",
       "  ('Modify/convert and concatate GROMACS trajectories',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_trj/gmx_trj/2022+galaxy2'),\n",
       "  ('GROMACS RMSD calculation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_rmsd/gmx_rmsd/2022+galaxy0'),\n",
       "  ('GROMACS RMSF calculation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_rmsf/gmx_rmsf/2022+galaxy0'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy0'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy1'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy2'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy3'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy4'),\n",
       "  ('xtb molecular optimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/xtb_molecular_optimization/xtb_molecular_optimization/6.6.1+galaxy1'),\n",
       "  ('Remove coordination complexes',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/rem_complex/rem_complex/1.0.0+galaxy2'),\n",
       "  ('QCxMS neutral run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_neutral_run/qcxms_neutral_run/5.2.1+galaxy1'),\n",
       "  ('QCxMS production run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_production_run/qcxms_production_run/5.2.1+galaxy2'),\n",
       "  ('QCxMS get results',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_getres/qcxms_getres/5.2.1+galaxy0'),\n",
       "  ('Remove coordination complexes',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/rem_complex/rem_complex/1.0.0+galaxy3'),\n",
       "  ('QCxMS neutral run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_neutral_run/qcxms_neutral_run/5.2.1+galaxy2'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.1+galaxy5'),\n",
       "  ('QCxMS neutral run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_neutral_run/qcxms_neutral_run/5.2.1+galaxy3'),\n",
       "  ('QCxMS get results',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_getres/qcxms_getres/5.2.1+galaxy1'),\n",
       "  (\"Convert Qiskit's XYZ files to\",\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/qiskit_xyz2pdb/qiskit_xyz2pdb/0.1.2+galaxy0'),\n",
       "  ('PULCHRA recontruction of all atom proteins',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/pulchra/pulchra/3.06+galaxy0'),\n",
       "  ('Align structures and compute relative RMSDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/chemteam/biopdb_align_and_rmsd/biopdb_align_and_rmsd/1.79+galaxy0'),\n",
       "  ('Remove coordination complexes',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/rem_complex/rem_complex/1.0.0+galaxy4'),\n",
       "  ('xtb molecular optimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/xtb_molecular_optimization/xtb_molecular_optimization/6.6.1+galaxy2'),\n",
       "  ('xtb molecular optimization',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/xtb_molecular_optimization/xtb_molecular_optimization/6.6.1+galaxy3'),\n",
       "  ('QCxMS neutral run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_neutral_run/qcxms_neutral_run/5.2.1+galaxy4'),\n",
       "  ('QCxMS production run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_production_run/qcxms_production_run/5.2.1+galaxy3'),\n",
       "  ('QCxMS get results',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_getres/qcxms_getres/5.2.1+galaxy2'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.2+galaxy0'),\n",
       "  ('Compound search',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_obgrep/openbabel_obgrep/3.1.1+galaxy1'),\n",
       "  ('Compute physico-chemical properties',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_genprop/openbabel_genProp/3.1.1+galaxy1'),\n",
       "  ('Visualisation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_svg_depiction/openbabel_svg_depiction/3.1.1+galaxy1'),\n",
       "  ('Spectrophores search',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_spectrophore_search/openbabel_spectrophore_search/3.1.1+galaxy1'),\n",
       "  ('Compound conversion',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_compound_convert/openbabel_compound_convert/3.1.1+galaxy1'),\n",
       "  ('Substructure Search',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/openbabel_subsearch/openbabel_subsearch/3.1.1+galaxy1'),\n",
       "  ('QCxMS neutral run',\n",
       "   'toolshed.g2.bx.psu.edu/repos/recetox/qcxms_neutral_run/qcxms_neutral_run/5.2.1+galaxy5'),\n",
       "  ('Alphafold 2',\n",
       "   'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/2.3.2+galaxy2'),\n",
       "  ('Sanntis biosynthetic gene clusters',\n",
       "   'toolshed.g2.bx.psu.edu/repos/ecology/sanntis_marine/sanntis_marine/0.9.3.5+galaxy0'),\n",
       "  ('CryptoGenotyper',\n",
       "   'toolshed.g2.bx.psu.edu/repos/nml/cryptogenotyper/CryptoGenotyper/1.0+galaxy0'),\n",
       "  ('Augustus', 'toolshed.g2.bx.psu.edu/repos/bgruening/augustus/augustus/0.3'),\n",
       "  ('tRNA prediction',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/trna_prediction/trnascan/0.3'),\n",
       "  ('tRNA and tmRNA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/trna_prediction/aragorn_trna/0.3'),\n",
       "  ('Annotate with DGI',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/dgidb_annotator/dgidb_annotator/0.1'),\n",
       "  ('psortb',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/Psortb/0.0.3'),\n",
       "  ('RXLR Motifs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/rxlr_motifs/0.0.7'),\n",
       "  ('Promoter 2.0',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/promoter2/0.0.6'),\n",
       "  ('TMHMM 2.0',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/tmhmm2/0.0.11'),\n",
       "  ('Annotate with DGI',\n",
       "   'toolshed.g2.bx.psu.edu/repos/devteam/dgidb_annotator/dgidb_annotate/0.1'),\n",
       "  ('HHsearch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/guerler/hhsearch/hhsearch/3.2.0+galaxy0'),\n",
       "  ('sistr_cmd',\n",
       "   'toolshed.g2.bx.psu.edu/repos/nml/sistr_cmd/sistr_cmd/1.1.1+galaxy1'),\n",
       "  ('seq2HLA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/seq2hla/seq2hla/2.2.0+galaxy1'),\n",
       "  ('ABRicate', 'toolshed.g2.bx.psu.edu/repos/iuc/abricate/abricate/1.0.1'),\n",
       "  ('ABRicate List',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/abricate/abricate_list/1.0.1'),\n",
       "  ('ABRicate Summary',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/abricate/abricate_summary/1.0.1'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/3.0.4+galaxy1'),\n",
       "  ('MITOS', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos/mitos/1.0.5+galaxy1'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.0.6+galaxy1'),\n",
       "  ('Glimmer ICM builder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/glimmer_build_icm/glimmer_build_icm/3.02'),\n",
       "  ('Glimmer3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/glimmer_knowledge_based/glimmer_knowledge_based/3.02'),\n",
       "  ('Glimmer3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/glimmer_not_knowledge_based/glimmer_not_knowledge_based/3.02'),\n",
       "  ('goseq', 'toolshed.g2.bx.psu.edu/repos/iuc/goseq/goseq/1.44.0+galaxy0'),\n",
       "  ('Maker', 'toolshed.g2.bx.psu.edu/repos/iuc/maker/maker/2.31.11+galaxy1'),\n",
       "  ('TB Variant Report',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbvcfreport/tbvcfreport/0.1.8+galaxy0'),\n",
       "  ('TB Variant Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tb_variant_filter/tb_variant_filter/0.1.3+galaxy0'),\n",
       "  ('TB Variant Report',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbvcfreport/tbvcfreport/0.1.7+galaxy0'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.12.0+galaxy1'),\n",
       "  ('Augustus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/augustus/augustus/3.4.0+galaxy0'),\n",
       "  ('Train Augustus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/augustus_training/augustus_training/3.4.0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.2.1+galaxy0'),\n",
       "  ('TB Variant Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tb_variant_filter/tb_variant_filter/0.3.5+galaxy1'),\n",
       "  ('hmmsearch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmsearch/hmmer_hmmsearch/3.3.2+galaxy0'),\n",
       "  ('hmmemit',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmemit/hmmer_hmmemit/3.3.2'),\n",
       "  ('hmmalign',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmalign/hmmer_hmmalign/3.3.2'),\n",
       "  ('nhmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmscan/hmmer_nhmmscan/3.3.2+galaxy0'),\n",
       "  ('jackhmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_jackhmmer/hmmer_jackhmmer/3.3.2+galaxy0'),\n",
       "  ('phmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_phmmer/hmmer_phmmer/3.3.2+galaxy0'),\n",
       "  ('hmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmscan/hmmer_hmmscan/3.3.2+galaxy0'),\n",
       "  ('hmmfetch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmfetch/hmmer_hmmfetch/3.3.2'),\n",
       "  ('alimask',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_alimask/hmmer_alimask/3.3.2'),\n",
       "  ('hmmconvert',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmconvert/hmmer_hmmconvert/3.3.2'),\n",
       "  ('hmmbuild',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmbuild/hmmer_hmmbuild/3.3.2+galaxy0'),\n",
       "  ('nhmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmer/hmmer_nhmmer/3.3.2+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.2.2+galaxy0'),\n",
       "  ('MOB-Recon',\n",
       "   'toolshed.g2.bx.psu.edu/repos/nml/mob_suite/mob_recon/3.0.3+galaxy0'),\n",
       "  ('MOB-Typer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/nml/mob_suite/mob_typer/3.0.3+galaxy0'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/3.0.6+galaxy0'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/3.0.7+galaxy0'),\n",
       "  ('lorikeet spoligotyping',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/lorikeet_spoligotype/lorikeet_spoligotype/20+galaxy0'),\n",
       "  ('SpoTyping',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spotyping/spotyping/2.1+galaxy2'),\n",
       "  ('RepeatModeler',\n",
       "   'toolshed.g2.bx.psu.edu/repos/csbl/repeatmodeler/repeatmodeler/2.0.2a+galaxy0'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/3.0.7+galaxy1'),\n",
       "  ('ChIPseeker',\n",
       "   'toolshed.g2.bx.psu.edu/repos/rnateam/chipseeker/chipseeker/1.28.3+galaxy0'),\n",
       "  ('Augustus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/augustus/augustus/3.4.0+galaxy1'),\n",
       "  ('Train Augustus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/augustus_training/augustus_training/3.4.0+galaxy1'),\n",
       "  ('RepeatMasker',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/repeat_masker/repeatmasker_wrapper/4.1.2-p1+galaxy0'),\n",
       "  ('TB Variant Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tb_variant_filter/tb_variant_filter/0.3.5+galaxy2'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/3.0.8+galaxy0'),\n",
       "  ('scanMotifGenomeWide',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_scanmotifgenomewide/homer_scanMotifGenomeWide/4.11+galaxy2'),\n",
       "  ('annotatePeaks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_annotatepeaks/homer_annotatePeaks/4.11+galaxy2'),\n",
       "  ('findMotifsGenome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_findmotifsgenome/homer_findMotifsGenome/4.11+galaxy2'),\n",
       "  ('Converts GTF to Annotations file for Homer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_gtf_to_annotations/homer_gtf_to_annotations/4.11+galaxy0'),\n",
       "  ('annotatePeaks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_annotatepeaks/homer_annotatePeaks/4.11+galaxy0'),\n",
       "  ('findMotifsGenome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_findmotifsgenome/homer_findMotifsGenome/4.11+galaxy0'),\n",
       "  ('hamronize',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hamronize_tool/hamronize_tool/1.0.3+galaxy0'),\n",
       "  ('summarize',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hamronize_summarize/hamronize_summarize/1.0.3+galaxy0'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.9+galaxy2'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.9+galaxy2'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.9+galaxy2'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.9+galaxy2'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.9+galaxy2'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.9+galaxy1'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.9+galaxy1'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.9+galaxy1'),\n",
       "  ('annotatePeaks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_annotatepeaks/homer_annotatePeaks/4.11+galaxy1'),\n",
       "  ('findMotifsGenome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_findmotifsgenome/homer_findMotifsGenome/4.11+galaxy1'),\n",
       "  ('hamronize',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hamronize_tool/hamronize_tool/1.0.3+galaxy1'),\n",
       "  ('hamronize: summarize',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hamronize_summarize/hamronize_summarize/1.0.3+galaxy1'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.2.2+galaxy2'),\n",
       "  ('Prokka', 'toolshed.g2.bx.psu.edu/repos/crs4/prokka/prokka/1.14.6+galaxy1'),\n",
       "  ('List spaln parameter tables',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/list_spaln_tables/2.4.6+galaxy0'),\n",
       "  ('Spaln: align cDNA or Protein to genome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/spaln/2.4.6+galaxy0'),\n",
       "  ('GECCO', 'toolshed.g2.bx.psu.edu/repos/althonos/gecco/gecco/0.8.5'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.52-86.0+galaxy1'),\n",
       "  ('TRANSIT Gumbel',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/transit_gumbel/transit_gumbel/3.0.2+galaxy1'),\n",
       "  ('TRANSIT Resampling',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/transit_resampling/transit_resampling/3.0.2+galaxy1'),\n",
       "  ('TRANSIT Tn5Gaps',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/transit_tn5gaps/transit_tn5gaps/3.0.2+galaxy1'),\n",
       "  ('TRANSIT HMM',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/transit_hmm/transit_hmm/3.0.2+galaxy1'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.52-86.0+galaxy0'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.54-87.0+galaxy1'),\n",
       "  ('List spaln parameter tables',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/list_spaln_tables/2.4.7+galaxy0'),\n",
       "  ('Spaln: align cDNA or Protein to genome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/spaln/2.4.7+galaxy0'),\n",
       "  ('Maker', 'toolshed.g2.bx.psu.edu/repos/iuc/maker/maker/2.31.11+galaxy2'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.54-87.0+galaxy2'),\n",
       "  ('KofamScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/kofamscan/kofamscan/1.3.0+galaxy1'),\n",
       "  ('legsta', 'toolshed.g2.bx.psu.edu/repos/iuc/legsta/legsta/0.5.1'),\n",
       "  ('legsta', 'toolshed.g2.bx.psu.edu/repos/iuc/legsta/legsta/0.5.1+galaxy1'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/4.1.1+galaxy0'),\n",
       "  ('TB Variant Report',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbvcfreport/tbvcfreport/0.1.10+galaxy0'),\n",
       "  ('Promoter 2.0',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/promoter2/0.0.13'),\n",
       "  ('psortb',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/Psortb/0.0.10'),\n",
       "  ('RXLR Motifs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/rxlr_motifs/0.0.17'),\n",
       "  ('SignalP 3.0',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/signalp3/0.0.20'),\n",
       "  ('TMHMM 2.0',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/tmhmm2/0.0.17'),\n",
       "  ('WoLF PSORT',\n",
       "   'toolshed.g2.bx.psu.edu/repos/peterjc/tmhmm_and_signalp/wolf_psort/0.0.12'),\n",
       "  ('RepeatModeler',\n",
       "   'toolshed.g2.bx.psu.edu/repos/csbl/repeatmodeler/repeatmodeler/2.0.3+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.3.1+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.3.2+galaxy0'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.55-88.0+galaxy3'),\n",
       "  ('List spaln parameter tables',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/list_spaln_tables/2.4.9+galaxy0'),\n",
       "  ('Spaln: align cDNA or Protein to genome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/spaln/spaln/2.4.9+galaxy0'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.0.6+galaxy2'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.14.0+galaxy0'),\n",
       "  ('Antismash',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/antismash/antismash/6.1.1+galaxy1'),\n",
       "  ('RepeatMasker',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/repeat_masker/repeatmasker_wrapper/4.1.2-p1+galaxy1'),\n",
       "  ('hamronize: summarize',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hamronize_summarize/hamronize_summarize/1.0.3+galaxy2'),\n",
       "  ('TEfinder', 'toolshed.g2.bx.psu.edu/repos/iuc/te_finder/te_finder/1.0.1'),\n",
       "  ('Gubbins',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/gubbins/gubbins/3.2.1+galaxy0'),\n",
       "  ('ISEScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/isescan/isescan/1.7.2.3+galaxy0'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.14.0+galaxy1'),\n",
       "  ('TEfinder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/te_finder/te_finder/1.0.1+galaxy1'),\n",
       "  ('Miniprot align',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot/miniprot/0.3+galaxy0'),\n",
       "  ('netboxr', 'toolshed.g2.bx.psu.edu/repos/bgruening/netboxr/netboxr/1.6.0'),\n",
       "  ('PlasmidFinder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/plasmidfinder/plasmidfinder/2.1.6+galaxy1'),\n",
       "  ('Miniprot index',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot_index/miniprot_index/0.3+galaxy0'),\n",
       "  ('Red', 'toolshed.g2.bx.psu.edu/repos/iuc/red/red/2018.09.10+galaxy1'),\n",
       "  ('Integron Finder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/integron_finder/integron_finder/2.0.2+galaxy0'),\n",
       "  ('edta', 'toolshed.g2.bx.psu.edu/repos/bgruening/edta/edta/2.1.0+galaxy0'),\n",
       "  ('InterProScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/interproscan/interproscan/5.59-91.0+galaxy3'),\n",
       "  ('TB Variant Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tb_variant_filter/tb_variant_filter/0.3.6+galaxy0'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/4.4.0+galaxy0'),\n",
       "  ('Bakta', 'toolshed.g2.bx.psu.edu/repos/iuc/bakta/bakta/1.6.1+galaxy0'),\n",
       "  ('goseq', 'toolshed.g2.bx.psu.edu/repos/iuc/goseq/goseq/1.50.0+galaxy0'),\n",
       "  ('Bakta', 'toolshed.g2.bx.psu.edu/repos/iuc/bakta/bakta/1.5.0+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.4.4+galaxy0'),\n",
       "  ('CPAT', 'toolshed.g2.bx.psu.edu/repos/bgruening/cpat/cpat/3.0.4+galaxy0'),\n",
       "  ('GffCompare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/gffcompare/gffcompare/0.12.6+galaxy0'),\n",
       "  ('PfamScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/pfamscan/pfamscan/1.6+galaxy0'),\n",
       "  ('GECCO', 'toolshed.g2.bx.psu.edu/repos/althonos/gecco/gecco/0.9.6'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.16.0+galaxy1'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.9+galaxy3'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.9+galaxy3'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.9+galaxy3'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.9+galaxy3'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.9+galaxy3'),\n",
       "  ('Integron Finder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/integron_finder/integron_finder/2.0.2+galaxy1'),\n",
       "  ('TB-Profiler Profile',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tbprofiler/tb_profiler_profile/4.4.1+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.4.5+galaxy0'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.1.3+galaxy0'),\n",
       "  ('bacteriophage annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pharokka/pharokka/1.2.1+galaxy0'),\n",
       "  ('seq2HLA', 'toolshed.g2.bx.psu.edu/repos/iuc/seq2hla/seq2hla/2.3+galaxy0'),\n",
       "  ('pharokka',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pharokka/pharokka/1.2.1+galaxy1'),\n",
       "  ('annotatePeaks',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_annotatepeaks/homer_annotatePeaks/4.11+galaxy3'),\n",
       "  ('findMotifs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_findmotifs/homer_findMotifs/4.11+galaxy3'),\n",
       "  ('findMotifsGenome',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_findmotifsgenome/homer_findMotifsGenome/4.11+galaxy3'),\n",
       "  ('scanMotifGenomeWide',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/homer_scanmotifgenomewide/homer_scanMotifGenomeWide/4.11+galaxy3'),\n",
       "  ('Miniprot align',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot/miniprot/0.10+galaxy0'),\n",
       "  ('Miniprot index',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot_index/miniprot_index/0.10+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.4.6+galaxy0'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy0'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.15+galaxy0'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.15+galaxy0'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.15+galaxy0'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.15+galaxy0'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy1'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.15+galaxy1'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.15+galaxy1'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.15+galaxy1'),\n",
       "  ('RepeatMasker',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/repeat_masker/repeatmasker_wrapper/4.1.5+galaxy0'),\n",
       "  ('RepeatModeler',\n",
       "   'toolshed.g2.bx.psu.edu/repos/csbl/repeatmodeler/repeatmodeler/2.0.4+galaxy0'),\n",
       "  ('RepeatModeler',\n",
       "   'toolshed.g2.bx.psu.edu/repos/csbl/repeatmodeler/repeatmodeler/2.0.4+galaxy1'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.15+galaxy1'),\n",
       "  ('MLST', 'toolshed.g2.bx.psu.edu/repos/iuc/mlst/mlst/2.22.0'),\n",
       "  ('MLST List', 'toolshed.g2.bx.psu.edu/repos/iuc/mlst/mlst_list/2.22.0'),\n",
       "  ('Braker genome annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker/braker/2.1.6+galaxy0'),\n",
       "  ('multiGSEA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/multigsea/multigsea/1.8.0+galaxy0'),\n",
       "  ('Miniprot align',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot/miniprot/0.12+galaxy0'),\n",
       "  ('Miniprot index',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/miniprot_index/miniprot_index/0.12+galaxy0'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.17.0+galaxy0'),\n",
       "  ('Bakta', 'toolshed.g2.bx.psu.edu/repos/iuc/bakta/bakta/1.8.1+galaxy1'),\n",
       "  ('pharokka',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/pharokka/pharokka/1.3.2+galaxy0'),\n",
       "  ('TB Variant Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/tb_variant_filter/tb_variant_filter/0.4.0+galaxy0'),\n",
       "  ('Get open reading frames (ORFs) or coding sequences (CDSs)',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_get_orfs/get_orfs_or_cdss/19.1.0.0'),\n",
       "  ('GFF3 Add Gene to CDS',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_gff_add_parents/edu.tamu.cpt.gff3.cdsParents/19.1.0.0'),\n",
       "  ('GFF3 Add Gene to CDS for Sixpack',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_fix_sixpack/edu.tamu.cpt.gff3.fixsixpack/19.1.0.0'),\n",
       "  ('Shine Find',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_shinefind/edu.tamu.cpt.genbank.shinefind/21.1.0.0'),\n",
       "  ('GFF3 Feature Type Filter',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_type_filter/edu.tamu.cpt.gff3.filter_type/19.1.0.0'),\n",
       "  ('GFF3 Filter: Require Phage Start',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_req_phage_start/edu.tamu.cpt.gff3.require_phage_start/19.1.0.0'),\n",
       "  ('Remove Annotation Feature',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_remove_annotations/edu.tamu.cpt.gff3.remove_annots/19.1.0.1'),\n",
       "  ('Prep GFF3 Input for Apollo',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_prep_for_apollo/edu.tamu.cpt.gff3.prepForApollo/20.8.0.0'),\n",
       "  ('Glimmer3 to GFF3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_convert_glimmer/edu.tamu.cpt2.util.glimmer3_to_gff3/19.1.0.0'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.17.0+galaxy1'),\n",
       "  ('AGAT', 'toolshed.g2.bx.psu.edu/repos/bgruening/agat/agat/1.2.0+galaxy0'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.3+galaxy0'),\n",
       "  ('Bakta', 'toolshed.g2.bx.psu.edu/repos/iuc/bakta/bakta/1.8.2+galaxy0'),\n",
       "  ('Helixer', 'toolshed.g2.bx.psu.edu/repos/genouest/helixer/helixer/0.3.2'),\n",
       "  ('Interrupted gene detection tool',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_intron_detection/edu.tamu.cpt2.phage.intron_detection/19.1.0.0'),\n",
       "  ('GFF3 Feature Sequence Export',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_gff_extract_seq/edu.tamu.cpt.gff3.export_seq/19.1.0.0'),\n",
       "  ('Identify Lipoboxes',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_lipory/edu.tamu.cpt.fasta.lipory/19.1.0.0'),\n",
       "  ('Renumber GenBank Genes',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_gbk_renumber/edu.tamu.cpt.genbank.RelabelTags/0.4'),\n",
       "  ('Genome Editor',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_genome_editor/edu.tamu.cpt.gff3.genome_editor/2.1'),\n",
       "  ('Phage QC',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_phageqc_annotations/edu.tamu.cpt2.phage.annotation_validator/1.9.0'),\n",
       "  ('Annotation Table',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_annotation_table/edu.tamu.cpt2.phage.annotation_table/20.0.0.1'),\n",
       "  ('Rebase GFF3 features',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_gff_rebase/gff3.rebase/19.1.0.0'),\n",
       "  ('Gff3 Filter: Require SD',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_require_shine/edu.tamu.cpt.gff3.require_sd/19.1.0.0'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.3+galaxy1'),\n",
       "  ('AMRFinderPlus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/amrfinderplus/amrfinderplus/3.11.26+galaxy0'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy2'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.15+galaxy2'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.15+galaxy2'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.15+galaxy2'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.15+galaxy2'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.3+galaxy2'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.15+galaxy3'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy3'),\n",
       "  ('RepeatExplorer (clustering)',\n",
       "   'toolshed.g2.bx.psu.edu/repos/gga/repeatexplorer_clustering/repeatexplorer_clustering/2.3.8+galaxy0'),\n",
       "  ('AMRFinderPlus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/amrfinderplus/amrfinderplus/3.11.26+galaxy1'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.15+galaxy3'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.15+galaxy3'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.15+galaxy3'),\n",
       "  ('DeepSig',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/deepsig/deepsig/1.2.5+galaxy0'),\n",
       "  ('Build a GeneNoteBook',\n",
       "   'toolshed.g2.bx.psu.edu/repos/gga/genenotebook_genenotebook_build/genenotebook_build/0.4.8+galaxy0'),\n",
       "  ('TSEBRA', 'toolshed.g2.bx.psu.edu/repos/iuc/tsebra/tsebra/1.1.2+galaxy0'),\n",
       "  ('hmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmscan/hmmer_hmmscan/3.3.2+galaxy1'),\n",
       "  ('hmmsearch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmsearch/hmmer_hmmsearch/3.3.2+galaxy1'),\n",
       "  ('nhmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmer/hmmer_nhmmer/3.3.2+galaxy1'),\n",
       "  ('nhmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmscan/hmmer_nhmmscan/3.3.2+galaxy1'),\n",
       "  ('phmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_phmmer/hmmer_phmmer/3.3.2+galaxy1'),\n",
       "  ('RepeatModeler',\n",
       "   'toolshed.g2.bx.psu.edu/repos/csbl/repeatmodeler/repeatmodeler/2.0.5+galaxy0'),\n",
       "  ('Busco', 'toolshed.g2.bx.psu.edu/repos/iuc/busco/busco/5.5.0+galaxy0'),\n",
       "  ('Build a GeneNoteBook',\n",
       "   'toolshed.g2.bx.psu.edu/repos/gga/genenotebook_genenotebook_build/genenotebook_build/0.4.9+galaxy0'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.6+galaxy2'),\n",
       "  ('WindowMasker mkcounts',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/windowmasker/windowmasker_mkcounts/1.0'),\n",
       "  ('WindowMasker ustat',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/windowmasker/windowmasker_ustat/1.0'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy4'),\n",
       "  ('Funannotate assembly clean',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_clean/funannotate_clean/1.8.15+galaxy4'),\n",
       "  ('Funannotate compare',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_compare/funannotate_compare/1.8.15+galaxy4'),\n",
       "  ('Funannotate predict annotation',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_predict/funannotate_predict/1.8.15+galaxy4'),\n",
       "  ('Sort assembly',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_sort/funannotate_sort/1.8.15+galaxy4'),\n",
       "  ('MITOS', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos/mitos/1.1.4+galaxy0'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.1.6+galaxy0'),\n",
       "  ('OrthoFinder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/orthofinder_onlygroups/orthofinder_onlygroups/2.5.5+galaxy0'),\n",
       "  ('ISEScan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/isescan/isescan/1.7.2.3+galaxy1'),\n",
       "  ('MITOS', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos/mitos/1.1.5+galaxy0'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.1.7+galaxy0'),\n",
       "  ('CPAT', 'toolshed.g2.bx.psu.edu/repos/bgruening/cpat/cpat/3.0.5+galaxy0'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.7+galaxy2'),\n",
       "  ('Build a GeneNoteBook',\n",
       "   'toolshed.g2.bx.psu.edu/repos/gga/genenotebook_genenotebook_build/genenotebook_build/0.4.12+galaxy0'),\n",
       "  ('MITOS2', 'toolshed.g2.bx.psu.edu/repos/iuc/mitos2/mitos2/2.1.7+galaxy1'),\n",
       "  ('BRAKER3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/genouest/braker3/braker3/3.0.7+galaxy3'),\n",
       "  ('TAPScan Classify',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/tapscan/tapscan_classify/4.74+galaxy0'),\n",
       "  ('AGAT', 'toolshed.g2.bx.psu.edu/repos/bgruening/agat/agat/1.2.0+galaxy1'),\n",
       "  ('Bakta', 'toolshed.g2.bx.psu.edu/repos/iuc/bakta/bakta/1.9.2+galaxy0'),\n",
       "  ('Funannotate functional',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/funannotate_annotate/funannotate_annotate/1.8.15+galaxy5'),\n",
       "  ('AEGeAn CanonGFF3',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/aegean_canongff3/aegean_canongff3/0.16.0+galaxy1'),\n",
       "  ('AEGeAn LocusPocus',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/aegean_locuspocus/aegean_locuspocus/0.16.0+galaxy1'),\n",
       "  ('AEGeAn ParsEval',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/aegean_parseval/aegean_parseval/0.16.0+galaxy1'),\n",
       "  ('AEGeAn GAEVAL',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/aegean_gaeval/aegean_gaeval/0.16.0+galaxy1'),\n",
       "  ('Fasta Translate',\n",
       "   'toolshed.g2.bx.psu.edu/repos/cpt/cpt_fasta_translate/edu.tamu.cpt.fasta.seq_trans/19.1.0.0'),\n",
       "  ('TAPScan Classify',\n",
       "   'toolshed.g2.bx.psu.edu/repos/bgruening/tapscan/tapscan_classify/4.76+galaxy0'),\n",
       "  ('annotateMyIDs',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/annotatemyids/annotatemyids/3.18.0+galaxy0'),\n",
       "  ('TSEBRA', 'toolshed.g2.bx.psu.edu/repos/iuc/tsebra/tsebra/1.1.2.3+galaxy0'),\n",
       "  ('multiGSEA',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/multigsea/multigsea/1.12.0+galaxy0'),\n",
       "  ('alimask',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_alimask/hmmer_alimask/3.4+galaxy0'),\n",
       "  ('hmmalign',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmalign/hmmer_hmmalign/3.4+galaxy0'),\n",
       "  ('hmmbuild',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmbuild/hmmer_hmmbuild/3.4+galaxy0'),\n",
       "  ('hmmconvert',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmconvert/hmmer_hmmconvert/3.4+galaxy0'),\n",
       "  ('hmmemit',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmemit/hmmer_hmmemit/3.4+galaxy0'),\n",
       "  ('hmmfetch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmfetch/hmmer_hmmfetch/3.4+galaxy0'),\n",
       "  ('hmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmscan/hmmer_hmmscan/3.4+galaxy0'),\n",
       "  ('hmmsearch',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_hmmsearch/hmmer_hmmsearch/3.4+galaxy0'),\n",
       "  ('jackhmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_jackhmmer/hmmer_jackhmmer/3.4+galaxy0'),\n",
       "  ('nhmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmer/hmmer_nhmmer/3.4+galaxy0'),\n",
       "  ('nhmmscan',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_nhmmscan/hmmer_nhmmscan/3.4+galaxy0'),\n",
       "  ('phmmer',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/hmmer_phmmer/hmmer_phmmer/3.4+galaxy0'),\n",
       "  ('Integron Finder',\n",
       "   'toolshed.g2.bx.psu.edu/repos/iuc/integron_finder/integron_finder/2.0.3+galaxy0'),\n",
       "  ('OMArk', 'toolshed.g2.bx.psu.edu/repos/iuc/omark/omark/0.3.0+galaxy0'),\n",
       "  ...])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## tools\n",
    "tools = exp.gi.tools.get_tools()\n",
    "len(tools), tools[0].keys(), [(k['name'], k['id']) for k in tools]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'model_class': 'Tool',\n",
       " 'id': 'cat1',\n",
       " 'name': 'Concatenate datasets',\n",
       " 'version': '1.0.0',\n",
       " 'description': 'tail-to-head',\n",
       " 'labels': [],\n",
       " 'edam_operations': ['operation_3436'],\n",
       " 'edam_topics': [],\n",
       " 'hidden': '',\n",
       " 'is_workflow_compatible': True,\n",
       " 'xrefs': [],\n",
       " 'panel_section_id': 'text_manipulation',\n",
       " 'panel_section_name': 'Text Manipulation',\n",
       " 'form_style': 'regular'}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t1 = exp.gi.tools.show_tool('cat1')\n",
    "t1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This is quite unrelated\n",
    "## Downloads of fastq files from ena are under heavy construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(f\"https://www.ebi.ac.uk/ena/portal/api/results?dataPortal=ena\")\n",
    "content = r.content.decode('utf-8')\n",
    "print(content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(f\"https://www.ebi.ac.uk/ena/portal/api/searchFields?result=read_run\")\n",
    "content = r.content.decode('utf-8')\n",
    "print(content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this actually works but it is not waht I want\n",
    "search = \"ERS1153743\"\n",
    "r = requests.get(f\"https://www.ebi.ac.uk/ena/portal/api/search?result=read_run&{search}?dataPortal=ena\")\n",
    "content = r.content.decode('utf-8')\n",
    "print(content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = requests.get(f\"https://www.ebi.ac.uk/ena/browser/api/fasta/{search}\")\n",
    "content = r.content.decode('utf-8')\n",
    "print(content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading local FASTQ files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momicsdem",
   "language": "python",
   "name": "momicsdem"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
