{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# App for running GECCO jobs in Galaxy\n",
    "\n",
    "1. Upload and run workflow.\n",
    "2. Monitor the job.\n",
    "3. Receive completion notification with some basic summary provided by Galaxy.\n",
    "\n",
    "Note: \n",
    "- \"Receiving\" the results (tentatively download) is part of the analysis pipeline.\n",
    "- included `.gb`, `.gbk` and `.embl` formats to be allowed for gecco\n",
    "  - `.embl` from ena comes as `.txt` upon download actually.\n",
    "  - ncbi genbank download full or normal always cames as `.gb`.\n",
    "  \n",
    "**IMPORTANT**\n",
    "You need to have a `.env` file looking like this\n",
    "```\n",
    "GALAXY_URL=\"https://earth-system.usegalaxy.eu/\"\n",
    "GALAXY_KEY=\"...\"\n",
    "```\n",
    "For `binder` users, you cannot upload the `.env` so you will need to hardcode the `API token`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODOs\n",
    "\n",
    "- save json, from the analysis dashboard, update the json (this will be heavily environmental dependent)\n",
    "  - note for which environments it works (like GColab env will get deleted etc).\n",
    "  - So far tested `linux local`\n",
    "    - to test: `win local`, `gcolab`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Binder\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Cloning into 'marine-omics'...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Repository marine-omics cloned\n",
      "Collecting git+https://github.com/palec87/marine-omics.git\n",
      "  Cloning https://github.com/palec87/marine-omics.git to /tmp/pip-req-build-c0m9qw_h\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  Running command git clone --filter=blob:none --quiet https://github.com/palec87/marine-omics.git /tmp/pip-req-build-c0m9qw_h\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Resolved https://github.com/palec87/marine-omics.git to commit 5cd67a5082c4336018cd1e8aac1fe2ca45077d71\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: bioblend>=1.4.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (1.4.0)\n",
      "Requirement already satisfied: fastparquet>=2024.11.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (2024.11.0)\n",
      "Requirement already satisfied: matplotlib>=3.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (3.10.0)\n",
      "Requirement already satisfied: numpy>=1.23.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (2.2.2)\n",
      "Requirement already satisfied: pandas>=1.2.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (2.2.3)\n",
      "Requirement already satisfied: panel>=1.6.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (1.6.1)\n",
      "Requirement already satisfied: plotly>=5.3.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (6.0.0)\n",
      "Requirement already satisfied: psutil>=5.8.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (6.1.1)\n",
      "Requirement already satisfied: python-dotenv>=1.0.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (1.0.1)\n",
      "Requirement already satisfied: requests>=2.26.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (2.32.3)\n",
      "Requirement already satisfied: scikit-bio>=0.6.3 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (0.6.3)\n",
      "Requirement already satisfied: seaborn>=0.11.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from momics==0.0.1) (0.13.2)\n",
      "Requirement already satisfied: requests-toolbelt!=0.9.0,>=0.5.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bioblend>=1.4.0->momics==0.0.1) (1.0.0)\n",
      "Requirement already satisfied: tuspy in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bioblend>=1.4.0->momics==0.0.1) (1.1.0)\n",
      "Requirement already satisfied: cramjam>=2.3 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from fastparquet>=2024.11.0->momics==0.0.1) (2.9.1)\n",
      "Requirement already satisfied: fsspec in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from fastparquet>=2024.11.0->momics==0.0.1) (2025.2.0)\n",
      "Requirement already satisfied: packaging in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from fastparquet>=2024.11.0->momics==0.0.1) (24.2)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (4.56.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (1.4.8)\n",
      "Requirement already satisfied: pillow>=8 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (11.1.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (3.2.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from matplotlib>=3.5->momics==0.0.1) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from pandas>=1.2.0->momics==0.0.1) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from pandas>=1.2.0->momics==0.0.1) (2025.1)\n",
      "Requirement already satisfied: bleach in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (6.2.0)\n",
      "Requirement already satisfied: bokeh<3.7.0,>=3.5.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (3.6.3)\n",
      "Requirement already satisfied: linkify-it-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (2.0.3)\n",
      "Requirement already satisfied: markdown in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (3.7)\n",
      "Requirement already satisfied: markdown-it-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (3.0.0)\n",
      "Requirement already satisfied: mdit-py-plugins in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (0.4.2)\n",
      "Requirement already satisfied: param<3.0,>=2.1.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (2.2.0)\n",
      "Requirement already satisfied: pyviz-comms>=2.0.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (3.0.4)\n",
      "Requirement already satisfied: tqdm in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (4.67.1)\n",
      "Requirement already satisfied: typing-extensions in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel>=1.6.0->momics==0.0.1) (4.12.2)\n",
      "Requirement already satisfied: narwhals>=1.15.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from plotly>=5.3.0->momics==0.0.1) (1.26.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests>=2.26.0->momics==0.0.1) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests>=2.26.0->momics==0.0.1) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests>=2.26.0->momics==0.0.1) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests>=2.26.0->momics==0.0.1) (2025.1.31)\n",
      "Requirement already satisfied: decorator>=3.4.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (5.1.1)\n",
      "Requirement already satisfied: natsort>=4.0.3 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (8.4.0)\n",
      "Requirement already satisfied: scipy>=1.9.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (1.15.1)\n",
      "Requirement already satisfied: h5py>=3.6.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (3.12.1)\n",
      "Requirement already satisfied: biom-format>=2.1.16 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (2.1.16)\n",
      "Requirement already satisfied: statsmodels>=0.14.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (0.14.4)\n",
      "Requirement already satisfied: patsy>=0.5.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from scikit-bio>=0.6.3->momics==0.0.1) (1.0.1)\n",
      "Requirement already satisfied: click in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from biom-format>=2.1.16->scikit-bio>=0.6.3->momics==0.0.1) (8.1.8)\n",
      "Requirement already satisfied: Jinja2>=2.9 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel>=1.6.0->momics==0.0.1) (3.1.5)\n",
      "Requirement already satisfied: PyYAML>=3.10 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel>=1.6.0->momics==0.0.1) (6.0.2)\n",
      "Requirement already satisfied: tornado>=6.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel>=1.6.0->momics==0.0.1) (6.4.2)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel>=1.6.0->momics==0.0.1) (2025.1.0)\n",
      "Requirement already satisfied: six>=1.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib>=3.5->momics==0.0.1) (1.17.0)\n",
      "Requirement already satisfied: webencodings in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bleach->panel>=1.6.0->momics==0.0.1) (0.5.1)\n",
      "Requirement already satisfied: uc-micro-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from linkify-it-py->panel>=1.6.0->momics==0.0.1) (1.0.3)\n",
      "Requirement already satisfied: mdurl~=0.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from markdown-it-py->panel>=1.6.0->momics==0.0.1) (0.1.2)\n",
      "Requirement already satisfied: tinydb>=3.5.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from tuspy->bioblend>=1.4.0->momics==0.0.1) (4.8.2)\n",
      "Requirement already satisfied: aiohttp>=3.6.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from tuspy->bioblend>=1.4.0->momics==0.0.1) (3.11.12)\n",
      "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (2.4.6)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (1.3.2)\n",
      "Requirement already satisfied: async-timeout<6.0,>=4.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (5.0.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (25.1.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (1.5.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (6.1.0)\n",
      "Requirement already satisfied: propcache>=0.2.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (0.2.1)\n",
      "Requirement already satisfied: yarl<2.0,>=1.17.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from aiohttp>=3.6.2->tuspy->bioblend>=1.4.0->momics==0.0.1) (1.18.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from Jinja2>=2.9->bokeh<3.7.0,>=3.5.0->panel>=1.6.0->momics==0.0.1) (3.0.2)\n",
      "Building wheels for collected packages: momics\n",
      "  Building wheel for momics (pyproject.toml): started\n",
      "  Building wheel for momics (pyproject.toml): finished with status 'done'\n",
      "  Created wheel for momics: filename=momics-0.0.1-py3-none-any.whl size=27408 sha256=40227ff92ff320a664f7e54731c6aa850cbbf97a624b98af37d849aeaa440bdc\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-sax2txdo/wheels/6a/2d/d9/3abe84bc0a11258798bbdfbbf5df57841c73165fd80e3fc016\n",
      "Successfully built momics\n",
      "Installing collected packages: momics\n",
      "Successfully installed momics-0.0.1\n",
      "momics installed\n",
      "Requirement already satisfied: panel in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (1.6.1)\n",
      "Collecting hvplot\n",
      "  Downloading hvplot-0.11.2-py3-none-any.whl.metadata (15 kB)\n",
      "Requirement already satisfied: bleach in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (6.2.0)\n",
      "Requirement already satisfied: bokeh<3.7.0,>=3.5.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (3.6.3)\n",
      "Requirement already satisfied: linkify-it-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (2.0.3)\n",
      "Requirement already satisfied: markdown in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (3.7)\n",
      "Requirement already satisfied: markdown-it-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (3.0.0)\n",
      "Requirement already satisfied: mdit-py-plugins in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (0.4.2)\n",
      "Requirement already satisfied: packaging in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (24.2)\n",
      "Requirement already satisfied: pandas>=1.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (2.2.3)\n",
      "Requirement already satisfied: param<3.0,>=2.1.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (2.2.0)\n",
      "Requirement already satisfied: pyviz-comms>=2.0.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (3.0.4)\n",
      "Requirement already satisfied: requests in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (2.32.3)\n",
      "Requirement already satisfied: tqdm in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (4.67.1)\n",
      "Requirement already satisfied: typing-extensions in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from panel) (4.12.2)\n",
      "Collecting colorcet>=2 (from hvplot)\n",
      "  Downloading colorcet-3.1.0-py3-none-any.whl.metadata (6.3 kB)\n",
      "Collecting holoviews>=1.19.0 (from hvplot)\n",
      "  Downloading holoviews-1.20.1-py3-none-any.whl.metadata (9.9 kB)\n",
      "Requirement already satisfied: numpy>=1.21 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from hvplot) (2.2.2)\n",
      "Requirement already satisfied: Jinja2>=2.9 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (3.1.5)\n",
      "Requirement already satisfied: contourpy>=1.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (1.3.1)\n",
      "Requirement already satisfied: pillow>=7.1.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (11.1.0)\n",
      "Requirement already satisfied: PyYAML>=3.10 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (6.0.2)\n",
      "Requirement already satisfied: tornado>=6.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (6.4.2)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bokeh<3.7.0,>=3.5.0->panel) (2025.1.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from pandas>=1.2->panel) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from pandas>=1.2->panel) (2025.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from pandas>=1.2->panel) (2025.1)\n",
      "Requirement already satisfied: webencodings in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from bleach->panel) (0.5.1)\n",
      "Requirement already satisfied: uc-micro-py in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from linkify-it-py->panel) (1.0.3)\n",
      "Requirement already satisfied: mdurl~=0.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from markdown-it-py->panel) (0.1.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests->panel) (3.4.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests->panel) (3.10)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests->panel) (2.3.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from requests->panel) (2025.1.31)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from Jinja2>=2.9->bokeh<3.7.0,>=3.5.0->panel) (3.0.2)\n",
      "Requirement already satisfied: six>=1.5 in /media/davidp/Data/coding/marine_omics/momics-demos/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas>=1.2->panel) (1.17.0)\n",
      "Downloading hvplot-0.11.2-py3-none-any.whl (161 kB)\n",
      "Downloading colorcet-3.1.0-py3-none-any.whl (260 kB)\n",
      "Downloading holoviews-1.20.1-py3-none-any.whl (5.0 MB)\n",
      "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m5.0/5.0 MB\u001b[0m \u001b[31m28.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: colorcet, holoviews, hvplot\n",
      "Successfully installed colorcet-3.1.0 holoviews-1.20.1 hvplot-0.11.2\n",
      "panel and hvplot installed\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import logging\n",
    "from IPython import get_ipython\n",
    "\n",
    "logger = logging.getLogger(name=\"GECCO galaxy runner\")\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    # clone the momics-demos repository to use the utils module from there\n",
    "    # TODO: eventually utils from momics will be used for that\n",
    "    try:\n",
    "        os.system('git clone https://github.com/palec87/momics-demos.git')\n",
    "        logger.info(f\"Repository cloned\")\n",
    "    except OSError as e:\n",
    "        logger.info(f\"An error occurred while cloning the repository: {e}\")\n",
    "\n",
    "    sys.path.insert(0,'/content/momics-demos')\n",
    "elif \"zmqshell\" in str(get_ipython()):\n",
    "    logger.info(\"Binder\")\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "\n",
    "    logger.info('Binder will not allow you to upload the \".env\" file')\n",
    "    os.environ[\"GALAXY_EARTH_URL\"] = \"https://earth-system.usegalaxy.eu/\"\n",
    "    #########################\n",
    "    ### INPUT TOKEN HERE ####\n",
    "    #########################\n",
    "    os.environ[\"GALAXY_EARTH_KEY\"] = 0\n",
    "    assert isinstance(os.environ[\"GALAXY_EARTH_KEY\"], str) is True, \"token must be a string\"\n",
    "\n",
    "else:\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))  # local utils, to be removed in the future\n",
    "\n",
    "    # downside of this is that all the deps need to be installed in the current (momics-demos) environment\n",
    "    sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '../../../marine-omics')))  # local momics package, to be removed too\n",
    "\n",
    "from utils import init_setup, get_notebook_environment\n",
    "init_setup()\n",
    "\n",
    "# Determine the notebook environment\n",
    "env = get_notebook_environment()\n",
    "logger.info(f\"Environment: {env}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This needs to be repeated here for the Pannel dashboard to work, WEIRD\n",
    "# TODO: report as possible bug\n",
    "import sys\n",
    "import os\n",
    "\n",
    "import bioblend.galaxy as g  # BioBlend is a Python library, wrapping the functionality of Galaxy and CloudMan APIs\n",
    "import panel as pn\n",
    "\n",
    "sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))\n",
    "from utils import init_setup\n",
    "init_setup()\n",
    "\n",
    "from bioblend.galaxy.datasets import DatasetClient\n",
    "\n",
    "from momics.galaxy import RemGalaxy, Gecco\n",
    "from momics.panel_utils import (\n",
    "    serve_app, close_server,\n",
    ")\n",
    "from momics.utils import memory_load, reconfig_logger\n",
    "\n",
    "# instead of the jupyter magic, you can also use\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Session setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEBUG = True  # enable stdout logging\n",
    "reconfig_logger()  # Set up logging\n",
    "\n",
    "# these variables need to be set in the .env file at the root of the project\n",
    "exp = RemGalaxy(\"GALAXY_EARTH_URL\", \"GALAXY_EARTH_KEY\")\n",
    "gecco_tool_id = \"toolshed.g2.bx.psu.edu/repos/althonos/gecco/gecco/0.9.6\"  # The id of the tool GECCO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# environmentally dependent paths.\n",
    "# TODO: why is this not in the momics package already?\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    root_folder = os.path.abspath(os.path.join('/content/momics-demos'))\n",
    "else:\n",
    "    root_folder = os.path.abspath(os.path.join('../'))\n",
    "\n",
    "assets_folder = os.path.join(root_folder, 'assets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## App setup\n",
    "\n",
    "TODO: do I really need to duplicate the `current_history`, `current_file` name and id? I could handle the tuples in the \"backend\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# buttons\n",
    "but_login = pn.widgets.Button(name=\"üîê Galaxy Login\")\n",
    "but_get_histories = pn.widgets.Button(name=\"üìö Refresh Histories\")\n",
    "but_create_history = pn.widgets.Button(name=\"üìù Create New History\")\n",
    "but_get_datasets = pn.widgets.Button(name=\"üìñ Refresh Datasets\")\n",
    "\n",
    "# input file handling\n",
    "but_upload_dataset = pn.widgets.Button(name=\"üì§ Upload Dataset to Galaxy\")\n",
    "\n",
    "# r'(*.fasta|*.gbk|*.embl|*.txt)'), this is a PR i suggested https://github.com/holoviz/panel/issues/7726\n",
    "file_input = pn.widgets.FileSelector('~', file_pattern=\"*\")\n",
    "file_source_checkbox = pn.widgets.Checkbox(name='Use a file from the Galaxy', value=True)\n",
    "\n",
    "but_submit = pn.widgets.Button(name=\"üöÄ Submit GECCO task\")\n",
    "\n",
    "# this perhaps can get solved with file selector\n",
    "history_name = pn.widgets.TextInput(name='New History Name', placeholder='Enter a string here...')\n",
    "\n",
    "current_history_name = pn.widgets.StaticText(name='Current History Name', value='No history selected')\n",
    "current_history_id = pn.widgets.StaticText(name='Current History ID', value='')\n",
    "\n",
    "current_file_name = pn.widgets.StaticText(name='Current filename for GECCO', value='No file selected')\n",
    "current_file_id = pn.widgets.StaticText(name='Current file ID for GECCO', value='')\n",
    "\n",
    "# selectors\n",
    "select_history = pn.widgets.Select(\n",
    "    name=\"Select History\",\n",
    "    options=[],\n",
    "    description=\"Your Galaxy histories, create a new one if needed\",\n",
    ")\n",
    "select_dataset = pn.widgets.Select(\n",
    "    name=\"Select Dataset\",\n",
    "    options=[(\"\", \"\")],\n",
    "    description=\"Your Galaxy fasta datasets\",\n",
    "    value=(\"\", \"\"),\n",
    ")\n",
    "\n",
    "# intro text\n",
    "intro_text = pn.pane.Markdown(\n",
    "    \"\"\"\n",
    "    # GECCO Galaxy Runner\n",
    "\n",
    "    GECCO (Gene Cluster prediction with Conditional Random Fields) is a fast and scalable method for identifying putative novel Biosynthetic Gene Clusters (BGCs) in genomic and metagenomic data using Conditional Random Fields (CRFs). It is developed in the Zeller group and is part of the suite of computational microbiome analysis tools hosted at EMBL.\n",
    "\n",
    "    ## Instructions\n",
    "\n",
    "    1. Log in to your Galaxy instance.\n",
    "    2. Select a history or create a new one.\n",
    "    3. Select or upload a fasta/GenBank/EMBL file to the selected history.\n",
    "    4. Submit the GECCO task.\n",
    "    5. Continue with the post-processing of the results in the analysis NB.\n",
    "\n",
    "    ## Output\n",
    "\n",
    "    GECCO will create the following files once done (using the same prefix as the input file):\n",
    "\n",
    "    - features.tsv: The genes file, containing the genes identified in the input sequences.\n",
    "    - features.tsv: The features file, containing the protein domains identified in the input sequences.\n",
    "    - clusters.tsv: A clusters file, containing the coordinates of the predicted clusters, along their putative biosynthetic type.\n",
    "    - {sequence}_cluster_{N}.gbk: If any BGCs were found, a GenBank file per cluster, containing the cluster sequence annotated with its member proteins and domains.\n",
    "\n",
    "    ## References\n",
    "    [link](https://doi.org/10.1101/2021.05.03.442509) Carroll, L. M., Larralde, M., Fleck, J. S., Ponnudurai, R., Milanese, A., Cappio, E., & Zeller, G. (2021). Accurate de novo identification of biosynthetic gene clusters with GECCO.\n",
    "    [GECCO tool page](https://usegalaxy.eu/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Falthonos%2Fgecco%2Fgecco%2F0.9.6)\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "# gecco params\n",
    "mask = pn.widgets.Checkbox(name='Enable masking of regions with unknown nucleotides', value=False)\n",
    "pad = pn.widgets.Checkbox(name='Enable padding of gene sequences smaller than the CRF window length', value=True)\n",
    "cds = pn.widgets.IntInput(name='IntInMinimum number of genes required for a clusterput',\n",
    "                          value=3, step=1, start=2, end=1000,\n",
    "                          )\n",
    "threshold = pn.widgets.FloatInput(name='Probability threshold for cluster detection',\n",
    "                                  value=0.05, step=0.01, start=0.0, end=1.0,\n",
    "                                  )\n",
    "postproc = pn.widgets.Select(\n",
    "    name=\"Post-processing method for gene cluster validation\",\n",
    "    options=[\"gecco\", \"antiSMASH\"],\n",
    ")\n",
    "gene_filter = pn.widgets.IntInput(\n",
    "    name='Number of genes from the contig edges to filter out',\n",
    "    value=0, step=1, start=0, end=100)\n",
    "\n",
    "antimash_sideload = pn.widgets.Checkbox(name='Generate an antiSMASH v6 sideload JSON file', value=False)\n",
    "\n",
    "# this is currently not Implemented for Galaxy tools\n",
    "# monitor here, https://github.com/galaxyproject/galaxy/issues/1364\n",
    "# email_input = pn.widgets.TextInput(name='Email notification', placeholder='Enter a string here...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gecco_params = {\n",
    "    \"select_history\": select_history,\n",
    "    \"current_history_name\": current_history_name,\n",
    "    \"current_history_id\": current_history_id,\n",
    "    \"select_dataset\": select_dataset,\n",
    "    \"current_file_name\": current_file_name,\n",
    "    \"current_file_id\": current_file_id,\n",
    "    \"file_source_checkbox\": file_source_checkbox,\n",
    "    \"file_input\": file_input,\n",
    "    \"mask\": mask,\n",
    "    \"pad\": pad,\n",
    "    \"cds\": cds,\n",
    "    \"threshold\": threshold,\n",
    "    \"postproc\": postproc,\n",
    "    \"gene_filter\": gene_filter,\n",
    "    \"antimash_sideload\": antimash_sideload,\n",
    "    \"history_name\": history_name,\n",
    "}\n",
    "\n",
    "gecco = Gecco(gecco_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@pn.depends(file_source=file_source_checkbox, watch=True)\n",
    "def _generate_data_input_flexbox(file_source):\n",
    "    if file_source:\n",
    "        return pn.Column(but_get_datasets, select_dataset)\n",
    "    else:\n",
    "        return pn.Column(\n",
    "            \"GECCO accepts `fasta`, GenBank (`.gbk`, `gb`) and EMBL ('.embl'). In principle all of those could be `.txt` files too.\",\n",
    "            file_input,\n",
    "            but_upload_dataset,\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension(\"tabulator\")\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    pn.extension(comms='colab')\n",
    "pn.extension(notifications=True)\n",
    "ACCENT = \"teal\"\n",
    "\n",
    "styles = {\n",
    "    \"box-shadow\": \"rgba(50, 50, 93, 0.25) 0px 6px 12px -2px, rgba(0, 0, 0, 0.3) 0px 3px 7px -3px\",\n",
    "    \"border-radius\": \"4px\",\n",
    "    \"padding\": \"10px\",\n",
    "}\n",
    "\n",
    "image = pn.pane.JPG(os.path.join(assets_folder, \"figs/logo_gecco.jpeg\"),\n",
    "                    width=200,\n",
    "                    )\n",
    "\n",
    "def app():\n",
    "    history_flexbox = pn.Column(\n",
    "        pn.Row(but_get_histories, but_create_history),\n",
    "        history_name,\n",
    "        select_history,\n",
    "        pn.bind(gecco.handle_get_histories, clicks=but_get_histories.param.clicks),\n",
    "        pn.bind(gecco.handle_create_history, clicks=but_create_history.param.clicks),\n",
    "        sizing_mode=\"stretch_width\",\n",
    "    )\n",
    "\n",
    "    gecco_flexbox = pn.FlexBox(\n",
    "        pn.Column(\n",
    "            mask, pad, cds, threshold, postproc, gene_filter, antimash_sideload,\n",
    "            pn.layout.Divider(),\n",
    "            \"## Submit\", current_file_name, current_history_name, but_submit,\n",
    "        ),\n",
    "        sizing_mode=\"stretch_both\",\n",
    "    )\n",
    "\n",
    "    tabs = pn.Tabs(\n",
    "        (\"Introduction\", intro_text),\n",
    "        ('File Upload', _generate_data_input_flexbox),\n",
    "        ('GECCO Parameters', gecco_flexbox),\n",
    "        styles=styles, sizing_mode=\"stretch_both\",\n",
    "    )\n",
    "\n",
    "    pn.bind(gecco.handle_update_current_file_name, select_dataset, watch=True)\n",
    "    pn.bind(gecco.handle_update_current_history_name, select_history, watch=True)\n",
    "    pn.bind(gecco.handle_update_current_history_id, select_history, watch=True)\n",
    "    pn.bind(gecco.handle_submit_gecco, but_submit, watch=True)\n",
    "    template = pn.template.FastListTemplate(\n",
    "        title=\"Run GECCO on Galaxy\",\n",
    "        sidebar=[image,\n",
    "                but_login, pn.bind(gecco.handle_login, clicks=but_login.param.clicks),\n",
    "                pn.layout.Divider(margin=(-20, 0, 0, 0)),\n",
    "                \"## Histories\", history_flexbox,\n",
    "                \"## Datasets\", file_source_checkbox,\n",
    "                pn.bind(gecco.handle_get_datasets, clicks=but_get_datasets.param.clicks),  # this callback cannot be moved, do I need it with clicks?\n",
    "                pn.bind(gecco.handle_upload_dataset, clicks=but_upload_dataset.param.clicks),\n",
    "                ],\n",
    "        main=[tabs,\n",
    "            ],\n",
    "        main_layout=None,\n",
    "        accent=ACCENT,\n",
    "    )\n",
    "    return template\n",
    "\n",
    "\n",
    "template = app()\n",
    "logger.info(\"Template created\")\n",
    "\n",
    "# serve the app\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    s = serve_app(template, env=env, name=\"GECCO_galaxy_runner\")\n",
    "else:\n",
    "    template.servable()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uncomment this if running if running ngrok tunnel which you want to quit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only use for the ngrok tunnel in GColab\n",
    "# close_server(s, env=env)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momicsdem",
   "language": "python",
   "name": "momicsdem"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "panel-cell-order": [
   "ed20ab75",
   "5d7b0436",
   "588e4c4d",
   "8ee8b5ae",
   "a5ba5dba",
   "9e01a365",
   "3886b6e4",
   "d70ab7ec"
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
