{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f5426a2",
   "metadata": {},
   "source": [
    "## Example of RO-Crates ingestion and fuseki SparQL\n",
    "- set up local SparQL DB\n",
    "- get several RO-Crates published\n",
    "- create a graph and add it to the SparQL\n",
    "\n",
    "This NB should mirror the [tutorial](https://py-ualg.github.io/biohap/biodata_pt/training_2511.html) working directly on `fuseki`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b86cd6",
   "metadata": {},
   "source": [
    "### Fuseki setup\n",
    "Fuseki is a java SparQL endpoint server.\n",
    "\n",
    "Get java\n",
    "```\n",
    "sudo apt update\n",
    "sudo apt install -y openjdk-17-jre-headless\n",
    "java -version\n",
    "```\n",
    "\n",
    "- download the zip file from https://jena.apache.org/download/index.cgi\n",
    "- run the fuseki server script\n",
    "```\n",
    "cd apache-jena-fuseki-5.6.0/\n",
    "./fuseki-server\n",
    "```\n",
    "the server is at\n",
    "http://localhost:3030/#/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c28d496b",
   "metadata": {},
   "source": [
    "### Preparation of a dataset\n",
    "- I want to keep this in pure python\n",
    "- use `rdflib rdflib-jsonld requests`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "035908fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import os\n",
    "import json\n",
    "import re\n",
    "from rdflib import Graph\n",
    "from urllib.parse import quote_plus, urljoin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc99c978",
   "metadata": {},
   "source": [
    "### Methods to get RO-Crates from GH\n",
    "- easier is to clone the `https://github.com/emo-bon/analysis-results-cluster-01-crate` repo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4059c78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def jsonld_to_rdflib(jsonld_text, base=None):\n",
    "    \"\"\"\n",
    "    Parse JSON-LD text into an rdflib.Graph and return the Graph.\n",
    "    \"\"\"\n",
    "    g = Graph()\n",
    "    # rdflib accepts a JSON-LD string as input; base is optional\n",
    "    g.parse(data=jsonld_text, format=\"json-ld\", publicID=base)\n",
    "    return g\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aa5037d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "rocrate_folder = \"/home/david-palecek/coding/ro-crates/analysis-results-cluster-01-crate/\"\n",
    "\n",
    "with open(os.path.join(rocrate_folder, \"EMOBON_AAOT_Wa_66-ro-crate.ttl\"), \"r\") as f:\n",
    "    ttl_content = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c80da2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the Graph Store Protocol endpoint for Fuseki\n",
    "fuseki_url = \"http://localhost:3030/rocrate_python/data\"\n",
    "headers = {\"Content-Type\": \"text/turtle\"}\n",
    "\n",
    "resp = requests.put(fuseki_url, data=ttl_content.encode(\"utf-8\"), headers=headers, timeout=60)\n",
    "\n",
    "# raise for HTTP error\n",
    "try:\n",
    "    resp.raise_for_status()\n",
    "except requests.HTTPError as e:\n",
    "    raise RuntimeError(f\"Upload failed: {resp.status_code} {resp.text}\") from e"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48505d04",
   "metadata": {},
   "source": [
    "### Loop over `ttl`s in a folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e5ffd15e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading (append): EMOBON_AAOT_Wa_67-ro-crate.ttl\n",
      "Uploading (append): EMOBON_HCMR-1_Wa_6-ro-crate.ttl\n",
      "Uploading (append): EMOBON_AAOT_Wa_66-ro-crate.ttl\n",
      "Uploading (append): EMOBON_OSD74_Wa_21-ro-crate.ttl\n",
      "All files uploaded (appended).\n"
     ]
    }
   ],
   "source": [
    "fuseki_host = \"http://localhost:3030\"\n",
    "dataset = \"rocrate_python\"\n",
    "gsp_endpoint = f\"{fuseki_host}/{dataset}/data\"\n",
    "headers = {\"Content-Type\": \"text/turtle\"}     # sending Turtle\n",
    "\n",
    "for filename in os.listdir(rocrate_folder):\n",
    "    if not filename.endswith(\"-ro-crate.ttl\"):\n",
    "        continue\n",
    "\n",
    "    path = os.path.join(rocrate_folder, filename)\n",
    "    print(\"Uploading (append):\", filename)\n",
    "    with open(path, \"rb\") as f:\n",
    "        ttl_bytes = f.read()\n",
    "\n",
    "    # POST appends triples to the named graph (PUT would replace)\n",
    "    resp = requests.post(\n",
    "        gsp_endpoint,\n",
    "        data=ttl_bytes,\n",
    "        headers=headers,\n",
    "        timeout=60,\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        resp.raise_for_status()\n",
    "    except requests.HTTPError as e:\n",
    "        print(\"Upload failed for\", filename, resp.status_code)\n",
    "        print(\"Server response:\", resp.text[:1000])\n",
    "        raise\n",
    "\n",
    "print(\"All files uploaded (appended).\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28f194e0",
   "metadata": {},
   "source": [
    "### sidetrack to dvc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1be0886",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/.github/workflows/rocrate_to_pages.yml', '/EMOBON_AAOT_Wa_66-ro-crate.ttl', '/EMOBON_AAOT_Wa_66-ro-crate/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged.cmsearch.all.tblout.deoverlapped.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged.fasta.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged.motus.tsv.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged.qc_summary', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged.unfiltered_fasta.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged_CDS.faa.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB.merged_CDS.ffn.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB_AAEPOSDA_1_1_HVWGWDSX5.UDI100_clean.fastq.trimmed.fasta.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB_AAEPOSDA_1_1_HVWGWDSX5.UDI100_clean.fastq.trimmed.qc_summary', '/EMOBON_AAOT_Wa_66-ro-crate/DBB_AAEPOSDA_1_2_HVWGWDSX5.UDI100_clean.fastq.trimmed.fasta.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/DBB_AAEPOSDA_1_2_HVWGWDSX5.UDI100_clean.fastq.trimmed.qc_summary', '/EMOBON_AAOT_Wa_66-ro-crate/config.yml', '/EMOBON_AAOT_Wa_66-ro-crate/fastp.html', '/EMOBON_AAOT_Wa_66-ro-crate/final.contigs.fa.bz2', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.emapper.summary.eggnog', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.hmm.tsv.gz', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.summary.go', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.summary.go_slim', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.summary.ips', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.summary.ko', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged.summary.pfam', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/DBB.merged_CDS.I5.tsv.gz', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/functional-annotation.ttl', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/go.stats', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/interproscan.stats', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/ko.stats', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/orf.stats', '/EMOBON_AAOT_Wa_66-ro-crate/functional-annotation/stats/pfam.stats', '/EMOBON_AAOT_Wa_66-ro-crate/ro-crate-metadata.json', '/EMOBON_AAOT_Wa_66-ro-crate/run.yml', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/5_8S.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/Bacteria_large_SRP.RF01854.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/Bacteria_small_SRP.RF00169.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/LSU_rRNA_archaea.RF02540.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/LSU_rRNA_bacteria.RF02541.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/LSU_rRNA_eukarya.RF02543.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/RNaseP_arch.RF00373.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/RNaseP_bact_a.RF00010.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/SSU_rRNA_archaea.RF01959.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/SSU_rRNA_bacteria.RF00177.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/SSU_rRNA_eukarya.RF01960.fa.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/alpha_tmRNA.RF01849.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/cyano_tmRNA.RF01851.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/tRNA-Sec.RF01852.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/tRNA.RF00005.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/sequence-categorisation/tmRNA.RF00023.fasta.gz', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.gz', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.tsv', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.txt', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_hdf5.biom', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_json.biom', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/LSU-taxonomy-summary.ttl', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/LSU/krona.html', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/RNA-counts', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/.gitignore', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.gz', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.tsv', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.txt', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_hdf5.biom', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_json.biom', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/krona.html', '/EMOBON_AAOT_Wa_67-ro-crate.ttl', '/EMOBON_AAOT_Wa_67-ro-crate/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged.cmsearch.all.tblout.deoverlapped.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged.fasta.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged.motus.tsv.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged.qc_summary', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged.unfiltered_fasta.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged_CDS.faa.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB.merged_CDS.ffn.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB_AAEQOSDA_1_1_HVWGWDSX5.UDI106_clean.fastq.trimmed.fasta.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB_AAEQOSDA_1_1_HVWGWDSX5.UDI106_clean.fastq.trimmed.qc_summary', '/EMOBON_AAOT_Wa_67-ro-crate/DBB_AAEQOSDA_1_2_HVWGWDSX5.UDI106_clean.fastq.trimmed.fasta.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/DBB_AAEQOSDA_1_2_HVWGWDSX5.UDI106_clean.fastq.trimmed.qc_summary', '/EMOBON_AAOT_Wa_67-ro-crate/config.yml', '/EMOBON_AAOT_Wa_67-ro-crate/fastp.html', '/EMOBON_AAOT_Wa_67-ro-crate/final.contigs.fa.bz2', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.emapper.summary.eggnog', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.hmm.tsv.gz', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.summary.go', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.summary.go_slim', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.summary.ips', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.summary.ko', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged.summary.pfam', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/DBB.merged_CDS.I5.tsv.gz', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/functional-annotation.ttl', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/go.stats', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/interproscan.stats', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/ko.stats', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/orf.stats', '/EMOBON_AAOT_Wa_67-ro-crate/functional-annotation/stats/pfam.stats', '/EMOBON_AAOT_Wa_67-ro-crate/ro-crate-metadata.json', '/EMOBON_AAOT_Wa_67-ro-crate/run.yml', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/5_8S.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/Bacteria_large_SRP.RF01854.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/Bacteria_small_SRP.RF00169.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/LSU_rRNA_archaea.RF02540.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/LSU_rRNA_bacteria.RF02541.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/LSU_rRNA_eukarya.RF02543.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/RNaseP_arch.RF00373.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/RNaseP_bact_a.RF00010.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/RNaseP_bact_b.RF00011.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/SSU_rRNA_archaea.RF01959.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/SSU_rRNA_bacteria.RF00177.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/SSU_rRNA_eukarya.RF01960.fa.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/alpha_tmRNA.RF01849.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/cyano_tmRNA.RF01851.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/tRNA-Sec.RF01852.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/tRNA.RF00005.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/sequence-categorisation/tmRNA.RF00023.fasta.gz', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.gz', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.tsv', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.txt', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_hdf5.biom', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_json.biom', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/LSU-taxonomy-summary.ttl', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/LSU/krona.html', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/RNA-counts', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/.gitignore', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.gz', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.tsv', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.txt', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_hdf5.biom', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_json.biom', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/krona.html', '/EMOBON_HCMR-1_Wa_6-ro-crate.ttl', '/EMOBON_HCMR-1_Wa_6-ro-crate/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged.cmsearch.all.tblout.deoverlapped.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged.fasta.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged.motus.tsv.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged.qc_summary', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged.unfiltered_fasta.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged_CDS.faa.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB.merged_CDS.ffn.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB_AABLOSDA_1_1_HMNJKDSX3.UDI279_clean.fastq.trimmed.fasta.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB_AABLOSDA_1_1_HMNJKDSX3.UDI279_clean.fastq.trimmed.qc_summary', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB_AABLOSDA_1_2_HMNJKDSX3.UDI279_clean.fastq.trimmed.fasta.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/DBB_AABLOSDA_1_2_HMNJKDSX3.UDI279_clean.fastq.trimmed.qc_summary', '/EMOBON_HCMR-1_Wa_6-ro-crate/config.yml', '/EMOBON_HCMR-1_Wa_6-ro-crate/fastp.html', '/EMOBON_HCMR-1_Wa_6-ro-crate/final.contigs.fa.bz2', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.emapper.summary.eggnog', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.hmm.tsv.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.summary.go', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.summary.go_slim', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.summary.ips', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.summary.ko', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged.summary.pfam', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/DBB.merged_CDS.I5.tsv.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/functional-annotation.ttl', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/go.stats', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/interproscan.stats', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/ko.stats', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/orf.stats', '/EMOBON_HCMR-1_Wa_6-ro-crate/functional-annotation/stats/pfam.stats', '/EMOBON_HCMR-1_Wa_6-ro-crate/ro-crate-metadata.json', '/EMOBON_HCMR-1_Wa_6-ro-crate/run.yml', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/5_8S.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/Bacteria_small_SRP.RF00169.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/LSU_rRNA_archaea.RF02540.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/LSU_rRNA_bacteria.RF02541.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/LSU_rRNA_eukarya.RF02543.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/Metazoa_SRP.RF00017.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/RNaseP_bact_a.RF00010.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/RNase_MRP.RF00030.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/SSU_rRNA_archaea.RF01959.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/SSU_rRNA_bacteria.RF00177.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/SSU_rRNA_eukarya.RF01960.fa.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/alpha_tmRNA.RF01849.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/cyano_tmRNA.RF01851.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/tRNA-Sec.RF01852.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/tRNA.RF00005.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/sequence-categorisation/tmRNA.RF00023.fasta.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.tsv', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.txt', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_hdf5.biom', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_json.biom', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/LSU-taxonomy-summary.ttl', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/LSU/krona.html', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/RNA-counts', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/.gitignore', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.gz', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.tsv', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.txt', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_hdf5.biom', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_json.biom', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/krona.html', '/EMOBON_OSD74_Wa_21-ro-crate.ttl', '/EMOBON_OSD74_Wa_21-ro-crate/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/.ro-crate-metadata.json.swp', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged.cmsearch.all.tblout.deoverlapped.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged.fasta.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged.motus.tsv.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged.qc_summary', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged.unfiltered_fasta.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged_CDS.faa.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB.merged_CDS.ffn.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB_AAHDOSDA_2_1_H5H2YDSX7.UDI239_clean.fastq.trimmed.fasta.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB_AAHDOSDA_2_1_H5H2YDSX7.UDI239_clean.fastq.trimmed.qc_summary', '/EMOBON_OSD74_Wa_21-ro-crate/DBB_AAHDOSDA_2_2_H5H2YDSX7.UDI239_clean.fastq.trimmed.fasta.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/DBB_AAHDOSDA_2_2_H5H2YDSX7.UDI239_clean.fastq.trimmed.qc_summary', '/EMOBON_OSD74_Wa_21-ro-crate/config.yml', '/EMOBON_OSD74_Wa_21-ro-crate/fastp.html', '/EMOBON_OSD74_Wa_21-ro-crate/final.contigs.fa.bz2', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.emapper.summary.eggnog', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.hmm.tsv.gz', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.summary.go', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.summary.go_slim', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.summary.ips', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.summary.ko', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged.summary.pfam', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/DBB.merged_CDS.I5.tsv.gz', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/functional-annotation.ttl', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/go.stats', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/interproscan.stats', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/ko.stats', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/orf.stats', '/EMOBON_OSD74_Wa_21-ro-crate/functional-annotation/stats/pfam.stats', '/EMOBON_OSD74_Wa_21-ro-crate/ro-crate-metadata.json', '/EMOBON_OSD74_Wa_21-ro-crate/run.yml', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/5S.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/5_8S.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/Bacteria_large_SRP.RF01854.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/Bacteria_small_SRP.RF00169.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/LSU_rRNA_archaea.RF02540.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/LSU_rRNA_bacteria.RF02541.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/LSU_rRNA_eukarya.RF02543.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/RNaseP_arch.RF00373.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/RNaseP_bact_a.RF00010.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/RNaseP_bact_b.RF00011.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/RNase_MRP.RF00030.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/SSU_rRNA_archaea.RF01959.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/SSU_rRNA_bacteria.RF00177.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/SSU_rRNA_eukarya.RF01960.fa.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/alpha_tmRNA.RF01849.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/cyano_tmRNA.RF01851.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/tRNA-Sec.RF01852.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/tRNA.RF00005.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/sequence-categorisation/tmRNA.RF00023.fasta.gz', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.gz', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.tsv', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq.txt', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_hdf5.biom', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/DBB.merged_LSU.fasta.mseq_json.biom', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/LSU-taxonomy-summary.ttl', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/LSU/krona.html', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/RNA-counts', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/.gitignore', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.gz', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.tsv', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq.txt', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_hdf5.biom', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/DBB.merged_SSU.fasta.mseq_json.biom', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/krona.html', '/README.md']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "os.environ['AWS_NO_SIGN_REQUEST'] = '1'\n",
    "os.environ.pop('AWS_PROFILE', None)   # avoid using a missing profile\n",
    "\n",
    "import dvc.api\n",
    "from dvc.api import DVCFileSystem\n",
    "import boto3\n",
    "from botocore import UNSIGNED\n",
    "from botocore.client import Config\n",
    "\n",
    "dvc_fs = DVCFileSystem(rocrate_folder)\n",
    "files = list(dvc_fs.find(\"/\", detail=False))\n",
    "print(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "d72d9743",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl', '/EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "filt_files = [f for f in files if f.endswith(\"SSU-taxonomy-summary.ttl\")]\n",
    "print(filt_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "3999b61b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_single_file_s3(repo_folder, path: str):\n",
    "    url = dvc.api.get_url(\n",
    "        path=path,\n",
    "        repo=repo_folder,\n",
    "    )\n",
    "    # Custom S3 endpoint (non-AWS)\n",
    "    endpoint_url = \"https://s3.mesocentre.uca.fr\"\n",
    "    bucket = \"mgf-data-products\"\n",
    "    key = url.split(f\"{bucket}/\")[-1]  # extract key from URL\n",
    "\n",
    "    # Create S3 client that does NOT require credentials\n",
    "    s3 = boto3.client(\n",
    "        \"s3\",\n",
    "        endpoint_url=endpoint_url,\n",
    "        config=Config(signature_version=UNSIGNED),\n",
    "    )\n",
    "\n",
    "    # Fetch the object\n",
    "    obj = s3.get_object(Bucket=bucket, Key=key)\n",
    "\n",
    "    # Read contents into memory\n",
    "    data = obj[\"Body\"].read()\n",
    "\n",
    "    filename = path.split(\"/\")[1] + \"_\" + path.split(\"/\")[-1]\n",
    "    # Save to a local file if needed\n",
    "    with open(filename, \"wb\") as f:\n",
    "        f.write(data)\n",
    "\n",
    "    print(\"Downloaded\", len(data), \"bytes from\", endpoint_url)\n",
    "    return path.split(\"/\")[1], data\n",
    "def create_upload_ds(name, contents):\n",
    "    fuseki_admin_url = \"http://localhost:3030/$/datasets\"\n",
    "    # Form data\n",
    "    form_data = {\n",
    "        \"dbName\": name,   # dataset name\n",
    "        \"dbType\": \"mem\"       # in-memory\n",
    "    }\n",
    "\n",
    "    # Headers to enforce form encoding\n",
    "    headers = {\n",
    "        \"Content-Type\": \"application/x-www-form-urlencoded\"\n",
    "    }\n",
    "\n",
    "    # POST request\n",
    "    resp = requests.post(fuseki_admin_url, data=form_data, headers=headers)\n",
    "    print(resp.status_code, resp.text)\n",
    "\n",
    "    # Now upload the data to the named graph\n",
    "    gsp_endpoint = f\"http://localhost:3030/{name}/data\"\n",
    "    headers = {\"Content-Type\": \"text/turtle\"}     # sending Turtle\n",
    "    resp = requests.post(\n",
    "        gsp_endpoint,\n",
    "        data=contents,\n",
    "        headers=headers,\n",
    "        timeout=60,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5070b6",
   "metadata": {},
   "source": [
    "### Get all the tables and upload them to separate graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5822c268",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded 233217 bytes from https://s3.mesocentre.uca.fr\n",
      "Uploading (append): /EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl\n",
      " 233217 bytes from https://s3.mesocentre.uca.fr\n",
      "Uploading (append): /EMOBON_AAOT_Wa_66-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl\n",
      "200 \n",
      "Downloaded 240253 bytes from https://s3.mesocentre.uca.fr\n",
      "Uploading (append): /EMOBON_AAOT_Wa_67-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl\n",
      "200 \n",
      "Downloaded 215932 bytes from https://s3.mesocentre.uca.fr\n",
      "Uploading (append): /EMOBON_HCMR-1_Wa_6-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl\n",
      "200 \n",
      "Downloaded 628259 bytes from https://s3.mesocentre.uca.fr\n",
      "Uploading (append): /EMOBON_OSD74_Wa_21-ro-crate/taxonomy-summary/SSU/SSU-taxonomy-summary.ttl\n",
      "200 \n"
     ]
    }
   ],
   "source": [
    "for file in filt_files:\n",
    "    name, contents = get_single_file_s3(rocrate_folder, file)\n",
    "    print(\"Uploading (append):\", file)\n",
    "\n",
    "    create_upload_ds(name, contents)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5475b99",
   "metadata": {},
   "source": [
    "### Serialization to .ttl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3c5e7272",
   "metadata": {},
   "outputs": [],
   "source": [
    "def serialize_rocrate_to_ttl(repo_url, output_file_path):\n",
    "    \"\"\"\n",
    "    Fetch RO-Crate metadata from GitHub, parse it, and serialize to a TTL file.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    repo_url : str\n",
    "        GitHub repo URL or raw file URL to ro-crate-metadata.json\n",
    "    output_file_path : str\n",
    "        Path where the TTL file should be saved\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        Path to the created TTL file\n",
    "    \"\"\"\n",
    "    # Fetch and parse the RO-Crate metadata\n",
    "    jsonld_text = fetch_rocrate_json_from_github(repo_url)\n",
    "    g = jsonld_to_rdflib(jsonld_text)\n",
    "    \n",
    "    # Serialize to TTL format and save to file\n",
    "    ttl_content = g.serialize(format='turtle')\n",
    "    \n",
    "    with open(output_file_path, 'w', encoding='utf-8') as f:\n",
    "        f.write(ttl_content)\n",
    "    \n",
    "    print(f\"RO-Crate serialized to TTL file: {output_file_path}\")\n",
    "    print(f\"Graph contains {len(g)} triples\")\n",
    "    \n",
    "    return output_file_path\n",
    "\n",
    "\n",
    "def serialize_multiple_rocrates_to_ttl(repo_urls, output_file_path):\n",
    "    \"\"\"\n",
    "    Fetch multiple RO-Crate metadata files from GitHub, parse them, \n",
    "    and serialize all to a single TTL file.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    repo_urls : list\n",
    "        List of GitHub repo URLs or raw file URLs to ro-crate-metadata.json\n",
    "    output_file_path : str\n",
    "        Path where the combined TTL file should be saved\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        Path to the created TTL file\n",
    "    \"\"\"\n",
    "    # Create a combined graph\n",
    "    combined_graph = Graph()\n",
    "    \n",
    "    for i, repo_url in enumerate(repo_urls):\n",
    "        try:\n",
    "            print(f\"Processing {i+1}/{len(repo_urls)}: {repo_url}\")\n",
    "            jsonld_text = fetch_rocrate_json_from_github(repo_url)\n",
    "            g = jsonld_to_rdflib(jsonld_text)\n",
    "            \n",
    "            # Add all triples from this graph to the combined graph\n",
    "            for triple in g:\n",
    "                combined_graph.add(triple)\n",
    "                \n",
    "            print(f\"  Added {len(g)} triples\")\n",
    "            \n",
    "        except Exception as e:\n",
    "            print(f\"  FAILED to process {repo_url}: {e}\")\n",
    "    \n",
    "    # Serialize combined graph to TTL format and save to file\n",
    "    ttl_content = combined_graph.serialize(format='turtle')\n",
    "    \n",
    "    with open(output_file_path, 'w', encoding='utf-8') as f:\n",
    "        f.write(ttl_content)\n",
    "    \n",
    "    print(f\"\\nCombined RO-Crates serialized to TTL file: {output_file_path}\")\n",
    "    print(f\"Combined graph contains {len(combined_graph)} triples\")\n",
    "    \n",
    "    return output_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2239823a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RO-Crate serialized to TTL file: emobon_bpns_so_34.ttl\n",
      "Graph contains 481 triples\n",
      "Single RO-Crate saved to: emobon_bpns_so_34.ttl\n",
      "\n",
      "==================================================\n",
      "\n",
      "Processing 1/2: https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json\n",
      "  Added 481 triples\n",
      "Processing 2/2: https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_1-ro-crate/ro-crate-metadata.json\n",
      "  Added 486 triples\n",
      "\n",
      "Combined RO-Crates serialized to TTL file: combined_emobon_rocrates.ttl\n",
      "Combined graph contains 655 triples\n",
      "Combined RO-Crates saved to: combined_emobon_rocrates.ttl\n"
     ]
    }
   ],
   "source": [
    "# Example usage of TTL serialization functions\n",
    "\n",
    "# Serialize a single RO-Crate to TTL\n",
    "single_repo = 'https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json'\n",
    "\n",
    "try:\n",
    "    ttl_file = serialize_rocrate_to_ttl(single_repo, \"emobon_bpns_so_34.ttl\")\n",
    "    print(f\"Single RO-Crate saved to: {ttl_file}\")\n",
    "except Exception as e:\n",
    "    print(f\"Error serializing single RO-Crate: {e}\")\n",
    "\n",
    "print(\"\\n\" + \"=\"*50 + \"\\n\")\n",
    "\n",
    "# Serialize multiple RO-Crates to a combined TTL file\n",
    "repos = [\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json\",\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_1-ro-crate/ro-crate-metadata.json\",\n",
    "    # \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_11-ro-crate/ro-crate-metadata.json\",\n",
    "    # 'https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_OOB_So_25-ro-crate/ro-crate-metadata.json',\n",
    "    # \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_RFormosa_So_11-ro-crate/ro-crate-metadata.json\"\n",
    "]\n",
    "\n",
    "try:\n",
    "    combined_ttl = serialize_multiple_rocrates_to_ttl(repos, \"combined_emobon_rocrates.ttl\")\n",
    "    print(f\"Combined RO-Crates saved to: {combined_ttl}\")\n",
    "except Exception as e:\n",
    "    print(f\"Error serializing multiple RO-Crates: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458d1987",
   "metadata": {},
   "source": [
    "### Upload of the combined graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e0cf6c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(combined_ttl, 'r', encoding='utf-8') as f:\n",
    "    ttl_content = f.read()\n",
    "\n",
    "uri = \"http://example.org/graphs/emobon_combined\"\n",
    "# Use the Graph Store Protocol endpoint for Fuseki\n",
    "gsp_url = \"http://localhost:3030/rocrate/data?graph=\" + uri\n",
    "\n",
    "headers = {\"Content-Type\": \"text/turtle\"}\n",
    "\n",
    "resp = requests.put(gsp_url, data=ttl_content.encode(\"utf-8\"), headers=headers, timeout=60)\n",
    "\n",
    "# raise for HTTP error\n",
    "try:\n",
    "    resp.raise_for_status()\n",
    "except requests.HTTPError as e:\n",
    "    raise RuntimeError(f\"Upload failed: {resp.status_code} {resp.text}\") from e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "944ab982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'http://example.org/graphs/emobon_combined'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uri"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb5182f5",
   "metadata": {},
   "source": [
    "### Original single graph upload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d864718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Upload successful: 200\n"
     ]
    }
   ],
   "source": [
    "repo = 'https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json'\n",
    "fuseki_base = \"http://localhost:3030/rocrate\"\n",
    "try:\n",
    "    r = process_github_rocrate(\n",
    "        repo,\n",
    "        fuseki_base,\n",
    "        graph_uri=\"http://example.org/rocrate/EMOBON_BPNS_So_34\",\n",
    "        fmt=\"turtle\")\n",
    "    print(\"Upload successful:\", r.status_code)\n",
    "except Exception as e:\n",
    "    print(\"Error:\", e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ffcd4897",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'head': {'vars': ['c']}, 'results': {'bindings': [{'c': {'type': 'literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#integer', 'value': '481'}}]}}\n"
     ]
    }
   ],
   "source": [
    "q = \"SELECT (COUNT(*) AS ?c) WHERE { ?s ?p ?o }\"\n",
    "r = requests.get(\"http://localhost:3030/rocrate/query\", params={\"query\": q}, headers={\"Accept\": \"application/sparql-results+json\"})\n",
    "print(r.json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b39f7af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sparql_json_to_df(sparql_json):\n",
    "    \"\"\"\n",
    "    Convert a SPARQL SELECT query JSON result to a pandas DataFrame.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    sparql_json : dict\n",
    "        JSON returned by Fuseki / SPARQL endpoint with Accept: application/sparql-results+json\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "    \"\"\"\n",
    "    vars_ = sparql_json.get(\"head\", {}).get(\"vars\", [])\n",
    "    rows = []\n",
    "\n",
    "    for binding in sparql_json.get(\"results\", {}).get(\"bindings\", []):\n",
    "        row = {}\n",
    "        for var in vars_:\n",
    "            # Some results might not bind all variables\n",
    "            if var in binding:\n",
    "                row[var] = binding[var][\"value\"]\n",
    "            else:\n",
    "                row[var] = None\n",
    "        rows.append(row)\n",
    "\n",
    "    df = pd.DataFrame(rows, columns=vars_)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "af2f709d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     c\n",
      "0  481\n"
     ]
    }
   ],
   "source": [
    "df = sparql_json_to_df(r.json())\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c553fff3",
   "metadata": {},
   "source": [
    "### Add more than one `repository`\n",
    "- ie. several emo-bon ro-crates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1b9dcd6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json -> 201\n",
      "https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_1-ro-crate/ro-crate-metadata.json -> 201\n",
      "https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_11-ro-crate/ro-crate-metadata.json -> 201\n",
      "https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_OOB_So_25-ro-crate/ro-crate-metadata.json -> 201\n",
      "https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_RFormosa_So_11-ro-crate/ro-crate-metadata.json -> 201\n"
     ]
    }
   ],
   "source": [
    "repos = [\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_BPNS_So_34-ro-crate/ro-crate-metadata.json\",\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_1-ro-crate/ro-crate-metadata.json\",\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_NRMCB_So_11-ro-crate/ro-crate-metadata.json\",\n",
    "    'https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_OOB_So_25-ro-crate/ro-crate-metadata.json',\n",
    "    \"https://raw.githubusercontent.com/emo-bon/analysis-results-cluster-01-crate/refs/heads/main/EMOBON_RFormosa_So_11-ro-crate/ro-crate-metadata.json\"\n",
    "]\n",
    "for r in repos:\n",
    "    try:\n",
    "        resp = process_github_rocrate(\n",
    "            r, \"http://localhost:3030/rocrate\",\n",
    "            graph_uri=f\"http://example.org/rocrate/{r.split('/')[-2]}\")\n",
    "        print(r, \"->\", resp.status_code)\n",
    "    except Exception as e:\n",
    "        print(\"FAILED\", r, e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "434ed646",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     c\n",
      "0  481\n"
     ]
    }
   ],
   "source": [
    "q = \"SELECT (COUNT(*) AS ?c) WHERE { ?s ?p ?o }\"\n",
    "r = requests.get(\"http://localhost:3030/rocrate/query\", params={\"query\": q}, headers={\"Accept\": \"application/sparql-results+json\"})\n",
    "df = sparql_json_to_df(r.json())\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38aff9ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momics-demos",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
