{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Querry data based on ro-Crates\n",
    "\n",
    "**Steps:** (for each metaGOflow step)\n",
    "1. Acess ro-crate metadata file and extract needed data sources\n",
    "2. Get the data\n",
    "3. Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import io\n",
    "import gc\n",
    "import logging\n",
    "import psutil\n",
    "\n",
    "from IPython import get_ipython\n",
    "logger = logging.getLogger(name=\"Quality Control app\")\n",
    "NUMBER_PERMUTATIONS = 999\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    print('Setting Google colab, you will need a ngrok account to make the dashboard display over the tunnel. \\\n",
    "    https://ngrok.com/')\n",
    "    # clone the momics-demos repository to use it to load data\n",
    "    try:\n",
    "        os.system('git clone https://github.com/palec87/momics-demos.git')\n",
    "        logger.info(f\"Repository cloned\")\n",
    "    except OSError as e:\n",
    "        logger.info(f\"An error occurred while cloning the repository: {e}\")\n",
    "\n",
    "    sys.path.insert(0,'/content/momics-demos')\n",
    "\n",
    "    # this step takes time beacause of many dependencies\n",
    "    os.system('pip install marine-omics')\n",
    "\n",
    "elif psutil.users() == []:\n",
    "    logger.info(\"Binder\")\n",
    "    NUMBER_PERMUTATIONS = 29  # permanova extremely slow on binder, therefore a change here\n",
    "else:\n",
    "    logger.info(\"Local\")\n",
    "\n",
    "from momics.utils import (\n",
    "    memory_load, reconfig_logger,\n",
    "    init_setup, get_notebook_environment,\n",
    ")\n",
    "\n",
    "# Set up logging\n",
    "reconfig_logger()\n",
    "\n",
    "# Determine the notebook environment\n",
    "env = get_notebook_environment()\n",
    "\n",
    "init_setup()\n",
    "logger.info(f\"Environment: {env}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This needs to be repeated here for the Pannel dashboard to work, WEIRD\n",
    "# TODO: report as possible bug\n",
    "import sys\n",
    "import os\n",
    "import io\n",
    "import requests\n",
    "import html\n",
    "\n",
    "from functools import partial\n",
    "import panel as pn\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "# All low level functions are imported from the momics package\n",
    "from momics.loader import (\n",
    "    get_rocrate_metadata_gh,\n",
    "    extract_data_by_name,\n",
    ")\n",
    "from momics.panel_utils import (\n",
    "    create_indicators_diversity,\n",
    "    serve_app,\n",
    "    close_server,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'google.colab' in str(get_ipython()):\n",
    "    root_folder = os.path.abspath(os.path.join('/content/momics-demos'))\n",
    "else:\n",
    "    root_folder = os.path.abspath(os.path.join('../'))\n",
    "\n",
    "assets_folder = os.path.join(root_folder, 'assets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dropdown and tabs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "archives = [\n",
    "    \"EMOBON_MBAL4_Wa_2\",\n",
    "    \"EMOBON_NRMCB_So_1\",\n",
    "    \"EMOBON_NRMCB_So_7\",\n",
    "    \"EMOBON_OSD74_Wa_2\",\n",
    "    \"EMOBON_PiEGetxo_Wa_4\",\n",
    "    \"EMOBON_RFormosa_Wa_6\",\n",
    "    \"EMOBON_ROSKOGO_So_1\",\n",
    "    \"EMOBON_VB_Wa_93\",\n",
    "]\n",
    "select_archive = pn.widgets.Select(\n",
    "    name=\"Archive\",\n",
    "    options= archives,\n",
    "    value=archives[0],\n",
    "    description=\"Select an archive for inspection\",\n",
    ")\n",
    "\n",
    "fastap_tab = pn.pane.HTML(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    "    sizing_mode=\"stretch_both\",\n",
    ")\n",
    "metagoflow_tab1 = pn.pane.Str(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    "    sizing_mode=\"stretch_both\",\n",
    ")\n",
    "metagoflow_tab2 = pn.pane.Str(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    "    sizing_mode=\"stretch_both\",\n",
    ")\n",
    "reads_qc_tab = pn.pane.Markdown(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    "    hard_line_break=True,\n",
    ")\n",
    "functional_tab = pn.pane.Str(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    ")\n",
    "krona1 = pn.pane.HTML(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    ")\n",
    "krona2 = pn.pane.HTML(\n",
    "    \"\"\"No data loaded yet. Please select an archive from the dropdown menu\"\"\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bindings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension('mathjax', 'plotly')\n",
    "\n",
    "def redraw_tabs(archive):\n",
    "    \"\"\"\n",
    "    Redraw the tabs with the selected archive\n",
    "    \"\"\"\n",
    "    if archive is None:\n",
    "        fastap_tab.object = \"No data loaded yet. Please select an archive from the dropdown menu\"\n",
    "        return\n",
    "\n",
    "    # Get the metadata from the ro-crate\n",
    "    metadata = get_rocrate_metadata_gh(archive)\n",
    "    data = extract_data_by_name(metadata, name='FASTP analysis of raw sequence data')\n",
    "    r = requests.get(data['downloadUrl'])\n",
    "    escaped_content = html.escape(r.content.decode('utf-8'))\n",
    "    iframe = f'<iframe srcdoc=\"{escaped_content}\" style=\"height:100%; width:900px;\"></iframe>'\n",
    "    fastap_tab.object = iframe\n",
    "\n",
    "    data = extract_data_by_name(metadata, name='MetaGOflow YAML configuration file')\n",
    "    r = requests.get(data['downloadUrl'])\n",
    "    metagoflow_tab1.object = r.content.decode('utf-8')\n",
    "\n",
    "    data = extract_data_by_name(metadata, name='MetaGOflow configuration in YAML')\n",
    "    r = requests.get(data['downloadUrl'])\n",
    "    metagoflow_tab2.object = r.content.decode('utf-8')\n",
    "\n",
    "    data = extract_data_by_name(metadata, name='Trimmed reverse reads QC summary')\n",
    "    data2 = extract_data_by_name(metadata, name='Trimmed forward reads QC summary')\n",
    "    data3 = extract_data_by_name(metadata, name='QC summary of merged reads')\n",
    "    r = requests.get(data['downloadUrl']).content.decode('utf-8')\n",
    "    r2 = requests.get(data2['downloadUrl']).content.decode('utf-8')\n",
    "    r3 = requests.get(data3['downloadUrl']).content.decode('utf-8')\n",
    "    reads_qc_tab.object = \"\"\"# Trimmed reads\\n## Forward reads\\n\"\"\" + r2 + \"\"\"\\n## Reverse reads\\n\"\"\" + r + \"\"\"\\n# Merged reads\\n\"\"\" + r3\n",
    "\n",
    "    data = extract_data_by_name(metadata, name='Geno Ontology summary statistics')\n",
    "    data2 = extract_data_by_name(metadata, name='InterProScan summary statistics')\n",
    "    data3 = extract_data_by_name(metadata, name='Kegg Ontology summary statistics')\n",
    "    data4 = extract_data_by_name(metadata, name='ORF summary statistics')\n",
    "    data5 = extract_data_by_name(metadata, name=\"Numbers of RNA's counted\")\n",
    "    data6 = extract_data_by_name(metadata, name='Merged contigs KO summary')\n",
    "    r = requests.get(data['downloadUrl']).content.decode('utf-8')\n",
    "    r2 = requests.get(data2['downloadUrl']).content.decode('utf-8')\n",
    "    r3 = requests.get(data3['downloadUrl']).content.decode('utf-8')\n",
    "    r4 = requests.get(data4['downloadUrl']).content.decode('utf-8')\n",
    "    r5 = requests.get(data5['downloadUrl']).content.decode('utf-8')\n",
    "    r6 = requests.get(data6['downloadUrl']).content.decode('utf-8')\n",
    "    functional_tab.object = concat_to_markdown([r, r2, r3, r4, r5, r6])\n",
    "\n",
    "    data = extract_data_by_name(metadata, name='Krona summary of LSU taxonomic inventory')\n",
    "    data2 = extract_data_by_name(metadata, name='Krona summary of SSU taxonomic inventory')\n",
    "    r = requests.get(data['downloadUrl']).content.decode('utf-8')\n",
    "    r2 = requests.get(data2['downloadUrl']).content.decode('utf-8')\n",
    "    escaped_content = html.escape(r)\n",
    "    escaped_content2 = html.escape(r2)\n",
    "    iframe = f'<iframe srcdoc=\"{escaped_content}\" style=\"height:700px; width:1000px;\"></iframe>'\n",
    "    krona1.object = iframe\n",
    "    iframe2 = f'<iframe srcdoc=\"{escaped_content2}\" style=\"height:700px; width:1000px;\"></iframe>'\n",
    "    krona2.object = iframe2\n",
    "\n",
    "\n",
    "def concat_to_markdown(data: list) -> str:\n",
    "    \"\"\"\n",
    "    Concatenate the data to a markdown string\n",
    "    \"\"\"\n",
    "    return \"\\n\".join([f\"{item}\\n\" for item in data])\n",
    "\n",
    "pn.bind(redraw_tabs,\n",
    "    archive=select_archive,\n",
    "    watch=True,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# populate the tabs with the first archive\n",
    "redraw_tabs(select_archive.value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Querry data based on ro-Crates\n",
    "\n",
    "**Steps:** (for each metaGOflow step)\n",
    "1. Acess ro-crate metadata file and extract needed data sources\n",
    "2. Get the data\n",
    "3. Visualize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## APP setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pn.extension(\"tabulator\")\n",
    "if 'google.colab' in str(get_ipython()):\n",
    "    pn.extension(comms='colab')\n",
    "ACCENT = \"teal\"\n",
    "\n",
    "styles = {\n",
    "    \"box-shadow\": \"rgba(50, 50, 93, 0.25) 0px 6px 12px -2px, rgba(0, 0, 0, 0.3) 0px 3px 7px -3px\",\n",
    "    \"border-radius\": \"4px\",\n",
    "    \"padding\": \"10px\",\n",
    "}\n",
    "\n",
    "# TODO: there is a bug in the panel library that does not allow to open png files, renoming does not help \n",
    "image = pn.pane.JPG(os.path.join(assets_folder, \"figs/metaGOflow_logo_italics.jpg\"),\n",
    "                    width=200,\n",
    "                    height=100,\n",
    "                    )\n",
    "tabs = pn.Tabs(\n",
    "    ('Fastap QC', fastap_tab),\n",
    "    ('Reads QC', reads_qc_tab),\n",
    "    ('Krona', pn.Column(\n",
    "        \"# Krona from LSU\",\n",
    "        krona1,\n",
    "        \"# Krona from SSU\",\n",
    "        krona2),\n",
    "    ),\n",
    "    ('Functional', functional_tab),\n",
    "    ('MGF yml 1', metagoflow_tab1),\n",
    "    ('MGF yml 2', metagoflow_tab2),\n",
    "    styles=styles,\n",
    "    margin=10\n",
    ")\n",
    "\n",
    "_, indicator_usage = create_indicators_diversity()\n",
    "\n",
    "def update_used_gb(event):\n",
    "    if not event:\n",
    "        return\n",
    "\n",
    "    used_gb, total_gb = memory_load()\n",
    "    indicator_usage.value = used_gb\n",
    "\n",
    "\n",
    "def app():\n",
    "    cb = pn.state.add_periodic_callback(\n",
    "        partial(update_used_gb, indicator_usage),\n",
    "        period=1000,\n",
    "        timeout=None,\n",
    "        )\n",
    "\n",
    "    toggle = pn.widgets.Toggle(\n",
    "        name='Toggle callback',\n",
    "        value=True,\n",
    "        button_type='success',)\n",
    "    toggle.link(cb, bidirectional=True, value='running')\n",
    "    template = pn.template.FastListTemplate(\n",
    "        title=\"Quality Control\",\n",
    "        sidebar=[image,\n",
    "                \"# Archive\", select_archive,\n",
    "                pn.layout.Divider(),\n",
    "                indicator_usage,\n",
    "                toggle,\n",
    "                ],\n",
    "        main=[pn.Column(\n",
    "                tabs,\n",
    "            )],\n",
    "        main_layout=None,\n",
    "        accent=ACCENT,\n",
    "    )\n",
    "    return template\n",
    "\n",
    "template = app()\n",
    "\n",
    "if 'google.colab' in str(get_ipython()):  \n",
    "    s = serve_app(template, env=env, name=\"quality_control\")\n",
    "else:\n",
    "    template.servable()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Uncomment this if running if running ngrok tunnel which you want to quit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# only use for the ngrok tunnel in GColab\n",
    "# close_server(s, env=env)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "momics-demos",
   "language": "python",
   "name": "momics-demos"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
